{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import pandas as pd, numpy as np, pyranges as pr\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a644ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.float_format', '{:.10f}'.format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189c8ada",
   "metadata": {},
   "source": [
    "[PyRanges v1.x GitHub](https://github.com/pyranges/pyranges_1.x)  \n",
    "[PyRanges v1.x Docs](https://pyranges1.readthedocs.io/en/latest/index.html)\n",
    "\n",
    "```{bash}\n",
    "mamba env export -n utr3.venv > environment.yml\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename\n",
    "\n",
    "def define_join_type(row):\n",
    "    gff_chr = row['Chromosome']\n",
    "    gff_start = row['']\n",
    "    gff_end = row['']\n",
    "    sv_chr = row['Chromosome']\n",
    "    sv_start = row['']\n",
    "    sv_end = row['']\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## Fetch data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.chr.gff3.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e446fd43",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b315f3",
   "metadata": {},
   "source": [
    "### Create PR/DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_pr = pr.read_gff3(str(gff_filepath))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only the necessary columns (62 of 600...)\n",
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ']\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End']\n",
    "sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=list(range(len(target_columns))), names=target_columns, comment='#')\n",
    "sv_df['Chromosome'] = sv_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "sv_pr = pr.PyRanges(sv_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b001f54",
   "metadata": {},
   "source": [
    "### EDA (Exploratory Data Analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788e1cc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_pr.Feature.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e7f9eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_pr.biotype.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cfc34fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_pr.svtype.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "011dad84",
   "metadata": {},
   "source": [
    "### Prepare GFF3 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f6e1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_pr = gff_pr[gff_pr.Feature == 'mRNA']\n",
    "three_utrs_pr = gff_pr[gff_pr.Feature == 'three_prime_UTR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1834b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation of 3'-UTRs with mRNA (parent) data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level'] # Replace the data in these empty 3'-UTRs columns with mRNA data\n",
    "mrna_pr_short = mrna_pr[target_columns]\n",
    "three_utrs_pr_annotated = three_utrs_pr.merge(mrna_pr_short, how='left', left_on='Parent', right_on='ID', suffixes=['_3utr', ''])\n",
    "three_utrs_pr_annotated[['Score', 'Frame']] = three_utrs_pr_annotated[['Score', 'Frame']].replace('.', np.nan)\n",
    "three_utrs_pr_annotated = three_utrs_pr_annotated.dropna(axis=1, how='all') # Remove empty columns\n",
    "\n",
    "# Stats\n",
    "three_utrs_pr_annotated['biotype'].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b10ec4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only MANE_select / MANE_Select|Ensembl_canonical\n",
    "three_utrs_pr_filtered = three_utrs_pr_annotated[three_utrs_pr_annotated['tag'].str.contains('MANE_Select', na=False, regex=True)]\n",
    "\n",
    "# Stats\n",
    "# MANE_select genes count = 19437 (release_1.5) from summary file: https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.5/\n",
    "three_utrs_pr_filtered['biotype'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ee4682",
   "metadata": {},
   "outputs": [],
   "source": [
    "three_utrs_pr_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20ce9d9",
   "metadata": {},
   "source": [
    "### Prepare gnomAD SV data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c1ed5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_del = sv_pr[sv_pr.svtype == \"DEL\"]\n",
    "sv_del.del_len = sv_del.End - sv_del.Start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3f5b08",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_del"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66228d5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_del.del_len.describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e87fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.histogram( \n",
    "    x=sv_del.del_len,\n",
    "    nbins=300,\n",
    "    log_y=True,\n",
    "    title='gnomAD SV DEL lenght distribution (log)'\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e0ca86",
   "metadata": {},
   "source": [
    "### Join GFF3 x gnomAD SV DEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b14ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "three_utr_sv_joined_pr = three_utrs_pr_filtered.join_overlaps(sv_del, suffix='_sv', report_overlap_column='overlap_length')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dee65c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "three_utr_sv_joined_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57e2487f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
