{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import json\n",
    "import pandas as pd, numpy as np, pyranges as pr\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 900,
   "id": "a644ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 500)\n",
    "pd.set_option('display.float_format', '{:.10f}'.format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Docs / notes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189c8ada",
   "metadata": {},
   "source": [
    "**Docs:**  \n",
    "* [PyRanges v1.x GitHub](https://github.com/pyranges/pyranges_1.x) / [PyRanges v1.x Docs](https://pyranges1.readthedocs.io/en/latest/index.html)\n",
    "\n",
    "**Notes:**  \n",
    " - !!! Нужно иметь ввиду, что позиции ФИЧ в GFF, например 3'-UTR, могут быть разделены на несколько интервалов, так как разделены, например, интронами.\n",
    "\n",
    "**Tasks:**  \n",
    "1. Привести все к одной нумерации позиций: GFF - 1-based, BED - 0-based\n",
    "2. Проверить правильность offtarget_id. Сейчас это по факту позиции SV.\n",
    "\n",
    "**Commands:**  \n",
    "```{bash}\n",
    "mamba env export -n utr3.venv > environment.yml\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d6c257",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "other_dir = main_path / 'data/other'\n",
    "output_dir = main_path / 'data/output'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(other_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(output_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 649,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def gunzip_file(input_filepath):\n",
    "    output_filepath = input_filepath.with_suffix('')\n",
    "    command = f'gunzip -c {input_filepath} > {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def sort_gff(input_filepath):\n",
    "    output_filepath = input_filepath.with_stem(input_filepath.stem + \".sorted\")\n",
    "    command = f'igvtools sort {input_filepath} {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def define_join_type(row):\n",
    "    feature_start = row['Start']\n",
    "    feature_end = row['End']\n",
    "    sv_start = row['Start_sv']\n",
    "    sv_end = row['End_sv']\n",
    "\n",
    "    join_type = np.nan\n",
    "    if feature_start <= sv_start and feature_end >= sv_end: # SV полностью в feature\n",
    "        join_type = 'sv_in_feature'\n",
    "    elif sv_start <= feature_start and sv_end >= feature_end: # feature полностью в sv\n",
    "        join_type = 'feature_in_sv'\n",
    "    elif sv_start < feature_start and sv_end <= sv_end:\n",
    "        join_type = 'sv_left_free'\n",
    "    elif sv_start > feature_start and sv_end >= sv_end:\n",
    "        join_type = 'sv_right_free'\n",
    "    elif sv_start == feature_start and sv_end == sv_end:\n",
    "        join_type = 'full_join'\n",
    "\n",
    "    return join_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## Fetch data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.gff3.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.dna.toplevel.fa.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/gnomad/gnomad.v4.1.sv.sites.bed.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/other/so.json' already there; not retrieving.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.gff3.gz', refs_dir)\n",
    "fasta_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)\n",
    "\n",
    "so_terms_filepath = fetch_file('https://raw.githubusercontent.com/The-Sequence-Ontology/SO-Ontologies/refs/heads/master/Ontology_Files/so.json', other_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61697b2",
   "metadata": {},
   "source": [
    "## Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 650,
   "id": "20681675",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sorting /Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.gff3  -> /Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.sorted.gff3\n",
      "Done\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "gff_gunzipped_filepath = gunzip_file(gff_filepath)\n",
    "gff_sorted_filepath = sort_gff(gff_gunzipped_filepath)\n",
    "index_gff(gff_sorted_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e446fd43",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Main</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b315f3",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Create PR/DF<span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF\n",
    "gff_pr = pr.read_gff3(str(gff_filepath))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SV\n",
    "# Keep only the necessary columns (62 of 600...) in SV\n",
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ']\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End']\n",
    "sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=list(range(len(target_columns))), names=target_columns, comment='#')\n",
    "sv_df['Chromosome'] = sv_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "sv_pr = pr.PyRanges(sv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 911,
   "id": "d0a4271f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SO terms\n",
    "with open(so_terms_filepath, 'r') as so_terms_file:\n",
    "    so_terms = json.load(so_terms_file)\n",
    "\n",
    "so_terms_filtered = {\n",
    "    'lbl': [],\n",
    "    'definition': []\n",
    "}\n",
    "for i in so_terms['graphs'][0]['nodes']:\n",
    "    lbl = i.get('lbl', np.nan)\n",
    "    definition = i.get('meta', {}).get('definition', {}).get('val', np.nan)\n",
    "    so_terms_filtered['lbl'].append(lbl)\n",
    "    so_terms_filtered['definition'].append(definition)\n",
    "\n",
    "so_terms_df = pd.DataFrame.from_dict(so_terms_filtered)\n",
    "so_terms_df['lbl'] = so_terms_df['lbl'].str.capitalize().replace('_', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b001f54",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">EDA (Exploratory Data Analysis)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788e1cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>count</th>\n",
       "      <th>definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon</td>\n",
       "      <td>3673949</td>\n",
       "      <td>A region of the transcript sequence within a gene which is not removed from the primary RNA transcript by RNA splicing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>2284258</td>\n",
       "      <td>A contiguous sequence which begins with, and includes, a start codon and ends with, and includes, a stop codon.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>426597</td>\n",
       "      <td>A region at the 5' end of a mature transcript (preceding the initiation codon) that is not translated into a protein.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>340566</td>\n",
       "      <td>A region at the 3' end of a mature transcript (following the stop codon) that is not translated into a protein.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>233574</td>\n",
       "      <td>Messenger RNA is the intermediate molecule between DNA and protein. It includes UTR and coding sequences. It does not contain introns.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>223403</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>biological_region</td>\n",
       "      <td>180392</td>\n",
       "      <td>A region defined by its disposition to be involved in a biological process.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>41946</td>\n",
       "      <td>A gene that encodes a non-coding RNA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>transcript</td>\n",
       "      <td>28799</td>\n",
       "      <td>An RNA synthesized on a DNA or RNA template by an RNA polymerase.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene</td>\n",
       "      <td>21547</td>\n",
       "      <td>A region (or regions) that includes all of the sequence elements necessary to encode a functional transcript. A gene may include regulatory regions, transcribed regions and/or other functional sequence regions.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>15201</td>\n",
       "      <td>A non functional descendant of a transcript, part of a pseudogene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>15198</td>\n",
       "      <td>A sequence that closely resembles a known functional gene, at another locus within a genome, that is non-functional as a consequence of (usually several) mutations that prevent either its transcription or translation (or both). In general, pseudogenes result from either reverse transcription of a transcript of their \"normal\" paralog (SO:0000043) (in which case the pseudogene typically lacks introns and includes a poly(A) tail) or from recombination (SO:0000044) (in which case the pseudogene ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>1906</td>\n",
       "      <td>A small nuclear RNA molecule involved in pre-mRNA splicing and processing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>1879</td>\n",
       "      <td>Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>unconfirmed_transcript</td>\n",
       "      <td>1108</td>\n",
       "      <td>This is used for non-spliced EST clusters that have polyA features. This category has been specifically created for the ENCODE project to highlight regions that could indicate the presence of protein coding genes that require experimental validation, either by 5' RACE or RT-PCR to extend the transcripts, or by confirming expression of the putatively-encoded peptide with specific antibodies.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>942</td>\n",
       "      <td>Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>V_gene_segment</td>\n",
       "      <td>253</td>\n",
       "      <td>Germline genomic DNA including L-part1, V-intron and V-exon, with the 5' UTR and 3' UTR.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>J_gene_segment</td>\n",
       "      <td>97</td>\n",
       "      <td>Germline genomic DNA of an immunoglobulin/T-cell receptor gene including J-region with 5' UTR (SO:0000204) and 3' UTR (SO:0000205), also designated as J-segment.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>scRNA</td>\n",
       "      <td>49</td>\n",
       "      <td>A small non coding RNA sequence, present in the cytoplasm.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>49</td>\n",
       "      <td>rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>D_gene_segment</td>\n",
       "      <td>42</td>\n",
       "      <td>Germline genomic DNA including D-region with 5' UTR and 3' UTR, also designated as D-segment.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>C_gene_segment</td>\n",
       "      <td>29</td>\n",
       "      <td>Genomic DNA of immunoglobulin/T-cell receptor gene including C-region (and introns if present) with 5' UTR (SO:0000204) and 3' UTR (SO:0000205).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chromosome</td>\n",
       "      <td>25</td>\n",
       "      <td>Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tRNA</td>\n",
       "      <td>22</td>\n",
       "      <td>Transfer RNA (tRNA) molecules are approximately 80 nucleotides in length. Their secondary structure includes four short double-helical elements and three loops (D, anti-codon, and T loops). Further hydrogen bonds mediate the characteristic L-shaped molecular structure. Transfer RNAs have two regions of fundamental functional importance: the anti-codon, which is responsible for specific mRNA codon recognition, and the 3' end, to which the tRNA's corresponding amino acid is attached (by aminoa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>12</td>\n",
       "      <td>A transcript for which no open reading frame has been identified and for which no other function has been determined.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Feature    count                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           definition\n",
       "0                     exon  3673949                                                                                                                                                                                                                                                                                                                                                                                              A region of the transcript sequence within a gene which is not removed from the primary RNA transcript by RNA splicing.\n",
       "1                      CDS  2284258                                                                                                                                                                                                                                                                                                                                                                                                      A contiguous sequence which begins with, and includes, a start codon and ends with, and includes, a stop codon.\n",
       "2           five_prime_UTR   426597                                                                                                                                                                                                                                                                                                                                                                                                A region at the 5' end of a mature transcript (preceding the initiation codon) that is not translated into a protein.\n",
       "3          three_prime_UTR   340566                                                                                                                                                                                                                                                                                                                                                                                                      A region at the 3' end of a mature transcript (following the stop codon) that is not translated into a protein.\n",
       "4                     mRNA   233574                                                                                                                                                                                                                                                                                                                                                                               Messenger RNA is the intermediate molecule between DNA and protein. It includes UTR and coding sequences. It does not contain introns.\n",
       "5                  lnc_RNA   223403                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN\n",
       "6        biological_region   180392                                                                                                                                                                                                                                                                                                                                                                                                                                          A region defined by its disposition to be involved in a biological process.\n",
       "7               ncRNA_gene    41946                                                                                                                                                                                                                                                                                                                                                                                                                                                                                A gene that encodes a non-coding RNA.\n",
       "8               transcript    28799                                                                                                                                                                                                                                                                                                                                                                                                                                                    An RNA synthesized on a DNA or RNA template by an RNA polymerase.\n",
       "9                     gene    21547                                                                                                                                                                                                                                                                                                   A region (or regions) that includes all of the sequence elements necessary to encode a functional transcript. A gene may include regulatory regions, transcribed regions and/or other functional sequence regions.\n",
       "10  pseudogenic_transcript    15201                                                                                                                                                                                                                                                                                                                                                                                                                                                   A non functional descendant of a transcript, part of a pseudogene.\n",
       "11              pseudogene    15198  A sequence that closely resembles a known functional gene, at another locus within a genome, that is non-functional as a consequence of (usually several) mutations that prevent either its transcription or translation (or both). In general, pseudogenes result from either reverse transcription of a transcript of their \"normal\" paralog (SO:0000043) (in which case the pseudogene typically lacks introns and includes a poly(A) tail) or from recombination (SO:0000044) (in which case the pseudogene ...\n",
       "12                   snRNA     1906                                                                                                                                                                                                                                                                                                                                                                                                                                           A small nuclear RNA molecule involved in pre-mRNA splicing and processing.\n",
       "13                   miRNA     1879                     Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.\n",
       "14  unconfirmed_transcript     1108                                                                                                            This is used for non-spliced EST clusters that have polyA features. This category has been specifically created for the ENCODE project to highlight regions that could indicate the presence of protein coding genes that require experimental validation, either by 5' RACE or RT-PCR to extend the transcripts, or by confirming expression of the putatively-encoded peptide with specific antibodies.\n",
       "15                  snoRNA      942                                                                                                      Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.\n",
       "16          V_gene_segment      253                                                                                                                                                                                                                                                                                                                                                                                                                             Germline genomic DNA including L-part1, V-intron and V-exon, with the 5' UTR and 3' UTR.\n",
       "17          J_gene_segment       97                                                                                                                                                                                                                                                                                                                                                    Germline genomic DNA of an immunoglobulin/T-cell receptor gene including J-region with 5' UTR (SO:0000204) and 3' UTR (SO:0000205), also designated as J-segment.\n",
       "18                   scRNA       49                                                                                                                                                                                                                                                                                                                                                                                                                                                           A small non coding RNA sequence, present in the cytoplasm.\n",
       "19                    rRNA       49                                                                                                                                                                                                                                                                                                                                                                                                          rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.\n",
       "20          D_gene_segment       42                                                                                                                                                                                                                                                                                                                                                                                                                        Germline genomic DNA including D-region with 5' UTR and 3' UTR, also designated as D-segment.\n",
       "21          C_gene_segment       29                                                                                                                                                                                                                                                                                                                                                                     Genomic DNA of immunoglobulin/T-cell receptor gene including C-region (and introns if present) with 5' UTR (SO:0000204) and 3' UTR (SO:0000205).\n",
       "22              chromosome       25                                                                                                                                                                                                                                                                                                                                           Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication.\n",
       "23                    tRNA       22  Transfer RNA (tRNA) molecules are approximately 80 nucleotides in length. Their secondary structure includes four short double-helical elements and three loops (D, anti-codon, and T loops). Further hydrogen bonds mediate the characteristic L-shaped molecular structure. Transfer RNAs have two regions of fundamental functional importance: the anti-codon, which is responsible for specific mRNA codon recognition, and the 3' end, to which the tRNA's corresponding amino acid is attached (by aminoa...\n",
       "24    processed_transcript       12                                                                                                                                                                                                                                                                                                                                                                                                A transcript for which no open reading frame has been identified and for which no other function has been determined."
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GFF features stats\n",
    "gff_features = gff_pr.Feature.value_counts().to_frame().reset_index()\n",
    "gff_features['tmp_id'] = gff_features['Feature'].str.capitalize().replace('_', '')\n",
    "gff_features = gff_features.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_features = gff_features.drop(['lbl', 'tmp_id'], axis=1)\n",
    "pd.DataFrame(gff_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e7f9eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "      <th>definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>231543</td>\n",
       "      <td>A gene which, when transcribed, can be translated into a protein.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>224032</td>\n",
       "      <td>A non-coding RNA generally longer than 200 nucleotides that cannot be classified as any other ncRNA subtype. Similar to mRNAs, lncRNAs are mainly transcribed by RNA polymerase II, are often capped by 7-methyl guanosine at their 5' ends, polyadenylated at their 3' ends and may be spliced.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>retained_intron</td>\n",
       "      <td>34239</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>26573</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>21949</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>18975</td>\n",
       "      <td>A pseudogene created via retrotranposition of the mRNA of a functional protein-coding parent gene followed by accumulation of deleterious mutations lacking introns and promoters, often including a polyA tail.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>4414</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>3898</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>3802</td>\n",
       "      <td>A small nuclear RNA molecule involved in pre-mRNA splicing and processing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>3758</td>\n",
       "      <td>Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>3176</td>\n",
       "      <td>A unprocessed pseudogene supported by locus-specific evidence of transcription.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2298</td>\n",
       "      <td>A processed_pseudogene overlapped by locus-specific evidence of transcription.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TEC</td>\n",
       "      <td>2127</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>1884</td>\n",
       "      <td>Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>994</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>transcribed_unitary_pseudogene</td>\n",
       "      <td>402</td>\n",
       "      <td>A species specific unprocessed pseudogene without a parent gene, as it has an active orthologue in another species.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IG_V_pseudogene</td>\n",
       "      <td>374</td>\n",
       "      <td>A pseudogenic variable region which closely resembles a known functional imunoglobulin variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>292</td>\n",
       "      <td>A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of an Immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TR_V_gene</td>\n",
       "      <td>214</td>\n",
       "      <td>A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of aT-cell receptor chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>unitary_pseudogene</td>\n",
       "      <td>178</td>\n",
       "      <td>A pseudogene, deactivated from original state by mutation, fixed in a population,where the ortholog in a reference species such as mouse remains functional.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TR_J_gene</td>\n",
       "      <td>158</td>\n",
       "      <td>A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of aT-cell receptor chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>non_stop_decay</td>\n",
       "      <td>105</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>scaRNA</td>\n",
       "      <td>98</td>\n",
       "      <td>A ncRNA, specific to the Cajal body, that has been demonstrated to function as a guide RNA in the site-specific synthesis of 2'-O-ribose-methylated nucleotides and pseudouridines in the RNA polymerase II-transcribed U1, U2, U4 and U5 spliceosomal small nuclear RNAs (snRNAs).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>94</td>\n",
       "      <td>rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>IG_D_gene</td>\n",
       "      <td>74</td>\n",
       "      <td>A gene that rearranges at the DNA level and codes the diversity region of the variable domain of an immunoglobuin (IG) gene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>protein_coding_LoF</td>\n",
       "      <td>74</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TR_V_pseudogene</td>\n",
       "      <td>66</td>\n",
       "      <td>A pseudogenic variable region which closely resembles a known functional T receptor variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>44</td>\n",
       "      <td>Mitochondrial transfer RNA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>artifact</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>IG_C_gene</td>\n",
       "      <td>37</td>\n",
       "      <td>A constant (C) gene, a gene that codes the constant region of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>IG_J_gene</td>\n",
       "      <td>36</td>\n",
       "      <td>A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "      <td>18</td>\n",
       "      <td>A pseudogenic constant region of an immunoglobulin gene which closely resembles a known functional Imunoglobulin constant gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ribozyme</td>\n",
       "      <td>16</td>\n",
       "      <td>An RNA with catalytic activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>TR_C_gene</td>\n",
       "      <td>12</td>\n",
       "      <td>A constant (C) gene, a gene that codes the constant region of a T-cell receptor chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>12</td>\n",
       "      <td>A transcript for which no open reading frame has been identified and for which no other function has been determined.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>10</td>\n",
       "      <td>A gene that rearranges at the DNA level and codes the diversity region of the variable domain of aT-cell receptor gene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>sRNA</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>vault_RNA</td>\n",
       "      <td>8</td>\n",
       "      <td>A family of RNAs are found as part of the enigmatic vault ribonucleoprotein complex. The complex consists of a major vault protein (MVP), two minor vault proteins (VPARP and TEP1), and several small untranslated RNA molecules. It has been suggested that the vault complex is involved in drug resistance.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>TR_J_pseudogene</td>\n",
       "      <td>8</td>\n",
       "      <td>A pseudogenic joining region which closely resembles a known functional T receptor (TR) joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>IG_J_pseudogene</td>\n",
       "      <td>6</td>\n",
       "      <td>A pseudogenic joining region which closely resembles a known functional imunoglobulin joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>translated_processed_pseudogene</td>\n",
       "      <td>4</td>\n",
       "      <td>A processed pseudogene where there is evidence, (mass spec data) suggesting that it is also translated.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>Mitochondrial rRNA is an RNA component of the small or large subunits of mitochondrial ribosomes.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>IG_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               biotype   count                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        definition\n",
       "0                       protein_coding  231543                                                                                                                                                                                                                                                                                                                                                                                                                                 A gene which, when transcribed, can be translated into a protein.\n",
       "1                               lncRNA  224032                                                                                                                                                                                                  A non-coding RNA generally longer than 200 nucleotides that cannot be classified as any other ncRNA subtype. Similar to mRNAs, lncRNAs are mainly transcribed by RNA polymerase II, are often capped by 7-methyl guanosine at their 5' ends, polyadenylated at their 3' ends and may be spliced.\n",
       "2                      retained_intron   34239                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "3       protein_coding_CDS_not_defined   26573                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "4              nonsense_mediated_decay   21949                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "5                 processed_pseudogene   18975                                                                                                                                                                                                                                                                                  A pseudogene created via retrotranposition of the mRNA of a functional protein-coding parent gene followed by accumulation of deleterious mutations lacking introns and promoters, often including a polyA tail.\n",
       "6                             misc_RNA    4414                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "7               unprocessed_pseudogene    3898                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "8                                snRNA    3802                                                                                                                                                                                                                                                                                                                                                                                                                        A small nuclear RNA molecule involved in pre-mRNA splicing and processing.\n",
       "9                                miRNA    3758  Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.\n",
       "10  transcribed_unprocessed_pseudogene    3176                                                                                                                                                                                                                                                                                                                                                                                                                   A unprocessed pseudogene supported by locus-specific evidence of transcription.\n",
       "11    transcribed_processed_pseudogene    2298                                                                                                                                                                                                                                                                                                                                                                                                                    A processed_pseudogene overlapped by locus-specific evidence of transcription.\n",
       "12                                 TEC    2127                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "13                              snoRNA    1884                                                                                   Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.\n",
       "14                     rRNA_pseudogene     994                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "15      transcribed_unitary_pseudogene     402                                                                                                                                                                                                                                                                                                                                                                               A species specific unprocessed pseudogene without a parent gene, as it has an active orthologue in another species.\n",
       "16                     IG_V_pseudogene     374                                                                                                                                                                              A pseudogenic variable region which closely resembles a known functional imunoglobulin variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.\n",
       "17                           IG_V_gene     292                                                                                                                                                                                                                                                                                                                                                                 A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of an Immunoglobulin chain.\n",
       "18                           TR_V_gene     214                                                                                                                                                                                                                                                                                                                                                                  A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of aT-cell receptor chain.\n",
       "19                  unitary_pseudogene     178                                                                                                                                                                                                                                                                                                                                      A pseudogene, deactivated from original state by mutation, fixed in a population,where the ortholog in a reference species such as mouse remains functional.\n",
       "20                           TR_J_gene     158                                                                                                                                                                                                                                                                                                                                                                    A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of aT-cell receptor chain.\n",
       "21                      non_stop_decay     105                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "22                              scaRNA      98                                                                                                                                                                                                               A ncRNA, specific to the Cajal body, that has been demonstrated to function as a guide RNA in the site-specific synthesis of 2'-O-ribose-methylated nucleotides and pseudouridines in the RNA polymerase II-transcribed U1, U2, U4 and U5 spliceosomal small nuclear RNAs (snRNAs).\n",
       "23                                rRNA      94                                                                                                                                                                                                                                                                                                                                                                                       rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.\n",
       "24                           IG_D_gene      74                                                                                                                                                                                                                                                                                                                                                                      A gene that rearranges at the DNA level and codes the diversity region of the variable domain of an immunoglobuin (IG) gene.\n",
       "25                  protein_coding_LoF      74                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "26                     TR_V_pseudogene      66                                                                                                                                                                                 A pseudogenic variable region which closely resembles a known functional T receptor variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.\n",
       "27                             Mt_tRNA      44                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Mitochondrial transfer RNA.\n",
       "28                            artifact      38                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "29                           IG_C_gene      37                                                                                                                                                                                                                                                                                                                                                                                                            A constant (C) gene, a gene that codes the constant region of an immunoglobulin chain.\n",
       "30                           IG_J_gene      36                                                                                                                                                                                                                                                                                                                                                                   A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.\n",
       "31                     IG_C_pseudogene      18                                                                                                                                                                                                                                              A pseudogenic constant region of an immunoglobulin gene which closely resembles a known functional Imunoglobulin constant gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon.\n",
       "32                            ribozyme      16                                                                                                                                                                                                                                                                                                                                                                                                                                                                   An RNA with catalytic activity.\n",
       "33                           TR_C_gene      12                                                                                                                                                                                                                                                                                                                                                                                                            A constant (C) gene, a gene that codes the constant region of a T-cell receptor chain.\n",
       "34                processed_transcript      12                                                                                                                                                                                                                                                                                                                                                                             A transcript for which no open reading frame has been identified and for which no other function has been determined.\n",
       "35                           TR_D_gene      10                                                                                                                                                                                                                                                                                                                                                                           A gene that rearranges at the DNA level and codes the diversity region of the variable domain of aT-cell receptor gene.\n",
       "36                                sRNA      10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN\n",
       "37                           vault_RNA       8                                                                                                                                                                                   A family of RNAs are found as part of the enigmatic vault ribonucleoprotein complex. The complex consists of a major vault protein (MVP), two minor vault proteins (VPARP and TEP1), and several small untranslated RNA molecules. It has been suggested that the vault complex is involved in drug resistance.\n",
       "38                     TR_J_pseudogene       8                                                                                                                                                        A pseudogenic joining region which closely resembles a known functional T receptor (TR) joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.\n",
       "39                     IG_J_pseudogene       6                                                                                                                                                          A pseudogenic joining region which closely resembles a known functional imunoglobulin joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.\n",
       "40     translated_processed_pseudogene       4                                                                                                                                                                                                                                                                                                                                                                                           A processed pseudogene where there is evidence, (mass spec data) suggesting that it is also translated.\n",
       "41                             Mt_rRNA       4                                                                                                                                                                                                                                                                                                                                                                                                 Mitochondrial rRNA is an RNA component of the small or large subunits of mitochondrial ribosomes.\n",
       "42                       IG_pseudogene       2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               NaN"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GFF biotypes stats\n",
    "gff_biotypes = gff_pr.biotype.value_counts().to_frame().reset_index()\n",
    "gff_biotypes['tmp_id'] = gff_biotypes['biotype'].str.capitalize().replace('_', '')\n",
    "gff_biotypes = gff_biotypes.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_biotypes = gff_biotypes.drop(['lbl', 'tmp_id'], axis=1)\n",
    "pd.DataFrame(gff_biotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "id": "8cfc34fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>svtype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DEL</td>\n",
       "      <td>1197080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BND</td>\n",
       "      <td>356035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DUP</td>\n",
       "      <td>269326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INS:ME:ALU</td>\n",
       "      <td>173374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>INS</td>\n",
       "      <td>83441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>INS:ME:LINE1</td>\n",
       "      <td>30223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>INS:ME:SVA</td>\n",
       "      <td>17607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CPX</td>\n",
       "      <td>15189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DEL:ME:LINE1</td>\n",
       "      <td>8505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>INV</td>\n",
       "      <td>2193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CNV</td>\n",
       "      <td>721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DEL:ME:HERVK</td>\n",
       "      <td>693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CTX</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          svtype    count\n",
       "0            DEL  1197080\n",
       "1            BND   356035\n",
       "2            DUP   269326\n",
       "3     INS:ME:ALU   173374\n",
       "4            INS    83441\n",
       "5   INS:ME:LINE1    30223\n",
       "6     INS:ME:SVA    17607\n",
       "7            CPX    15189\n",
       "8   DEL:ME:LINE1     8505\n",
       "9            INV     2193\n",
       "10           CNV      721\n",
       "11  DEL:ME:HERVK      693\n",
       "12           CTX       99"
      ]
     },
     "execution_count": 547,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_pr.svtype.value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "011dad84",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Prepare GFF3 data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 856,
   "id": "bb4ed9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Self annotation of GFF with mRNA (parent) data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level', 'feature_len', 'feature_id'] # Replace the data in these empty columns with mRNA data\n",
    "annotation_gff_pr = gff_pr[target_columns].dropna(subset='ID')\n",
    "\n",
    "annotation_suffix = '_annotation'\n",
    "annotated_gff_pr = gff_pr.merge(annotation_gff_pr, how='left', left_on='Parent', right_on='ID', suffixes=['', annotation_suffix])\n",
    "\n",
    "annotated_columns = [f'{column}{annotation_suffix}' for column in target_columns]\n",
    "fill_dict = dict(zip(target_columns, annotated_columns))\n",
    "for target, annotation in fill_dict.items():\n",
    "    annotated_gff_pr[target] = annotated_gff_pr[target].fillna(annotated_gff_pr[annotation])\n",
    "annotated_gff_pr = annotated_gff_pr.drop(annotated_columns, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8f64c00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add additional columns\n",
    "annotated_gff_pr['feature_len'] = (annotated_gff_pr['End'] - annotated_gff_pr['Start']).astype('int')\n",
    "annotated_gff_pr['feature_id'] = 'feature_' + annotated_gff_pr['Chromosome'].astype('str') + ':' + annotated_gff_pr['Start'].astype('str') + '-' + annotated_gff_pr['End'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f6e1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create specific PRs (mRNAm, 3'-UTR, offtargets)\n",
    "mrna_pr = annotated_gff_pr.query('Feature == \"mRNA\"')\n",
    "three_utrs_pr = annotated_gff_pr.query('Feature == \"three_prime_UTR\"')\n",
    "\n",
    "offtarget_features = ['five_prime_UTR',\n",
    "                               'CDS',\n",
    "                               #'lnc_RNA',\n",
    "                               'ncRNA_gene',\n",
    "                               #'snRNA',\n",
    "                               #'snoRNA',\n",
    "                               #'scRNA',\n",
    "                               'rRNA',\n",
    "                               'tRNA',\n",
    "                               'processed_transcript',\n",
    "                               ]\n",
    "offtarget_features_pr = gff_pr.query('Feature in @offtarget_features')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 912,
   "id": "d1834b24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>190254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>21947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding_LoF</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IG_C_gene</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_C_gene</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   biotype   count\n",
       "0           protein_coding  190254\n",
       "1  nonsense_mediated_decay   21947\n",
       "2       protein_coding_LoF      43\n",
       "3                IG_C_gene      22\n",
       "4                TR_C_gene       6\n",
       "5                IG_V_gene       2"
      ]
     },
     "execution_count": 912,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3'-UTR biotype stats\n",
    "# Считаем именно уникальные вхождения, так как один 3'-UTR может быть разбить на несколько интервалов в GFF (разделен интронами).\n",
    "# В особенности это выражено у nonsense_mediated_decay транскриптов (нам они не пригодятся, но тем не менее).\n",
    "three_utrs_pr.drop_duplicates(subset='ID')['biotype'].value_counts().to_frame().reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 913,
   "id": "3b10ec4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          biotype  count\n",
       "0  protein_coding  19757"
      ]
     },
     "execution_count": 913,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only MANE_select / MANE_Select|Ensembl_canonical\n",
    "three_utrs_pr_filtered = three_utrs_pr[three_utrs_pr['tag'].str.contains('MANE_Select', na=False, regex=True)]\n",
    "\n",
    "# Stats\n",
    "# MANE_select genes count = 19437 (release_1.5) from summary file: https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.5/\n",
    "three_utrs_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20ce9d9",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Prepare SV data</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2a71e19",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">gnomAD</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c1ed5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep target SVs (DEL)\n",
    "sv_targets = ['DEL']\n",
    "target_sv_pr = sv_pr.query('svtype in @sv_targets')\n",
    "\n",
    "# Add columns\n",
    "target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
    "target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 914,
   "id": "970eda5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>1197080.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean</td>\n",
       "      <td>0.0388642708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>std</td>\n",
       "      <td>0.1348195999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>min</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10%</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25%</td>\n",
       "      <td>0.0000160000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>50%</td>\n",
       "      <td>0.0000500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>75%</td>\n",
       "      <td>0.0003650000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>90%</td>\n",
       "      <td>0.0076041001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>95%</td>\n",
       "      <td>0.5000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>99%</td>\n",
       "      <td>0.5000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>99.9%</td>\n",
       "      <td>0.9285709858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>max</td>\n",
       "      <td>1.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index                 AF\n",
       "0   count 1197080.0000000000\n",
       "1    mean       0.0388642708\n",
       "2     std       0.1348195999\n",
       "3     min       0.0000080000\n",
       "4     10%       0.0000080000\n",
       "5     25%       0.0000160000\n",
       "6     50%       0.0000500000\n",
       "7     75%       0.0003650000\n",
       "8     90%       0.0076041001\n",
       "9     95%       0.5000000000\n",
       "10    99%       0.5000000000\n",
       "11  99.9%       0.9285709858\n",
       "12    max       1.0000000000"
      ]
     },
     "execution_count": 914,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SV percentiles by AF\n",
    "target_sv_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66228d5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sv_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>1197080.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean</td>\n",
       "      <td>5086.8619674541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>std</td>\n",
       "      <td>154413.4211853280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>min</td>\n",
       "      <td>51.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10%</td>\n",
       "      <td>99.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25%</td>\n",
       "      <td>324.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>50%</td>\n",
       "      <td>609.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>75%</td>\n",
       "      <td>964.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>90%</td>\n",
       "      <td>5847.1000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>95%</td>\n",
       "      <td>10927.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>99%</td>\n",
       "      <td>82869.0499999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>99.9%</td>\n",
       "      <td>412913.1350000836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>max</td>\n",
       "      <td>87762440.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index              sv_len\n",
       "0   count  1197080.0000000000\n",
       "1    mean     5086.8619674541\n",
       "2     std   154413.4211853280\n",
       "3     min       51.0000000000\n",
       "4     10%       99.0000000000\n",
       "5     25%      324.0000000000\n",
       "6     50%      609.0000000000\n",
       "7     75%      964.0000000000\n",
       "8     90%     5847.1000000001\n",
       "9     95%    10927.0000000000\n",
       "10    99%    82869.0499999998\n",
       "11  99.9%   412913.1350000836\n",
       "12    max 87762440.0000000000"
      ]
     },
     "execution_count": 915,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SV percentiles by SV length\n",
    "target_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 908,
   "id": "79e87fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"fig = px.histogram( \\n    x=target_sv_pr['sv_len'],\\n    nbins=300,\\n    log_y=True,\\n    title='gnomAD SV DEL lenght distribution (log)'\\n)\\nfig.show()\""
      ]
     },
     "execution_count": 908,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''fig = px.histogram( \n",
    "    x=target_sv_pr['sv_len'],\n",
    "    nbins=300,\n",
    "    log_y=True,\n",
    "    title='gnomAD SV DEL lenght distribution (log)'\n",
    ")\n",
    "fig.show()'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e0ca86",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Join GFF3 x gnomAD SV DEL</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 867,
   "id": "45b14ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join GFF x SV\n",
    "features_sv_joined_pr = three_utrs_pr_filtered.join_overlaps(target_sv_pr, suffix='_sv', report_overlap_column='join_len')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6885c5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add additional columns (join_start, join_end, join_id)\n",
    "features_sv_joined_pr['join_start'] = np.maximum(features_sv_joined_pr['Start'], features_sv_joined_pr['Start_sv'])\n",
    "features_sv_joined_pr['join_end'] = np.minimum(features_sv_joined_pr['End'], features_sv_joined_pr['End_sv'])\n",
    "features_sv_joined_pr['join_id'] = 'join_' + features_sv_joined_pr['Chromosome'].astype('str') + ':' + features_sv_joined_pr['join_start'].astype('str') + '-' + features_sv_joined_pr['join_end'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 916,
   "id": "33ec54b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>join_type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>feature_in_sv</td>\n",
       "      <td>20428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>8318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>3701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sv_left_free</td>\n",
       "      <td>3455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       join_type  count\n",
       "0  feature_in_sv  20428\n",
       "1  sv_in_feature   8318\n",
       "2  sv_right_free   3701\n",
       "3   sv_left_free   3455"
      ]
     },
     "execution_count": 916,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define joint type\n",
    "features_sv_joined_pr['join_type'] = features_sv_joined_pr.apply(define_join_type, axis=1)\n",
    "features_sv_joined_pr['join_type'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ae52f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count joints by SV\n",
    "joints_count_by_sv = features_sv_joined_pr['sv_id'].value_counts().to_frame().reset_index()\n",
    "joints_count_by_sv = joints_count_by_sv.rename(columns={'count': 'joins_count'})\n",
    "\n",
    "# Add joints count data\n",
    "features_sv_joined_pr = features_sv_joined_pr.merge(joints_count_by_sv, how='left')\n",
    "#len(features_sv_joined_pr)\n",
    "\n",
    "# Filter joins by join type\n",
    "targey_joint_types = ['sv_in_feature', 'sv_right_free', 'sv_left_free']\n",
    "features_sv_joined_pr_filtered = features_sv_joined_pr.query('join_type in @targey_joint_types and ((Strand == \"+\" and join_type == \"sv_right_free\") or (Strand == \"-\" and join_type == \"sv_left_free\") or join_type == \"sv_in_feature\") and joins_count == 1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b449576",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join type stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a69d46c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>join_type</th>\n",
       "      <th>sv_in_feature</th>\n",
       "      <th>sv_left_free</th>\n",
       "      <th>sv_right_free</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>8124.0000000000</td>\n",
       "      <td>1589.0000000000</td>\n",
       "      <td>1924.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>538.8242245199</td>\n",
       "      <td>4629.9559471366</td>\n",
       "      <td>4550.9636174636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>632.0622349426</td>\n",
       "      <td>14678.7805179974</td>\n",
       "      <td>15540.4454833119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>51.0000000000</td>\n",
       "      <td>52.0000000000</td>\n",
       "      <td>53.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10%</th>\n",
       "      <td>77.0000000000</td>\n",
       "      <td>454.6000000000</td>\n",
       "      <td>487.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>175.0000000000</td>\n",
       "      <td>627.0000000000</td>\n",
       "      <td>628.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>515.0000000000</td>\n",
       "      <td>1325.0000000000</td>\n",
       "      <td>1102.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>657.0000000000</td>\n",
       "      <td>5075.0000000000</td>\n",
       "      <td>4697.5000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90%</th>\n",
       "      <td>796.0000000000</td>\n",
       "      <td>9015.6000000000</td>\n",
       "      <td>9576.1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95%</th>\n",
       "      <td>1135.8500000000</td>\n",
       "      <td>13867.2000000000</td>\n",
       "      <td>13220.1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99%</th>\n",
       "      <td>3200.0100000000</td>\n",
       "      <td>43648.5199999999</td>\n",
       "      <td>47440.5099999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99.9%</th>\n",
       "      <td>7484.1490000002</td>\n",
       "      <td>225928.0960000205</td>\n",
       "      <td>179193.6130000250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>14078.0000000000</td>\n",
       "      <td>371108.0000000000</td>\n",
       "      <td>455766.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "join_type    sv_in_feature      sv_left_free     sv_right_free\n",
       "count      8124.0000000000   1589.0000000000   1924.0000000000\n",
       "mean        538.8242245199   4629.9559471366   4550.9636174636\n",
       "std         632.0622349426  14678.7805179974  15540.4454833119\n",
       "min          51.0000000000     52.0000000000     53.0000000000\n",
       "10%          77.0000000000    454.6000000000    487.0000000000\n",
       "25%         175.0000000000    627.0000000000    628.0000000000\n",
       "50%         515.0000000000   1325.0000000000   1102.0000000000\n",
       "75%         657.0000000000   5075.0000000000   4697.5000000000\n",
       "90%         796.0000000000   9015.6000000000   9576.1000000000\n",
       "95%        1135.8500000000  13867.2000000000  13220.1000000000\n",
       "99%        3200.0100000000  43648.5199999999  47440.5099999999\n",
       "99.9%      7484.1490000002 225928.0960000205 179193.6130000250\n",
       "max       14078.0000000000 371108.0000000000 455766.0000000000"
      ]
     },
     "execution_count": 873,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stats by join type\n",
    "features_sv_joined_pr_filtered.groupby('join_type')['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 917,
   "id": "cc037975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>11637</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          biotype  count\n",
       "0  protein_coding  11637"
      ]
     },
     "execution_count": 917,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_joined_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4135ee",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Offtarget features exclusion</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1142,
   "id": "ab7f6e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "features_sv_joined_pr_filtered_cp = features_sv_joined_pr_filtered.copy()\n",
    "features_sv_joined_pr_filtered_cp[['Start_tmp', 'End_tmp']] = features_sv_joined_pr_filtered_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_joined_pr_filtered_cp[['Start', 'End']] = features_sv_joined_pr_filtered_cp[['Start_sv', 'End_sv']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1143,
   "id": "2bd478da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# features_sv_offtereget_joined_pr - только для отслеживания пересечений с offtarget. Фильтрованный от offtarget pr - features_sv_no_offtereget_joined_pr (через overlap, потому что join находя пересечения с offtarget (через left) также оставляет оригинальную строку).DS_Storefeatures_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len')\n",
    "features_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len', strand_behavior='ignore')\n",
    "\n",
    "features_sv_offtereget_joined_pr[['Start', 'End']] = features_sv_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_offtereget_joined_pr = features_sv_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_offtereget_joined_pr['offtarget_id'] = 'offtarget_' + features_sv_offtereget_joined_pr['Chromosome'].astype('str') + ':' + features_sv_offtereget_joined_pr['Start_offtarget'].astype('str') + '-' + features_sv_offtereget_joined_pr['End_offtarget'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c516d79f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>join_type</th>\n",
       "      <th>Feature_offtarget</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">sv_in_feature</th>\n",
       "      <th>CDS</th>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>1512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>processed_transcript</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">sv_left_free</th>\n",
       "      <th>CDS</th>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoRNA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">sv_right_free</th>\n",
       "      <th>CDS</th>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>498</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    count\n",
       "join_type     Feature_offtarget          \n",
       "sv_in_feature CDS                     166\n",
       "              five_prime_UTR            6\n",
       "              ncRNA_gene             1512\n",
       "              processed_transcript      1\n",
       "sv_left_free  CDS                     112\n",
       "              five_prime_UTR           31\n",
       "              ncRNA_gene              403\n",
       "              snoRNA                    1\n",
       "sv_right_free CDS                      91\n",
       "              five_prime_UTR           57\n",
       "              ncRNA_gene              498"
      ]
     },
     "execution_count": 1153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features offtargets stats\n",
    "# Суммарные значения данной статистики будут выше чем разница всех джоинов - оффтартет джоинов, так как один и тот же SV (напр. DEL) могут пересекать разные изоформы одного и того же транскрипта.\n",
    "features_sv_offtereget_joined_pr.groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])\n",
    "#features_sv_offtereget_joined_pr.drop_duplicates(subset='sv_id').groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1147,
   "id": "1dc11ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclude offtargets (SV x offtarget overlaps)\n",
    "features_sv_no_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.overlap(offtarget_features_pr, invert=True, strand_behavior='ignore')\n",
    "\n",
    "features_sv_no_offtereget_joined_pr[['Start', 'End']] = features_sv_no_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_joined_pr = features_sv_no_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "623083fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before offtarget filtering\n",
      "       join_type  count\n",
      "0  sv_in_feature   8124\n",
      "1  sv_right_free   1924\n",
      "2   sv_left_free   1589\n",
      "\n",
      "After offtarget filtering\n",
      "       join_type  count\n",
      "0  sv_in_feature   6439\n",
      "1  sv_right_free   1278\n",
      "2   sv_left_free   1042\n"
     ]
    }
   ],
   "source": [
    "# Joint type stats after offtargets exclution\n",
    "print('Before offtargets filtering')\n",
    "print(features_sv_joined_pr_filtered['join_type'].value_counts().to_frame().reset_index())\n",
    "print('\\nAfter offtargets filtering')\n",
    "print(features_sv_no_offtereget_joined_pr['join_type'].value_counts().to_frame().reset_index())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1dec50d",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">GFF output for IGV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1134,
   "id": "f2648581",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create universal pr fot igv\n",
    "pr_for_igv = features_sv_no_offtereget_joined_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1135,
   "id": "0bd81149",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create empty GFF pr\n",
    "output_gff_columns = ['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'ID', 'color', 'type', 'Name', 'biotype', 'tag', 'transcript_id']\n",
    "output_gff_pr = pr.PyRanges(columns=output_gff_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1136,
   "id": "a51b5d79",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_16198/3865225082.py:4: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_16198/3865225082.py:5: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "/Users/andrejnekrasov/mambaforge/envs/utr3.venv/lib/python3.12/site-packages/pyranges/methods/concat.py:70: FutureWarning:\n",
      "\n",
      "The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5428"
      ]
     },
     "execution_count": 1136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add GFF data\n",
    "output_features_pr = pr_for_igv[['Chromosome', 'Feature', 'Start', 'End', 'Strand', 'Name', 'biotype', 'tag', 'ID']]\n",
    "output_features_pr = output_features_pr.drop_duplicates()\n",
    "output_features_pr['Source'] = 'GFF'\n",
    "output_features_pr['color'] = 'green'\n",
    "\n",
    "output_gff_pr = pr.concat([output_gff_pr, output_features_pr])\n",
    "len(output_gff_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1137,
   "id": "70a7aa35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16025"
      ]
     },
     "execution_count": 1137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add SV data\n",
    "output_sv_pr = pr_for_igv[['Chromosome', 'Start_sv', 'End_sv', 'svtype']]\n",
    "output_sv_pr = output_sv_pr.drop_duplicates()\n",
    "output_sv_pr['Source'] = 'gnomAD_SV'\n",
    "output_sv_pr['color'] = 'red'\n",
    "output_sv_pr['ID'] = pr_for_igv['sv_id']\n",
    "\n",
    "output_sv_pr = pr.PyRanges(output_sv_pr.rename(columns={'Start_sv': 'Start', 'End_sv': 'End', 'svtype': 'Feature'}))\n",
    "output_gff_pr = pr.concat([output_gff_pr, output_sv_pr])\n",
    "len(output_gff_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1138,
   "id": "596c3487",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26603"
      ]
     },
     "execution_count": 1138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add joins data\n",
    "output_joins_pr = pr_for_igv[['Chromosome', 'join_start', 'join_end', 'join_type']]\n",
    "output_joins_pr = output_joins_pr.drop_duplicates()\n",
    "output_joins_pr['Source'] = 'join'\n",
    "output_joins_pr['color'] = 'orange'\n",
    "output_joins_pr['ID'] = pr_for_igv['join_id']\n",
    "\n",
    "output_joins_pr = pr.PyRanges(output_joins_pr.rename(columns={'join_start': 'Start', 'join_end': 'End', 'join_type': 'Feature'}))\n",
    "output_gff_pr = pr.concat([output_gff_pr, output_joins_pr])\n",
    "len(output_gff_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1139,
   "id": "3ad5753d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30416"
      ]
     },
     "execution_count": 1139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add offtarget GFF (features) data\n",
    "output_offtargets_pr = features_sv_offtereget_joined_pr[['Chromosome', 'Start_offtarget', 'End_offtarget', 'Feature_offtarget', 'offtarget_id']]\n",
    "output_offtargets_pr = output_offtargets_pr.query('Start_offtarget == Start_offtarget')\n",
    "output_offtargets_pr = output_offtargets_pr.drop_duplicates()\n",
    "output_offtargets_pr['Source'] = 'GFF_offtarget'\n",
    "output_offtargets_pr['color'] = '#a3adac'\n",
    "\n",
    "output_offtargets_pr = pr.PyRanges(output_offtargets_pr.rename(columns={'Start_offtarget': 'Start', 'End_offtarget': 'End', 'Feature_offtarget': 'Feature', 'offtarget_id': 'ID'}))\n",
    "output_gff_pr = pr.concat([output_gff_pr, output_offtargets_pr])\n",
    "len(output_gff_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1140,
   "id": "5b16cad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gnomAD_SV</td>\n",
       "      <td>10597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>join</td>\n",
       "      <td>10578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GFF</td>\n",
       "      <td>5428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GFF_offtarget</td>\n",
       "      <td>3813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Source  count\n",
       "0      gnomAD_SV  10597\n",
       "1           join  10578\n",
       "2            GFF   5428\n",
       "3  GFF_offtarget   3813"
      ]
     },
     "execution_count": 1140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stats\n",
    "output_gff_pr['Source'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1141,
   "id": "f2cb8479",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill NA and save\n",
    "output_gff_pr = output_gff_pr.reset_index(drop=True)\n",
    "output_gff_pr = pr.PyRanges(output_gff_pr.fillna('.'))\n",
    "\n",
    "output_gff_pr.to_gff3(output_dir / 'regions_for_igv.gff3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541f18c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a29f19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "668dfa8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "662de559",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e9d4e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1fd440d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
