{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import json\n",
    "import pandas as pd, numpy as np, pyranges as pr\n",
    "import plotly.express as px\n",
    "import clickhouse_connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a644ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 500)\n",
    "pd.set_option('display.float_format', '{:.10f}'.format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Docs / notes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189c8ada",
   "metadata": {},
   "source": [
    "**Docs:**  \n",
    "* [PyRanges v1.x GitHub](https://github.com/pyranges/pyranges_1.x) / [PyRanges v1.x Docs](https://pyranges1.readthedocs.io/en/latest/index.html)\n",
    "\n",
    "**Notes:**  \n",
    " - !!! Нужно иметь ввиду, что позиции ФИЧ в GFF, например 3'-UTR, могут быть разделены на несколько интервалов, так как разделены, например, интронами.\n",
    "\n",
    "**Sources**\n",
    " - [miRBase](https://www.mirbase.org): hsa.gff - human\n",
    " - [multiMiR](http://multimir.org)\n",
    " - [TarBase](https://dianalab.e-ce.uth.gr/tarbasev9/downloads)\n",
    " - [miRTarBase](https://awi.cuhk.edu.cn/~miRTarBase/miRTarBase_2025/php/index.php)\n",
    " - [MiRanda](http://mirtoolsgallery.tech/mirtoolsgallery/node/1055): ineractions prediction\n",
    "\n",
    "**Tasks:**  \n",
    "1. Привести все к одной нумерации позиций: GFF - 1-based, BED - 0-based\n",
    "2. \n",
    "3. Проверить правильность offtarget_id. Сейчас это по факту позиции SV.\n",
    "\n",
    "**Commands:**  \n",
    "```{bash}\n",
    "mamba env export -n utr3.venv > environment.yml\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d6c257",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "rcmg_btk_sv_dir = main_path / 'data/rcmg_btk_sv'\n",
    "clinvar_dir = main_path / 'data/clinvar'\n",
    "mirna_dir = main_path / 'data/mirna'\n",
    "mirbase_dir = mirna_dir / 'mirbase'\n",
    "tarbase_dir = mirna_dir / 'tarbase'\n",
    "domino_dir = main_path / 'data/domino'\n",
    "other_dir = main_path / 'data/other'\n",
    "output_dir = main_path / 'data/output'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(rcmg_btk_sv_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(clinvar_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(mirbase_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(tarbase_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(domino_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(other_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(output_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def gunzip_file(input_filepath):\n",
    "    output_filepath = input_filepath.with_suffix('')\n",
    "    command = f'gunzip -c {input_filepath} > {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def sort_gff(input_filepath):\n",
    "    output_filepath = input_filepath.with_stem(input_filepath.stem + \".sorted\")\n",
    "    command = f'igvtools sort {input_filepath} {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def define_join_type(row):\n",
    "    feature_start = row['Start']\n",
    "    feature_end = row['End']\n",
    "    sv_start = row['Start_sv']\n",
    "    sv_end = row['End_sv']\n",
    "\n",
    "    join_type = np.nan\n",
    "    if feature_start <= sv_start and feature_end >= sv_end: # SV полностью в feature\n",
    "        join_type = 'sv_in_feature'\n",
    "    elif sv_start <= feature_start and sv_end >= feature_end: # feature полностью в sv\n",
    "        join_type = 'feature_in_sv'\n",
    "    elif sv_start < feature_start and sv_end <= sv_end:\n",
    "        join_type = 'sv_left_free'\n",
    "    elif sv_start > feature_start and sv_end >= sv_end:\n",
    "        join_type = 'sv_right_free'\n",
    "    elif sv_start == feature_start and sv_end == sv_end:\n",
    "        join_type = 'full_join'\n",
    "\n",
    "    return join_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Fetch data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.gff3.gz', refs_dir)\n",
    "fasta_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)\n",
    "clinvar_filepath = fetch_file('https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz', clinvar_dir)\n",
    "mirbase_filepath = fetch_file('https://www.mirbase.org/download/hsa.gff3', mirbase_dir)\n",
    "tarbase_filepath = fetch_file('https://dianalab.e-ce.uth.gr/tarbasev9/data/Homo_sapiens_TarBase-v9.tsv.gz', tarbase_dir)\n",
    "domino_filepath = fetch_file('https://domino.iob.ch/score_all_final_19.02.19.txt', domino_dir)\n",
    "\n",
    "so_terms_filepath = fetch_file('https://raw.githubusercontent.com/The-Sequence-Ontology/SO-Ontologies/refs/heads/master/Ontology_Files/so.json', other_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e76bd71b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RCMG BTK SV\n",
    "chromosomes = list(range(1, 23)) + ['X', 'Y', 'M']\n",
    "rcmg_btk_sv_filepaths = []\n",
    "\n",
    "try:\n",
    "    client = clickhouse_connect.get_client(host='192.168.86.236', user='rdr', password='Dru_Todd3')\n",
    "    db_name = \"ngs\"\n",
    "    client.database = db_name\n",
    "except:\n",
    "    for chrom in chromosomes:\n",
    "        output_filename =  f'rcmg_btk_sv_chr{chrom}.csv.gz'\n",
    "        output_filepath = rcmg_btk_sv_dir / output_filename\n",
    "        rcmg_btk_sv_filepaths.append(output_filepath)\n",
    "\n",
    "        if not output_filepath.exists():\n",
    "            rcmg_bkt_sv_df = client.query_df(f\"SELECT * from ngs.GENOME_hg38_2_sv_chr{chrom} WHERE SVTYPE='DEL';\")\n",
    "\n",
    "            rcmg_bkt_sv_df['Chromosome'] = str(chrom)\n",
    "            rcmg_bkt_sv_df.to_csv(output_filepath, index=False, compression='gzip')\n",
    "    try:\n",
    "        client.close()\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61697b2",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Main</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20681675",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_gunzipped_filepath = gunzip_file(gff_filepath)\n",
    "gff_sorted_filepath = sort_gff(gff_gunzipped_filepath)\n",
    "index_gff(gff_sorted_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b315f3",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Prepare data</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "007545ae",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Prepare GFF3 data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF\n",
    "gff_pr = pr.read_gff3(str(gff_filepath))\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Add additional metrics\n",
    "gff_pr['feature_len'] = (gff_pr['End'] - gff_pr['Start']).astype('int')\n",
    "gff_pr['feature_id'] = 'feature_' + gff_pr['Chromosome'].astype('str') + ':' + gff_pr['Start'].astype('str') + '-' + gff_pr['End'].astype('str')\n",
    "\n",
    "# Self annotation of GFF with mRNA (parent) data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level', 'feature_len', 'feature_id'] # Replace the data in these empty columns with mRNA data\n",
    "annotation_gff_pr = gff_pr[target_columns].dropna(subset='ID')\n",
    "\n",
    "annotation_suffix = '_annotation'\n",
    "annotated_gff_pr = gff_pr.merge(annotation_gff_pr, how='left', left_on='Parent', right_on='ID', suffixes=['', annotation_suffix])\n",
    "\n",
    "annotated_columns = [f'{column}{annotation_suffix}' for column in target_columns]\n",
    "fill_dict = dict(zip(target_columns, annotated_columns))\n",
    "for target, annotation in fill_dict.items():\n",
    "    annotated_gff_pr[target] = annotated_gff_pr[target].fillna(annotated_gff_pr[annotation])\n",
    "annotated_gff_pr = annotated_gff_pr.drop(annotated_columns, axis=1)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Create specific PRs (mRNAm, 3'-UTR, offtargets)\n",
    "mrna_pr = annotated_gff_pr.query('Feature == \"mRNA\"')\n",
    "three_utrs_pr = annotated_gff_pr.query('Feature == \"three_prime_UTR\"')\n",
    "\n",
    "offtarget_features = ['five_prime_UTR',\n",
    "                               'CDS',\n",
    "                               #'lnc_RNA',\n",
    "                               'ncRNA_gene',\n",
    "                               #'snRNA',\n",
    "                               #'snoRNA',\n",
    "                               #'scRNA',\n",
    "                               'rRNA',\n",
    "                               'tRNA',\n",
    "                               'processed_transcript',\n",
    "                               ]\n",
    "offtarget_features_pr = gff_pr.query('Feature in @offtarget_features')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# 3'-UTR biotype stats\n",
    "# Считаем именно уникальные вхождения, так как один 3'-UTR может быть разбить на несколько интервалов в GFF (разделен интронами).\n",
    "# В особенности это выражено у nonsense_mediated_decay транскриптов (нам они не пригодятся, но тем не менее).\n",
    "three_utrs_pr.drop_duplicates(subset='ID')['biotype'].value_counts().to_frame().reset_index()\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Keep only MANE_select / MANE_Select|Ensembl_canonical\n",
    "three_utrs_pr_filtered = three_utrs_pr[three_utrs_pr['tag'].str.contains('MANE_Select', na=False, regex=True)]\n",
    "\n",
    "# Stats\n",
    "# MANE_select genes count = 19437 (release_1.5) from summary file: https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.5/\n",
    "three_utrs_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ad65e24",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">gnomAD SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gnomAD SV\n",
    "# Keep only the necessary columns (63 of 600...) in SV\n",
    "with gzip.open(str(gnomad_sv_filepath), 'rt') as sv_file:\n",
    "    sv_header = sv_file.readline().strip().split('\\t')\n",
    "\n",
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ', 'FILTER']\n",
    "target_columns_indexes = [sv_header.index(i) for i in target_columns]\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End'] # Rename columns\n",
    "\n",
    "sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=target_columns_indexes, names=target_columns, comment='#')\n",
    "sv_df['Chromosome'] = sv_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "sv_pr = pr.PyRanges(sv_df)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Keep target SVs (DEL)\n",
    "sv_targets = ['DEL']\n",
    "sv_filter = ['PASS']\n",
    "target_sv_pr = sv_pr.query('svtype in @sv_targets and FILTER in @sv_filter')\n",
    "\n",
    "# Add columns\n",
    "target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
    "target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# SV percentiles by AF\n",
    "target_sv_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()\n",
    "\n",
    "# SV percentiles by SV length\n",
    "target_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb02969",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">ClinVar</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381eb36e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ClinVar\n",
    "clinvar_df = pd.read_csv(clinvar_filepath, sep='\\t', compression='gzip')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "clinvar_df = clinvar_df.replace(['na', '-'], np.nan)\n",
    "clinvar_df = clinvar_df.rename(columns={'Stop': 'End'})\n",
    "clinvar_df_filtered = clinvar_df.query('Assembly == \"GRCh38\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "693042b0",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">RCMG BTK SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b6188b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RCMG BTK SV\n",
    "rcmg_btk_sv_filepath = rcmg_btk_sv_dir / 'rcmg_btk_sv.csv.gz'\n",
    "\n",
    "if not rcmg_btk_sv_filepath.exists():\n",
    "    rcmg_bkt_sv_df = pd.DataFrame() # concated filtered aggregated\n",
    "    for rcmg_btk_sv_file in rcmg_btk_sv_filepaths:\n",
    "        print(rcmg_btk_sv_file)\n",
    "        current_rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_file), compression='gzip')\n",
    "\n",
    "        # Add columns\n",
    "        current_rcmg_bkt_sv_df['AF'] = current_rcmg_bkt_sv_df['AD'] / current_rcmg_bkt_sv_df['DP']\n",
    "        current_rcmg_bkt_sv_df['sv_len'] = abs(current_rcmg_bkt_sv_df['END'] - current_rcmg_bkt_sv_df['POS'])\n",
    "        current_rcmg_bkt_sv_df['sv_id'] = 'sv_' + current_rcmg_bkt_sv_df['Chromosome'].astype('str') + ':' + current_rcmg_bkt_sv_df['POS'].astype('str') + '-' + current_rcmg_bkt_sv_df['END'].astype('str')\n",
    "\n",
    "        # Drop columns\n",
    "        current_rcmg_bkt_sv_df = current_rcmg_bkt_sv_df.drop(columns=['REF', 'ALT'])\n",
    "\n",
    "        # Filter\n",
    "        current_rcmg_bkt_sv_df_filtered = current_rcmg_bkt_sv_df.query('caller == \"manta\"')\n",
    "\n",
    "        # Aggregate\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered.groupby(['Chromosome', 'POS', 'END', 'SVTYPE', 'caller', 'sv_len', 'sv_id'], as_index=False).agg({\n",
    "            'SID': [list, 'count'],\n",
    "            'DP': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AD': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AF': [list, 'min', 'median', 'mean', 'max']\n",
    "        })\n",
    "\n",
    "        # Rename columns\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated.columns = ['_'.join(col).strip('_') for col in current_rcmg_bkt_sv_df_filtered_aggregated.columns]\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.rename(columns={'POS': 'Start', 'END': 'End'})\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.reset_index(drop=True)\n",
    "\n",
    "        # Concat\n",
    "        rcmg_bkt_sv_df['Chromosome'] = rcmg_bkt_sv_df['Chromosome'].astype('str')\n",
    "        rcmg_bkt_sv_df = pd.concat([rcmg_bkt_sv_df, current_rcmg_bkt_sv_df_filtered_aggregated], ignore_index=False)\n",
    "    rcmg_bkt_sv_df.to_csv(str(rcmg_btk_sv_filepath), index=False, compression='gzip')\n",
    "else:\n",
    "    rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_filepath), compression='gzip')\n",
    "    rcmg_bkt_sv_df['Chromosome'] = rcmg_bkt_sv_df['Chromosome'].astype('str')\n",
    "\n",
    "rcmg_btk_sv_pr = pr.PyRanges(rcmg_bkt_sv_df)\n",
    "rcmg_btk_sv_pr_filtered = rcmg_btk_sv_pr.query('SID_count < 30 and sv_len < 1000000')\n",
    "rcmg_btk_sv_pr_filtered = rcmg_btk_sv_pr_filtered.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dc5b242",
   "metadata": {},
   "outputs": [],
   "source": [
    "rcmg_btk_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d32aa6",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">miRBase</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dfaa41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# miRBase\n",
    "mirbase_pr = pr.read_gff3(str(mirbase_filepath))\n",
    "mirbase_pr['Chromosome'] = mirbase_pr['Chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "718e8ef5",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">TarBase</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a86eb49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TarBase\n",
    "tarbase_df = pd.read_csv(str(tarbase_filepath), sep='\\t', compression='gzip')\n",
    "tarbase_df = tarbase_df.rename(columns={'chromosome': 'Chromosome', 'start': 'Start', 'end': 'End', 'strand': 'Strand'})\n",
    "tarbase_df['Chromosome'] = tarbase_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "tarbase_pr = pr.PyRanges(tarbase_df)\n",
    "tarbase_pr = tarbase_pr.dropna(subset=['Chromosome', 'Start', 'End'], how='any')\n",
    "tarbase_pr[['Start', 'End']] = tarbase_pr[['Start', 'End']].astype('int64')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f18e8a51",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Sequence Ontology terms</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0a4271f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SO terms\n",
    "with open(so_terms_filepath, 'r') as so_terms_file:\n",
    "    so_terms = json.load(so_terms_file)\n",
    "\n",
    "so_terms_filtered = {\n",
    "    'lbl': [],\n",
    "    'definition': []\n",
    "}\n",
    "for i in so_terms['graphs'][0]['nodes']:\n",
    "    lbl = i.get('lbl', np.nan)\n",
    "    definition = i.get('meta', {}).get('definition', {}).get('val', np.nan)\n",
    "    so_terms_filtered['lbl'].append(lbl)\n",
    "    so_terms_filtered['definition'].append(definition)\n",
    "\n",
    "so_terms_df = pd.DataFrame.from_dict(so_terms_filtered)\n",
    "so_terms_df['lbl'] = so_terms_df['lbl'].str.capitalize().replace('_', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b001f54",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">EDA (Exploratory Data Analysis)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788e1cc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF features stats\n",
    "gff_features = gff_pr.Feature.value_counts().to_frame().reset_index()\n",
    "gff_features['tmp_id'] = gff_features['Feature'].str.capitalize().replace('_', '')\n",
    "gff_features = gff_features.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_features = gff_features.drop(['lbl', 'tmp_id'], axis=1)\n",
    "#pd.DataFrame(gff_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e7f9eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF biotypes stats\n",
    "gff_biotypes = gff_pr.biotype.value_counts().to_frame().reset_index()\n",
    "gff_biotypes['tmp_id'] = gff_biotypes['biotype'].str.capitalize().replace('_', '')\n",
    "gff_biotypes = gff_biotypes.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_biotypes = gff_biotypes.drop(['lbl', 'tmp_id'], axis=1)\n",
    "#pd.DataFrame(gff_biotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cfc34fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sv_pr.svtype.value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a2175ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "tarbase_pr['gene_location'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e0ca86",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Join GFF3 x SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64921092",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_for_join_pr = rcmg_btk_sv_pr_filtered # target_sv_pr / rcmg_btk_sv_pr_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b14ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join GFF x SV\n",
    "features_sv_joined_pr = three_utrs_pr_filtered.join_overlaps(sv_for_join_pr, suffix='_sv', report_overlap_column='join_len', strand_behavior='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6885c5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add additional columns (join_start, join_end, join_id)\n",
    "features_sv_joined_pr['join_start'] = np.maximum(features_sv_joined_pr['Start'], features_sv_joined_pr['Start_sv'])\n",
    "features_sv_joined_pr['join_end'] = np.minimum(features_sv_joined_pr['End'], features_sv_joined_pr['End_sv'])\n",
    "features_sv_joined_pr['join_id'] = 'join_' + features_sv_joined_pr['Chromosome'].astype('str') + ':' + features_sv_joined_pr['join_start'].astype('str') + '-' + features_sv_joined_pr['join_end'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ec54b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define joint type\n",
    "features_sv_joined_pr['join_type'] = features_sv_joined_pr.apply(define_join_type, axis=1)\n",
    "features_sv_joined_pr['join_type'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ae52f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count joints by SV\n",
    "joints_count_by_sv = features_sv_joined_pr['sv_id'].value_counts().to_frame().reset_index()\n",
    "joints_count_by_sv = joints_count_by_sv.rename(columns={'count': 'joins_count'})\n",
    "\n",
    "# Add joints count data\n",
    "features_sv_joined_pr = features_sv_joined_pr.merge(joints_count_by_sv, how='left')\n",
    "#len(features_sv_joined_pr)\n",
    "\n",
    "# Filter joins by join type\n",
    "targey_joint_types = ['sv_in_feature', 'sv_right_free', 'sv_left_free']\n",
    "features_sv_joined_pr_filtered = features_sv_joined_pr.query('join_type in @targey_joint_types and ((Strand == \"+\" and join_type == \"sv_right_free\") or (Strand == \"-\" and join_type == \"sv_left_free\") or join_type == \"sv_in_feature\") and joins_count == 1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b449576",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join type stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a69d46c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stats by join type\n",
    "features_sv_joined_pr_filtered.groupby('join_type')['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc037975",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_joined_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4135ee",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Offtarget features exclusion</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ce637a",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">features_sv_joined_pr_filtered copy</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab7f6e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Offtarget stats\n",
    "features_sv_joined_pr_filtered_cp = features_sv_joined_pr_filtered.copy()\n",
    "features_sv_joined_pr_filtered_cp[['Start_tmp', 'End_tmp']] = features_sv_joined_pr_filtered_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_joined_pr_filtered_cp[['Start', 'End']] = features_sv_joined_pr_filtered_cp[['Start_sv', 'End_sv']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157fc717",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Filter offtargets</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dc11ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exclude offtargets (SV x offtarget overlaps)\n",
    "features_sv_no_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.overlap(offtarget_features_pr, invert=True, strand_behavior='ignore')\n",
    "\n",
    "features_sv_no_offtereget_joined_pr[['Start', 'End']] = features_sv_no_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_joined_pr = features_sv_no_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "len(features_sv_no_offtereget_joined_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "623083fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Joint type stats after offtargets exclution\n",
    "print('Before offtargets filtering')\n",
    "print(features_sv_joined_pr_filtered['join_type'].value_counts().to_frame().reset_index())\n",
    "print('\\nAfter offtargets filtering')\n",
    "print(features_sv_no_offtereget_joined_pr['join_type'].value_counts().to_frame().reset_index())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ddbe504",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Offtarget stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd478da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# features_sv_offtereget_joined_pr - только для отслеживания пересечений с offtarget. Фильтрованный от offtarget pr - features_sv_no_offtereget_joined_pr (через overlap, потому что join находя пересечения с offtarget (через left) также оставляет оригинальную строку).DS_Storefeatures_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len')\n",
    "features_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len', strand_behavior='ignore')\n",
    "\n",
    "features_sv_offtereget_joined_pr[['Start', 'End']] = features_sv_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_offtereget_joined_pr = features_sv_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_offtereget_joined_pr['offtarget_id'] = 'offtarget_' + features_sv_offtereget_joined_pr['Chromosome'].astype('str') + ':' + features_sv_offtereget_joined_pr['Start_offtarget'].astype('str') + '-' + features_sv_offtereget_joined_pr['End_offtarget'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c516d79f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Features offtargets stats\n",
    "# Суммарные значения данной статистики будут выше чем разница всех джоинов - оффтартет джоинов, так как один и тот же SV (напр. DEL) могут пересекать разные изоформы одного и того же транскрипта.\n",
    "features_sv_offtereget_joined_pr.groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])\n",
    "#features_sv_offtereget_joined_pr.drop_duplicates(subset='sv_id').groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d8bf654",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">miRNA join</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1e4a88",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">features_sv_no_offtereget_joined_pr copy</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e88e7aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Offtarget stats\n",
    "features_sv_no_offtereget_joined_pr_cp = features_sv_no_offtereget_joined_pr.copy()\n",
    "features_sv_no_offtereget_joined_pr_cp[['Start_tmp', 'End_tmp']] = features_sv_no_offtereget_joined_pr_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_no_offtereget_joined_pr_cp[['Start', 'End']] = features_sv_no_offtereget_joined_pr_cp[['join_start', 'join_end']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99af8a69",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c871563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join TarBase\n",
    "features_sv_no_offtereget_tarbase_joined_pr = features_sv_no_offtereget_joined_pr_cp.join_overlaps(tarbase_pr, suffix='_tarbase', strand_behavior='ignore', report_overlap_column='tarbase_join_len')\n",
    "\n",
    "features_sv_no_offtereget_tarbase_joined_pr[['Start', 'End']] = features_sv_no_offtereget_tarbase_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_tarbase_joined_pr = features_sv_no_offtereget_tarbase_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_no_offtereget_tarbase_joined_pr['tarbase_id'] = 'tarbase_' + features_sv_no_offtereget_tarbase_joined_pr['Chromosome'].astype('str') + ':' + features_sv_no_offtereget_tarbase_joined_pr['Start_tarbase'].astype('str') + '-' + features_sv_no_offtereget_tarbase_joined_pr['End_tarbase'].astype('str') + ':' + features_sv_no_offtereget_tarbase_joined_pr['mirna_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6984057b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#features_sv_no_offtereget_tarbase_joined_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "653cb4b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['gene_location'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae28975b",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['interaction_group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e230a118",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['regulation'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "084ddca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['transcript_name'].value_counts().to_frame().reset_index().to_csv(output_dir / 'mirna_genes.csv', index=False)\n",
    "features_sv_no_offtereget_tarbase_joined_pr['transcript_name'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eaea44e",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['experimental_method'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53825b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['join_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e67fb804",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['join_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f2cecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['Chromosome'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e3fc4b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9903c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(features_sv_no_offtereget_tarbase_joined_pr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1dec50d",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">GFF output for IGV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1769e350",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr_for_igv = features_sv_no_offtereget_tarbase_joined_pr\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Add GFF\n",
    "output_features_pr = pr_for_igv[['Chromosome', 'Feature', 'Start', 'End', 'Strand', 'Name', 'biotype', 'tag', 'ID']]\n",
    "output_features_pr = output_features_pr.drop_duplicates()\n",
    "output_features_pr['Source'] = 'GFF'\n",
    "output_features_pr['color'] = 'blue'\n",
    "\n",
    "# Add gnomAD SV\n",
    "'''output_sv_pr = pr_for_igv[['Chromosome', 'Start_sv', 'End_sv', 'svtype', 'name', 'sv_id', 'AF', 'N_HET', 'N_HOMALT']]\n",
    "output_sv_pr = output_sv_pr.drop_duplicates()\n",
    "output_sv_pr['Source'] = 'gnomAD_SV'\n",
    "output_sv_pr['color'] = 'red'\n",
    "output_sv_pr = pr.PyRanges(output_sv_pr.rename(columns={'Start_sv': 'Start', 'End_sv': 'End', 'svtype': 'Feature', 'name': 'Name', 'sv_id': 'ID'}))'''\n",
    "\n",
    "# Add RCMG BTK SV\n",
    "output_sv_pr = pr_for_igv[['Chromosome', 'Start_sv', 'End_sv', 'SVTYPE', 'SID_count', 'DP_median', 'AF_median', 'SID_list', 'sv_id']]\n",
    "output_sv_pr = output_sv_pr.drop_duplicates()\n",
    "output_sv_pr['Source'] = 'RCMG_BTK_SV'\n",
    "output_sv_pr['color'] = 'red'\n",
    "output_sv_pr = pr.PyRanges(output_sv_pr.rename(columns={'Start_sv': 'Start', 'End_sv': 'End', 'svtype': 'Feature', 'name': 'Name', 'sv_id': 'ID'}))\n",
    "\n",
    "# Add joins data\n",
    "output_joins_pr = pr_for_igv[['Chromosome', 'join_start', 'join_end', 'join_type', 'join_id']]\n",
    "output_joins_pr = output_joins_pr.drop_duplicates()\n",
    "output_joins_pr['Source'] = 'join'\n",
    "output_joins_pr['color'] = 'orange'\n",
    "output_joins_pr = pr.PyRanges(output_joins_pr.rename(columns={'join_start': 'Start', 'join_end': 'End', 'join_type': 'Feature', 'join_id': 'ID'}))\n",
    "\n",
    "# Add miRNA TarBase\n",
    "output_tarbase_pr = pr_for_igv[['Chromosome', 'Start_tarbase', 'End_tarbase', 'Strand_tarbase', 'mirna_name', 'interaction_group', 'mirna_id', 'tarbase_id']]\n",
    "output_tarbase_pr = output_tarbase_pr.drop_duplicates(subset=['mirna_id', 'tarbase_id'])\n",
    "output_tarbase_pr['Source'] = 'tarbase'\n",
    "output_tarbase_pr['color'] = 'green'\n",
    "output_tarbase_pr = pr.PyRanges(output_tarbase_pr.rename(columns={'Start_tarbase': 'Start', 'End_tarbase': 'End', 'Strand_tarbase': 'Strand', 'interaction_group': 'Feature', 'mirna_name': 'Name', 'tarbase_id': 'ID'}))\n",
    "\n",
    "# Add offtarget GFF (features) data\n",
    "output_offtargets_pr = features_sv_offtereget_joined_pr[['Chromosome', 'Start_offtarget', 'End_offtarget', 'Feature_offtarget', 'offtarget_id']]\n",
    "output_offtargets_pr = output_offtargets_pr.query('Start_offtarget == Start_offtarget')\n",
    "output_offtargets_pr = output_offtargets_pr.drop_duplicates()\n",
    "output_offtargets_pr['Source'] = 'GFF_offtarget'\n",
    "output_offtargets_pr['color'] = '#d9d9d9'\n",
    "output_offtargets_pr = pr.PyRanges(output_offtargets_pr.rename(columns={'Start_offtarget': 'Start', 'End_offtarget': 'End', 'Feature_offtarget': 'Feature', 'offtarget_id': 'ID'}))\n",
    "\n",
    "output_igv_pr = pd.concat([output_features_pr, output_sv_pr, output_joins_pr, output_tarbase_pr, output_offtargets_pr], ignore_index=True)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Fill NA and save\n",
    "output_igv_pr = output_igv_pr.reset_index(drop=True)\n",
    "#output_igv_pr = pr.PyRanges(output_igv_pr.fillna('.'))\n",
    "\n",
    "output_igv_pr.to_gff3(output_dir / 'regions_for_igv.gff3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b16cad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stats\n",
    "output_igv_pr['Source'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541f18c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a29f19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "668dfa8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "662de559",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e9d4e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1fd440d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
