{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import json\n",
    "import pandas as pd, numpy as np, pyranges as pr\n",
    "import plotly.express as px\n",
    "import clickhouse_connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a644ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 500)\n",
    "pd.set_option('display.float_format', '{:.10f}'.format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Docs / notes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189c8ada",
   "metadata": {},
   "source": [
    "**Docs:**  \n",
    "* [PyRanges v1.x GitHub](https://github.com/pyranges/pyranges_1.x) / [PyRanges v1.x Docs](https://pyranges1.readthedocs.io/en/latest/index.html)\n",
    "\n",
    "**Notes:**  \n",
    " - !!! Нужно иметь ввиду, что позиции ФИЧ в GFF, например 3'-UTR, могут быть разделены на несколько интервалов, так как разделены, например, интронами.\n",
    "\n",
    "**Sources**\n",
    " - [miRBase](https://www.mirbase.org): hsa.gff - human\n",
    " - [multiMiR](http://multimir.org)\n",
    " - [TarBase](https://dianalab.e-ce.uth.gr/tarbasev9/downloads)\n",
    " - [miRTarBase](https://awi.cuhk.edu.cn/~miRTarBase/miRTarBase_2025/php/index.php)\n",
    " - [MiRanda](http://mirtoolsgallery.tech/mirtoolsgallery/node/1055): ineractions prediction\n",
    "\n",
    "**Tasks:**  \n",
    "1. Привести все к одной нумерации позиций: GFF - 1-based, BED - 0-based\n",
    "2. \n",
    "3. Проверить правильность offtarget_id. Сейчас это по факту позиции SV.\n",
    "\n",
    "**Commands:**  \n",
    "```{bash}\n",
    "mamba env export -n utr3.venv > environment.yml\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d6c257",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "rcmg_btk_sv_dir = main_path / 'data/rcmg_btk_sv'\n",
    "clinvar_dir = main_path / 'data/clinvar'\n",
    "mirna_dir = main_path / 'data/mirna'\n",
    "mirbase_dir = mirna_dir / 'mirbase'\n",
    "tarbase_dir = mirna_dir / 'tarbase'\n",
    "other_dir = main_path / 'data/other'\n",
    "output_dir = main_path / 'data/output'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(rcmg_btk_sv_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(clinvar_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(mirbase_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(tarbase_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(other_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(output_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def gunzip_file(input_filepath):\n",
    "    output_filepath = input_filepath.with_suffix('')\n",
    "    command = f'gunzip -c {input_filepath} > {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def sort_gff(input_filepath):\n",
    "    output_filepath = input_filepath.with_stem(input_filepath.stem + \".sorted\")\n",
    "    command = f'igvtools sort {input_filepath} {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def define_join_type(row):\n",
    "    feature_start = row['Start']\n",
    "    feature_end = row['End']\n",
    "    sv_start = row['Start_sv']\n",
    "    sv_end = row['End_sv']\n",
    "\n",
    "    join_type = np.nan\n",
    "    if feature_start <= sv_start and feature_end >= sv_end: # SV полностью в feature\n",
    "        join_type = 'sv_in_feature'\n",
    "    elif sv_start <= feature_start and sv_end >= feature_end: # feature полностью в sv\n",
    "        join_type = 'feature_in_sv'\n",
    "    elif sv_start < feature_start and sv_end <= sv_end:\n",
    "        join_type = 'sv_left_free'\n",
    "    elif sv_start > feature_start and sv_end >= sv_end:\n",
    "        join_type = 'sv_right_free'\n",
    "    elif sv_start == feature_start and sv_end == sv_end:\n",
    "        join_type = 'full_join'\n",
    "\n",
    "    return join_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Fetch data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.gff3.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.dna.toplevel.fa.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/gnomad/gnomad.v4.1.sv.sites.bed.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/clinvar/variant_summary.txt.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/mirna/mirbase/hsa.gff3' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/mirna/tarbase/Homo_sapiens_TarBase-v9.tsv.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/other/so.json' already there; not retrieving.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.gff3.gz', refs_dir)\n",
    "fasta_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)\n",
    "clinvar_filepath = fetch_file('https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz', clinvar_dir)\n",
    "mirbase_filepath = fetch_file('https://www.mirbase.org/download/hsa.gff3', mirbase_dir)\n",
    "tarbase_filepath = fetch_file('https://dianalab.e-ce.uth.gr/tarbasev9/data/Homo_sapiens_TarBase-v9.tsv.gz', tarbase_dir)\n",
    "\n",
    "so_terms_filepath = fetch_file('https://raw.githubusercontent.com/The-Sequence-Ontology/SO-Ontologies/refs/heads/master/Ontology_Files/so.json', other_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e76bd71b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'ConnectTimeoutError(<HTTPConnection(host='192.168.86.236', port=8123) at 0x16cf16810>, 'Connection to 192.168.86.236 timed out. (connect timeout=10)')': /?\n",
      "WARNING:clickhouse_connect.driver.httpclient:Unexpected Http Driver Exception\n"
     ]
    }
   ],
   "source": [
    "# RCMG BTK SV\n",
    "chromosomes = list(range(1, 23)) + ['X', 'Y', 'M']\n",
    "rcmg_btk_sv_filepaths = []\n",
    "\n",
    "try:\n",
    "    client = clickhouse_connect.get_client(host='192.168.86.236', user='rdr', password='Dru_Todd3')\n",
    "    db_name = \"ngs\"\n",
    "    client.database = db_name\n",
    "except:\n",
    "    for chrom in chromosomes:\n",
    "        output_filename =  f'rcmg_btk_sv_chr{chrom}.csv.gz'\n",
    "        output_filepath = rcmg_btk_sv_dir / output_filename\n",
    "        rcmg_btk_sv_filepaths.append(output_filepath)\n",
    "\n",
    "        if not output_filepath.exists():\n",
    "            rcmg_bkt_sv_df = client.query_df(f\"SELECT * from ngs.GENOME_hg38_2_sv_chr{chrom} WHERE SVTYPE='DEL';\")\n",
    "\n",
    "            rcmg_bkt_sv_df['Chromosome'] = chrom\n",
    "            rcmg_bkt_sv_df.to_csv(output_filepath, index=False, compression='gzip')\n",
    "    try:\n",
    "        client.close()\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61697b2",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Prepare data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "20681675",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sorting /Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.gff3  -> /Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.sorted.gff3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: A restricted method in java.lang.System has been called\n",
      "WARNING: java.lang.System::load has been called by org.xerial.snappy.SnappyLoader in module snappy.java (file:/Users/andrejnekrasov/mambaforge/envs/utr3.venv/share/igvtools-2.17.3-0/lib/snappy-java-1.1.10.5.jar)\n",
      "WARNING: Use --enable-native-access=snappy.java to avoid a warning for callers in this module\n",
      "WARNING: Restricted methods will be blocked in a future release unless native access is enabled\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "gff_gunzipped_filepath = gunzip_file(gff_filepath)\n",
    "gff_sorted_filepath = sort_gff(gff_gunzipped_filepath)\n",
    "index_gff(gff_sorted_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b315f3",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Create PR/DF<span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "007545ae",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Prepare GFF3 data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF\n",
    "gff_pr = pr.read_gff3(str(gff_filepath))\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Add additional metrics\n",
    "gff_pr['feature_len'] = (gff_pr['End'] - gff_pr['Start']).astype('int')\n",
    "gff_pr['feature_id'] = 'feature_' + gff_pr['Chromosome'].astype('str') + ':' + gff_pr['Start'].astype('str') + '-' + gff_pr['End'].astype('str')\n",
    "\n",
    "# Self annotation of GFF with mRNA (parent) data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level', 'feature_len', 'feature_id'] # Replace the data in these empty columns with mRNA data\n",
    "annotation_gff_pr = gff_pr[target_columns].dropna(subset='ID')\n",
    "\n",
    "annotation_suffix = '_annotation'\n",
    "annotated_gff_pr = gff_pr.merge(annotation_gff_pr, how='left', left_on='Parent', right_on='ID', suffixes=['', annotation_suffix])\n",
    "\n",
    "annotated_columns = [f'{column}{annotation_suffix}' for column in target_columns]\n",
    "fill_dict = dict(zip(target_columns, annotated_columns))\n",
    "for target, annotation in fill_dict.items():\n",
    "    annotated_gff_pr[target] = annotated_gff_pr[target].fillna(annotated_gff_pr[annotation])\n",
    "annotated_gff_pr = annotated_gff_pr.drop(annotated_columns, axis=1)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Create specific PRs (mRNAm, 3'-UTR, offtargets)\n",
    "mrna_pr = annotated_gff_pr.query('Feature == \"mRNA\"')\n",
    "three_utrs_pr = annotated_gff_pr.query('Feature == \"three_prime_UTR\"')\n",
    "\n",
    "offtarget_features = ['five_prime_UTR',\n",
    "                               'CDS',\n",
    "                               #'lnc_RNA',\n",
    "                               'ncRNA_gene',\n",
    "                               #'snRNA',\n",
    "                               #'snoRNA',\n",
    "                               #'scRNA',\n",
    "                               'rRNA',\n",
    "                               'tRNA',\n",
    "                               'processed_transcript',\n",
    "                               ]\n",
    "offtarget_features_pr = gff_pr.query('Feature in @offtarget_features')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# 3'-UTR biotype stats\n",
    "# Считаем именно уникальные вхождения, так как один 3'-UTR может быть разбить на несколько интервалов в GFF (разделен интронами).\n",
    "# В особенности это выражено у nonsense_mediated_decay транскриптов (нам они не пригодятся, но тем не менее).\n",
    "three_utrs_pr.drop_duplicates(subset='ID')['biotype'].value_counts().to_frame().reset_index()\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Keep only MANE_select / MANE_Select|Ensembl_canonical\n",
    "three_utrs_pr_filtered = three_utrs_pr[three_utrs_pr['tag'].str.contains('MANE_Select', na=False, regex=True)]\n",
    "\n",
    "# Stats\n",
    "# MANE_select genes count = 19437 (release_1.5) from summary file: https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.5/\n",
    "three_utrs_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ad65e24",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">gnomAD SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/1131200204.py:10: DtypeWarning: Columns (10,11,24,25,26,28,29,30,31,32,33,36,37,38,39,40,42,57,58,59) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=target_columns_indexes, names=target_columns, comment='#')\n"
     ]
    }
   ],
   "source": [
    "# gnomAD SV\n",
    "# Keep only the necessary columns (63 of 600...) in SV\n",
    "with gzip.open(str(gnomad_sv_filepath), 'rt') as sv_file:\n",
    "    sv_header = sv_file.readline().strip().split('\\t')\n",
    "\n",
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ', 'FILTER']\n",
    "target_columns_indexes = [sv_header.index(i) for i in target_columns]\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End'] # Rename columns\n",
    "\n",
    "sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=target_columns_indexes, names=target_columns, comment='#')\n",
    "sv_df['Chromosome'] = sv_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "sv_pr = pr.PyRanges(sv_df)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Keep target SVs (DEL)\n",
    "sv_targets = ['DEL']\n",
    "sv_filter = ['PASS']\n",
    "target_sv_pr = sv_pr.query('svtype in @sv_targets and FILTER in @sv_filter')\n",
    "\n",
    "# Add columns\n",
    "target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
    "target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# SV percentiles by AF\n",
    "target_sv_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()\n",
    "\n",
    "# SV percentiles by SV length\n",
    "target_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb02969",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">ClinVar</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381eb36e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/1108798041.py:2: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinvar_df = pd.read_csv(clinvar_filepath, sep='\\t', compression='gzip')\n"
     ]
    }
   ],
   "source": [
    "# ClinVar\n",
    "clinvar_df = pd.read_csv(clinvar_filepath, sep='\\t', compression='gzip')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "clinvar_df = clinvar_df.replace(['na', '-'], np.nan)\n",
    "clinvar_df = clinvar_df.rename(columns={'Stop': 'End'})\n",
    "clinvar_df_filtered = clinvar_df.query('Assembly == \"GRCh38\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "693042b0",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">RCMG BTK SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b6188b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RCMG BTK SV\n",
    "rcmg_btk_sv_filepath = rcmg_btk_sv_dir / 'rcmg_btk_sv.csv.gz'\n",
    "\n",
    "if not rcmg_btk_sv_filepath.exists():\n",
    "    rcmg_bkt_sv_df = pd.DataFrame() # concated filtered aggregated\n",
    "    for rcmg_btk_sv_file in rcmg_btk_sv_filepaths:\n",
    "        print(rcmg_btk_sv_file)\n",
    "        current_rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_file), compression='gzip')\n",
    "\n",
    "        # Add columns\n",
    "        current_rcmg_bkt_sv_df['AF'] = current_rcmg_bkt_sv_df['AD'] / current_rcmg_bkt_sv_df['DP']\n",
    "        current_rcmg_bkt_sv_df['SVLEN'] = abs(current_rcmg_bkt_sv_df['END'] - current_rcmg_bkt_sv_df['POS'])\n",
    "\n",
    "        # Drop columns\n",
    "        current_rcmg_bkt_sv_df = current_rcmg_bkt_sv_df.drop(columns=['REF', 'ALT'])\n",
    "\n",
    "        # Filter\n",
    "        current_rcmg_bkt_sv_df_filtered = current_rcmg_bkt_sv_df.query('caller == \"manta\"')\n",
    "\n",
    "        # Aggregate\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered.groupby(['Chromosome', 'POS', 'END', 'SVTYPE', 'caller', 'SVLEN'], as_index=False).agg({\n",
    "            'SID': [list, 'count'],\n",
    "            'DP': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AD': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AF': [list, 'min', 'median', 'mean', 'max']\n",
    "        })\n",
    "\n",
    "        # Rename columns\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated.columns = ['_'.join(col).strip('_') for col in current_rcmg_bkt_sv_df_filtered_aggregated.columns]\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.rename(columns={'POS': 'Start', 'END': 'End'})\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.reset_index(drop=True)\n",
    "\n",
    "        # Concat\n",
    "        rcmg_bkt_sv_df = pd.concat([rcmg_bkt_sv_df, current_rcmg_bkt_sv_df_filtered_aggregated], ignore_index=False)\n",
    "    rcmg_bkt_sv_df.to_csv(str(rcmg_btk_sv_filepath), index=False, compression='gzip')\n",
    "else:\n",
    "    rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_filepath), compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d32aa6",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">miRBase</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9dfaa41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# miRBase\n",
    "mirbase_pr = pr.read_gff3(str(mirbase_filepath))\n",
    "mirbase_pr['Chromosome'] = mirbase_pr['Chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "718e8ef5",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">TarBase</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7a86eb49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/1318660838.py:2: DtypeWarning: Columns (5,14,15,19,21) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  tarbase_df = pd.read_csv(str(tarbase_filepath), sep='\\t', compression='gzip')\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/1318660838.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tarbase_pr[['Start', 'End']] = tarbase_pr[['Start', 'End']].astype('int64')\n"
     ]
    }
   ],
   "source": [
    "# TarBase\n",
    "tarbase_df = pd.read_csv(str(tarbase_filepath), sep='\\t', compression='gzip')\n",
    "tarbase_df = tarbase_df.rename(columns={'chromosome': 'Chromosome', 'start': 'Start', 'end': 'End', 'strand': 'Strand'})\n",
    "tarbase_df['Chromosome'] = tarbase_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "tarbase_pr = pr.PyRanges(tarbase_df)\n",
    "tarbase_pr = tarbase_pr.dropna(subset=['Chromosome', 'Start', 'End'], how='any')\n",
    "tarbase_pr[['Start', 'End']] = tarbase_pr[['Start', 'End']].astype('int64')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f18e8a51",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Sequence Ontology terms</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d0a4271f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SO terms\n",
    "with open(so_terms_filepath, 'r') as so_terms_file:\n",
    "    so_terms = json.load(so_terms_file)\n",
    "\n",
    "so_terms_filtered = {\n",
    "    'lbl': [],\n",
    "    'definition': []\n",
    "}\n",
    "for i in so_terms['graphs'][0]['nodes']:\n",
    "    lbl = i.get('lbl', np.nan)\n",
    "    definition = i.get('meta', {}).get('definition', {}).get('val', np.nan)\n",
    "    so_terms_filtered['lbl'].append(lbl)\n",
    "    so_terms_filtered['definition'].append(definition)\n",
    "\n",
    "so_terms_df = pd.DataFrame.from_dict(so_terms_filtered)\n",
    "so_terms_df['lbl'] = so_terms_df['lbl'].str.capitalize().replace('_', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b001f54",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">EDA (Exploratory Data Analysis)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "788e1cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>count</th>\n",
       "      <th>definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon</td>\n",
       "      <td>3683354</td>\n",
       "      <td>A region of the transcript sequence within a gene which is not removed from the primary RNA transcript by RNA splicing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>2284414</td>\n",
       "      <td>A contiguous sequence which begins with, and includes, a start codon and ends with, and includes, a stop codon.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>426629</td>\n",
       "      <td>A region at the 5' end of a mature transcript (preceding the initiation codon) that is not translated into a protein.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>340592</td>\n",
       "      <td>A region at the 3' end of a mature transcript (following the stop codon) that is not translated into a protein.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>233603</td>\n",
       "      <td>Messenger RNA is the intermediate molecule between DNA and protein. It includes UTR and coding sequences. It does not contain introns.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>225629</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>biological_region</td>\n",
       "      <td>182818</td>\n",
       "      <td>A region defined by its disposition to be involved in a biological process.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>42124</td>\n",
       "      <td>A gene that encodes a non-coding RNA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>transcript</td>\n",
       "      <td>28808</td>\n",
       "      <td>An RNA synthesized on a DNA or RNA template by an RNA polymerase.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene</td>\n",
       "      <td>21571</td>\n",
       "      <td>A region (or regions) that includes all of the sequence elements necessary to encode a functional transcript. A gene may include regulatory regions, transcribed regions and/or other functional sequence regions.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>15207</td>\n",
       "      <td>A non functional descendant of a transcript, part of a pseudogene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>15204</td>\n",
       "      <td>A sequence that closely resembles a known functional gene, at another locus within a genome, that is non-functional as a consequence of (usually several) mutations that prevent either its transcription or translation (or both). In general, pseudogenes result from either reverse transcription of a transcript of their \"normal\" paralog (SO:0000043) (in which case the pseudogene typically lacks introns and includes a poly(A) tail) or from recombination (SO:0000044) (in which case the pseudogene ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>1915</td>\n",
       "      <td>A small nuclear RNA molecule involved in pre-mRNA splicing and processing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>1879</td>\n",
       "      <td>Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>unconfirmed_transcript</td>\n",
       "      <td>1108</td>\n",
       "      <td>This is used for non-spliced EST clusters that have polyA features. This category has been specifically created for the ENCODE project to highlight regions that could indicate the presence of protein coding genes that require experimental validation, either by 5' RACE or RT-PCR to extend the transcripts, or by confirming expression of the putatively-encoded peptide with specific antibodies.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>942</td>\n",
       "      <td>Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>V_gene_segment</td>\n",
       "      <td>253</td>\n",
       "      <td>Germline genomic DNA including L-part1, V-intron and V-exon, with the 5' UTR and 3' UTR.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>scaffold</td>\n",
       "      <td>169</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>J_gene_segment</td>\n",
       "      <td>97</td>\n",
       "      <td>Germline genomic DNA of an immunoglobulin/T-cell receptor gene including J-region with 5' UTR (SO:0000204) and 3' UTR (SO:0000205), also designated as J-segment.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>55</td>\n",
       "      <td>rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>scRNA</td>\n",
       "      <td>49</td>\n",
       "      <td>A small non coding RNA sequence, present in the cytoplasm.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>D_gene_segment</td>\n",
       "      <td>42</td>\n",
       "      <td>Germline genomic DNA including D-region with 5' UTR and 3' UTR, also designated as D-segment.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>C_gene_segment</td>\n",
       "      <td>29</td>\n",
       "      <td>Genomic DNA of immunoglobulin/T-cell receptor gene including C-region (and introns if present) with 5' UTR (SO:0000204) and 3' UTR (SO:0000205).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chromosome</td>\n",
       "      <td>25</td>\n",
       "      <td>Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>tRNA</td>\n",
       "      <td>22</td>\n",
       "      <td>Transfer RNA (tRNA) molecules are approximately 80 nucleotides in length. Their secondary structure includes four short double-helical elements and three loops (D, anti-codon, and T loops). Further hydrogen bonds mediate the characteristic L-shaped molecular structure. Transfer RNAs have two regions of fundamental functional importance: the anti-codon, which is responsible for specific mRNA codon recognition, and the 3' end, to which the tRNA's corresponding amino acid is attached (by aminoa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>12</td>\n",
       "      <td>A transcript for which no open reading frame has been identified and for which no other function has been determined.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Feature    count  \\\n",
       "0                     exon  3683354   \n",
       "1                      CDS  2284414   \n",
       "2           five_prime_UTR   426629   \n",
       "3          three_prime_UTR   340592   \n",
       "4                     mRNA   233603   \n",
       "5                  lnc_RNA   225629   \n",
       "6        biological_region   182818   \n",
       "7               ncRNA_gene    42124   \n",
       "8               transcript    28808   \n",
       "9                     gene    21571   \n",
       "10  pseudogenic_transcript    15207   \n",
       "11              pseudogene    15204   \n",
       "12                   snRNA     1915   \n",
       "13                   miRNA     1879   \n",
       "14  unconfirmed_transcript     1108   \n",
       "15                  snoRNA      942   \n",
       "16          V_gene_segment      253   \n",
       "17                scaffold      169   \n",
       "18          J_gene_segment       97   \n",
       "19                    rRNA       55   \n",
       "20                   scRNA       49   \n",
       "21          D_gene_segment       42   \n",
       "22          C_gene_segment       29   \n",
       "23              chromosome       25   \n",
       "24                    tRNA       22   \n",
       "25    processed_transcript       12   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             definition  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                               A region of the transcript sequence within a gene which is not removed from the primary RNA transcript by RNA splicing.  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                       A contiguous sequence which begins with, and includes, a start codon and ends with, and includes, a stop codon.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                 A region at the 5' end of a mature transcript (preceding the initiation codon) that is not translated into a protein.  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                       A region at the 3' end of a mature transcript (following the stop codon) that is not translated into a protein.  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                Messenger RNA is the intermediate molecule between DNA and protein. It includes UTR and coding sequences. It does not contain introns.  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   NaN  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                           A region defined by its disposition to be involved in a biological process.  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 A gene that encodes a non-coding RNA.  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                                     An RNA synthesized on a DNA or RNA template by an RNA polymerase.  \n",
       "9                                                                                                                                                                                                                                                                                                    A region (or regions) that includes all of the sequence elements necessary to encode a functional transcript. A gene may include regulatory regions, transcribed regions and/or other functional sequence regions.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                   A non functional descendant of a transcript, part of a pseudogene.  \n",
       "11  A sequence that closely resembles a known functional gene, at another locus within a genome, that is non-functional as a consequence of (usually several) mutations that prevent either its transcription or translation (or both). In general, pseudogenes result from either reverse transcription of a transcript of their \"normal\" paralog (SO:0000043) (in which case the pseudogene typically lacks introns and includes a poly(A) tail) or from recombination (SO:0000044) (in which case the pseudogene ...  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                           A small nuclear RNA molecule involved in pre-mRNA splicing and processing.  \n",
       "13                     Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.  \n",
       "14                                                                                                            This is used for non-spliced EST clusters that have polyA features. This category has been specifically created for the ENCODE project to highlight regions that could indicate the presence of protein coding genes that require experimental validation, either by 5' RACE or RT-PCR to extend the transcripts, or by confirming expression of the putatively-encoded peptide with specific antibodies.  \n",
       "15                                                                                                      Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.  \n",
       "16                                                                                                                                                                                                                                                                                                                                                                                                                             Germline genomic DNA including L-part1, V-intron and V-exon, with the 5' UTR and 3' UTR.  \n",
       "17                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "18                                                                                                                                                                                                                                                                                                                                                    Germline genomic DNA of an immunoglobulin/T-cell receptor gene including J-region with 5' UTR (SO:0000204) and 3' UTR (SO:0000205), also designated as J-segment.  \n",
       "19                                                                                                                                                                                                                                                                                                                                                                                                          rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.  \n",
       "20                                                                                                                                                                                                                                                                                                                                                                                                                                                           A small non coding RNA sequence, present in the cytoplasm.  \n",
       "21                                                                                                                                                                                                                                                                                                                                                                                                                        Germline genomic DNA including D-region with 5' UTR and 3' UTR, also designated as D-segment.  \n",
       "22                                                                                                                                                                                                                                                                                                                                                                     Genomic DNA of immunoglobulin/T-cell receptor gene including C-region (and introns if present) with 5' UTR (SO:0000204) and 3' UTR (SO:0000205).  \n",
       "23                                                                                                                                                                                                                                                                                                                                           Structural unit composed of a nucleic acid molecule which controls its own replication through the interaction of specific proteins at one or more origins of replication.  \n",
       "24  Transfer RNA (tRNA) molecules are approximately 80 nucleotides in length. Their secondary structure includes four short double-helical elements and three loops (D, anti-codon, and T loops). Further hydrogen bonds mediate the characteristic L-shaped molecular structure. Transfer RNAs have two regions of fundamental functional importance: the anti-codon, which is responsible for specific mRNA codon recognition, and the 3' end, to which the tRNA's corresponding amino acid is attached (by aminoa...  \n",
       "25                                                                                                                                                                                                                                                                                                                                                                                                A transcript for which no open reading frame has been identified and for which no other function has been determined.  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GFF features stats\n",
    "gff_features = gff_pr.Feature.value_counts().to_frame().reset_index()\n",
    "gff_features['tmp_id'] = gff_features['Feature'].str.capitalize().replace('_', '')\n",
    "gff_features = gff_features.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_features = gff_features.drop(['lbl', 'tmp_id'], axis=1)\n",
    "pd.DataFrame(gff_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a9e7f9eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "      <th>definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>231596</td>\n",
       "      <td>A gene which, when transcribed, can be translated into a protein.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>226412</td>\n",
       "      <td>A non-coding RNA generally longer than 200 nucleotides that cannot be classified as any other ncRNA subtype. Similar to mRNAs, lncRNAs are mainly transcribed by RNA polymerase II, are often capped by 7-methyl guanosine at their 5' ends, polyadenylated at their 3' ends and may be spliced.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>retained_intron</td>\n",
       "      <td>34239</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding_CDS_not_defined</td>\n",
       "      <td>26573</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>21949</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>18979</td>\n",
       "      <td>A pseudogene created via retrotranposition of the mRNA of a functional protein-coding parent gene followed by accumulation of deleterious mutations lacking introns and promoters, often including a polyA tail.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>4432</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>3898</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>snRNA</td>\n",
       "      <td>3820</td>\n",
       "      <td>A small nuclear RNA molecule involved in pre-mRNA splicing and processing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>3758</td>\n",
       "      <td>Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>3176</td>\n",
       "      <td>A unprocessed pseudogene supported by locus-specific evidence of transcription.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2298</td>\n",
       "      <td>A processed_pseudogene overlapped by locus-specific evidence of transcription.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TEC</td>\n",
       "      <td>2127</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>snoRNA</td>\n",
       "      <td>1884</td>\n",
       "      <td>Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>994</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>transcribed_unitary_pseudogene</td>\n",
       "      <td>402</td>\n",
       "      <td>A species specific unprocessed pseudogene without a parent gene, as it has an active orthologue in another species.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IG_V_pseudogene</td>\n",
       "      <td>374</td>\n",
       "      <td>A pseudogenic variable region which closely resembles a known functional imunoglobulin variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>292</td>\n",
       "      <td>A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of an Immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TR_V_gene</td>\n",
       "      <td>214</td>\n",
       "      <td>A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of aT-cell receptor chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>unitary_pseudogene</td>\n",
       "      <td>178</td>\n",
       "      <td>A pseudogene, deactivated from original state by mutation, fixed in a population,where the ortholog in a reference species such as mouse remains functional.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TR_J_gene</td>\n",
       "      <td>158</td>\n",
       "      <td>A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of aT-cell receptor chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rRNA</td>\n",
       "      <td>106</td>\n",
       "      <td>rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>non_stop_decay</td>\n",
       "      <td>105</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>scaRNA</td>\n",
       "      <td>98</td>\n",
       "      <td>A ncRNA, specific to the Cajal body, that has been demonstrated to function as a guide RNA in the site-specific synthesis of 2'-O-ribose-methylated nucleotides and pseudouridines in the RNA polymerase II-transcribed U1, U2, U4 and U5 spliceosomal small nuclear RNAs (snRNAs).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>protein_coding_LoF</td>\n",
       "      <td>74</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>IG_D_gene</td>\n",
       "      <td>74</td>\n",
       "      <td>A gene that rearranges at the DNA level and codes the diversity region of the variable domain of an immunoglobuin (IG) gene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TR_V_pseudogene</td>\n",
       "      <td>66</td>\n",
       "      <td>A pseudogenic variable region which closely resembles a known functional T receptor variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>44</td>\n",
       "      <td>Mitochondrial transfer RNA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>artifact</td>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>IG_C_gene</td>\n",
       "      <td>37</td>\n",
       "      <td>A constant (C) gene, a gene that codes the constant region of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>IG_J_gene</td>\n",
       "      <td>36</td>\n",
       "      <td>A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "      <td>18</td>\n",
       "      <td>A pseudogenic constant region of an immunoglobulin gene which closely resembles a known functional Imunoglobulin constant gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ribozyme</td>\n",
       "      <td>16</td>\n",
       "      <td>An RNA with catalytic activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>12</td>\n",
       "      <td>A transcript for which no open reading frame has been identified and for which no other function has been determined.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>TR_C_gene</td>\n",
       "      <td>12</td>\n",
       "      <td>A constant (C) gene, a gene that codes the constant region of a T-cell receptor chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>TR_D_gene</td>\n",
       "      <td>10</td>\n",
       "      <td>A gene that rearranges at the DNA level and codes the diversity region of the variable domain of aT-cell receptor gene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>sRNA</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>TR_J_pseudogene</td>\n",
       "      <td>8</td>\n",
       "      <td>A pseudogenic joining region which closely resembles a known functional T receptor (TR) joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>vault_RNA</td>\n",
       "      <td>8</td>\n",
       "      <td>A family of RNAs are found as part of the enigmatic vault ribonucleoprotein complex. The complex consists of a major vault protein (MVP), two minor vault proteins (VPARP and TEP1), and several small untranslated RNA molecules. It has been suggested that the vault complex is involved in drug resistance.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>pseudogene</td>\n",
       "      <td>8</td>\n",
       "      <td>A sequence that closely resembles a known functional gene, at another locus within a genome, that is non-functional as a consequence of (usually several) mutations that prevent either its transcription or translation (or both). In general, pseudogenes result from either reverse transcription of a transcript of their \"normal\" paralog (SO:0000043) (in which case the pseudogene typically lacks introns and includes a poly(A) tail) or from recombination (SO:0000044) (in which case the pseudogene ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>IG_J_pseudogene</td>\n",
       "      <td>6</td>\n",
       "      <td>A pseudogenic joining region which closely resembles a known functional imunoglobulin joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>translated_processed_pseudogene</td>\n",
       "      <td>4</td>\n",
       "      <td>A processed pseudogene where there is evidence, (mass spec data) suggesting that it is also translated.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>Mitochondrial rRNA is an RNA component of the small or large subunits of mitochondrial ribosomes.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>IG_pseudogene</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               biotype   count  \\\n",
       "0                       protein_coding  231596   \n",
       "1                               lncRNA  226412   \n",
       "2                      retained_intron   34239   \n",
       "3       protein_coding_CDS_not_defined   26573   \n",
       "4              nonsense_mediated_decay   21949   \n",
       "5                 processed_pseudogene   18979   \n",
       "6                             misc_RNA    4432   \n",
       "7               unprocessed_pseudogene    3898   \n",
       "8                                snRNA    3820   \n",
       "9                                miRNA    3758   \n",
       "10  transcribed_unprocessed_pseudogene    3176   \n",
       "11    transcribed_processed_pseudogene    2298   \n",
       "12                                 TEC    2127   \n",
       "13                              snoRNA    1884   \n",
       "14                     rRNA_pseudogene     994   \n",
       "15      transcribed_unitary_pseudogene     402   \n",
       "16                     IG_V_pseudogene     374   \n",
       "17                           IG_V_gene     292   \n",
       "18                           TR_V_gene     214   \n",
       "19                  unitary_pseudogene     178   \n",
       "20                           TR_J_gene     158   \n",
       "21                                rRNA     106   \n",
       "22                      non_stop_decay     105   \n",
       "23                              scaRNA      98   \n",
       "24                  protein_coding_LoF      74   \n",
       "25                           IG_D_gene      74   \n",
       "26                     TR_V_pseudogene      66   \n",
       "27                             Mt_tRNA      44   \n",
       "28                            artifact      38   \n",
       "29                           IG_C_gene      37   \n",
       "30                           IG_J_gene      36   \n",
       "31                     IG_C_pseudogene      18   \n",
       "32                            ribozyme      16   \n",
       "33                processed_transcript      12   \n",
       "34                           TR_C_gene      12   \n",
       "35                           TR_D_gene      10   \n",
       "36                                sRNA      10   \n",
       "37                     TR_J_pseudogene       8   \n",
       "38                           vault_RNA       8   \n",
       "39                          pseudogene       8   \n",
       "40                     IG_J_pseudogene       6   \n",
       "41     translated_processed_pseudogene       4   \n",
       "42                             Mt_rRNA       4   \n",
       "43                       IG_pseudogene       2   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             definition  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                                                                                     A gene which, when transcribed, can be translated into a protein.  \n",
       "1                                                                                                                                                                                                                      A non-coding RNA generally longer than 200 nucleotides that cannot be classified as any other ncRNA subtype. Similar to mRNAs, lncRNAs are mainly transcribed by RNA polymerase II, are often capped by 7-methyl guanosine at their 5' ends, polyadenylated at their 3' ends and may be spliced.  \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   NaN  \n",
       "3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   NaN  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   NaN  \n",
       "5                                                                                                                                                                                                                                                                                                      A pseudogene created via retrotranposition of the mRNA of a functional protein-coding parent gene followed by accumulation of deleterious mutations lacking introns and promoters, often including a polyA tail.  \n",
       "6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   NaN  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   NaN  \n",
       "8                                                                                                                                                                                                                                                                                                                                                                                                                                            A small nuclear RNA molecule involved in pre-mRNA splicing and processing.  \n",
       "9                      Small, ~22-nt, RNA molecule that is the endogenous transcript of a miRNA gene (or the product of other non coding RNA genes). Micro RNAs are produced from precursor molecules (SO:0001244) that can form local hairpin structures, which ordinarily are processed (usually via the Dicer pathway) such that a single miRNA molecule accumulates from one arm of a hairpin precursor molecule. Micro RNAs may trigger the cleavage of their target molecules or act as translational repressors.  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                      A unprocessed pseudogene supported by locus-specific evidence of transcription.  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                       A processed_pseudogene overlapped by locus-specific evidence of transcription.  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "13                                                                                                      Small nucleolar RNAs (snoRNAs) are short non-coding RNAs enriched in the nucleolus as components of small nucleolar ribonucleoproteins. They guide ribose methylation and pseudouridylation of rRNAs and snRNAs, and a subgroup regulate excision of rRNAs from rRNA precursor transcripts. snoRNAs may also guide rRNA acetylation and tRNA methylation, and regulate mRNA abundance and alternative splicing.  \n",
       "14                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "15                                                                                                                                                                                                                                                                                                                                                                                                  A species specific unprocessed pseudogene without a parent gene, as it has an active orthologue in another species.  \n",
       "16                                                                                                                                                                                                 A pseudogenic variable region which closely resembles a known functional imunoglobulin variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.  \n",
       "17                                                                                                                                                                                                                                                                                                                                                                                    A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of an Immunoglobulin chain.  \n",
       "18                                                                                                                                                                                                                                                                                                                                                                                     A variable gene that rearranges at the DNA level and codes the variable region of the variable domain of aT-cell receptor chain.  \n",
       "19                                                                                                                                                                                                                                                                                                                                                         A pseudogene, deactivated from original state by mutation, fixed in a population,where the ortholog in a reference species such as mouse remains functional.  \n",
       "20                                                                                                                                                                                                                                                                                                                                                                                       A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of aT-cell receptor chain.  \n",
       "21                                                                                                                                                                                                                                                                                                                                                                                                          rRNA is an RNA component of a ribosome that can provide both structural scaffolding and catalytic activity.  \n",
       "22                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "23                                                                                                                                                                                                                                  A ncRNA, specific to the Cajal body, that has been demonstrated to function as a guide RNA in the site-specific synthesis of 2'-O-ribose-methylated nucleotides and pseudouridines in the RNA polymerase II-transcribed U1, U2, U4 and U5 spliceosomal small nuclear RNAs (snRNAs).  \n",
       "24                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "25                                                                                                                                                                                                                                                                                                                                                                                         A gene that rearranges at the DNA level and codes the diversity region of the variable domain of an immunoglobuin (IG) gene.  \n",
       "26                                                                                                                                                                                                    A pseudogenic variable region which closely resembles a known functional T receptor variable gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the variable region of an immunoglobulin chain.  \n",
       "27                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Mitochondrial transfer RNA.  \n",
       "28                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "29                                                                                                                                                                                                                                                                                                                                                                                                                               A constant (C) gene, a gene that codes the constant region of an immunoglobulin chain.  \n",
       "30                                                                                                                                                                                                                                                                                                                                                                                      A joining gene that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.  \n",
       "31                                                                                                                                                                                                                                                                 A pseudogenic constant region of an immunoglobulin gene which closely resembles a known functional Imunoglobulin constant gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon.  \n",
       "32                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      An RNA with catalytic activity.  \n",
       "33                                                                                                                                                                                                                                                                                                                                                                                                A transcript for which no open reading frame has been identified and for which no other function has been determined.  \n",
       "34                                                                                                                                                                                                                                                                                                                                                                                                                               A constant (C) gene, a gene that codes the constant region of a T-cell receptor chain.  \n",
       "35                                                                                                                                                                                                                                                                                                                                                                                              A gene that rearranges at the DNA level and codes the diversity region of the variable domain of aT-cell receptor gene.  \n",
       "36                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  \n",
       "37                                                                                                                                                                           A pseudogenic joining region which closely resembles a known functional T receptor (TR) joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.  \n",
       "38                                                                                                                                                                                                      A family of RNAs are found as part of the enigmatic vault ribonucleoprotein complex. The complex consists of a major vault protein (MVP), two minor vault proteins (VPARP and TEP1), and several small untranslated RNA molecules. It has been suggested that the vault complex is involved in drug resistance.  \n",
       "39  A sequence that closely resembles a known functional gene, at another locus within a genome, that is non-functional as a consequence of (usually several) mutations that prevent either its transcription or translation (or both). In general, pseudogenes result from either reverse transcription of a transcript of their \"normal\" paralog (SO:0000043) (in which case the pseudogene typically lacks introns and includes a poly(A) tail) or from recombination (SO:0000044) (in which case the pseudogene ...  \n",
       "40                                                                                                                                                                             A pseudogenic joining region which closely resembles a known functional imunoglobulin joining gene but in which the coding region has stop codons, frameshift mutations or a mutation that effects the initiation codon that rearranges at the DNA level and codes the joining region of the variable domain of an immunoglobulin chain.  \n",
       "41                                                                                                                                                                                                                                                                                                                                                                                                              A processed pseudogene where there is evidence, (mass spec data) suggesting that it is also translated.  \n",
       "42                                                                                                                                                                                                                                                                                                                                                                                                                    Mitochondrial rRNA is an RNA component of the small or large subunits of mitochondrial ribosomes.  \n",
       "43                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GFF biotypes stats\n",
    "gff_biotypes = gff_pr.biotype.value_counts().to_frame().reset_index()\n",
    "gff_biotypes['tmp_id'] = gff_biotypes['biotype'].str.capitalize().replace('_', '')\n",
    "gff_biotypes = gff_biotypes.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_biotypes = gff_biotypes.drop(['lbl', 'tmp_id'], axis=1)\n",
    "pd.DataFrame(gff_biotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8cfc34fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>svtype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DEL</td>\n",
       "      <td>1197080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BND</td>\n",
       "      <td>356035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DUP</td>\n",
       "      <td>269326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INS:ME:ALU</td>\n",
       "      <td>173374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>INS</td>\n",
       "      <td>83441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>INS:ME:LINE1</td>\n",
       "      <td>30223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>INS:ME:SVA</td>\n",
       "      <td>17607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CPX</td>\n",
       "      <td>15189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DEL:ME:LINE1</td>\n",
       "      <td>8505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>INV</td>\n",
       "      <td>2193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CNV</td>\n",
       "      <td>721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DEL:ME:HERVK</td>\n",
       "      <td>693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CTX</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          svtype    count\n",
       "0            DEL  1197080\n",
       "1            BND   356035\n",
       "2            DUP   269326\n",
       "3     INS:ME:ALU   173374\n",
       "4            INS    83441\n",
       "5   INS:ME:LINE1    30223\n",
       "6     INS:ME:SVA    17607\n",
       "7            CPX    15189\n",
       "8   DEL:ME:LINE1     8505\n",
       "9            INV     2193\n",
       "10           CNV      721\n",
       "11  DEL:ME:HERVK      693\n",
       "12           CTX       99"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_pr.svtype.value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3a2175ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_location\n",
       "CDS     2502793\n",
       "3UTR    2056551\n",
       "5UTR        175\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tarbase_pr['gene_location'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "011dad84",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Prepare GFF3 data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bb4ed9e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add additional metrics\n",
    "gff_pr['feature_len'] = (gff_pr['End'] - gff_pr['Start']).astype('int')\n",
    "gff_pr['feature_id'] = 'feature_' + gff_pr['Chromosome'].astype('str') + ':' + gff_pr['Start'].astype('str') + '-' + gff_pr['End'].astype('str')\n",
    "\n",
    "# Self annotation of GFF with mRNA (parent) data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level', 'feature_len', 'feature_id'] # Replace the data in these empty columns with mRNA data\n",
    "annotation_gff_pr = gff_pr[target_columns].dropna(subset='ID')\n",
    "\n",
    "annotation_suffix = '_annotation'\n",
    "annotated_gff_pr = gff_pr.merge(annotation_gff_pr, how='left', left_on='Parent', right_on='ID', suffixes=['', annotation_suffix])\n",
    "\n",
    "annotated_columns = [f'{column}{annotation_suffix}' for column in target_columns]\n",
    "fill_dict = dict(zip(target_columns, annotated_columns))\n",
    "for target, annotation in fill_dict.items():\n",
    "    annotated_gff_pr[target] = annotated_gff_pr[target].fillna(annotated_gff_pr[annotation])\n",
    "annotated_gff_pr = annotated_gff_pr.drop(annotated_columns, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "79f6e1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create specific PRs (mRNAm, 3'-UTR, offtargets)\n",
    "mrna_pr = annotated_gff_pr.query('Feature == \"mRNA\"')\n",
    "three_utrs_pr = annotated_gff_pr.query('Feature == \"three_prime_UTR\"')\n",
    "\n",
    "offtarget_features = ['five_prime_UTR',\n",
    "                               'CDS',\n",
    "                               #'lnc_RNA',\n",
    "                               'ncRNA_gene',\n",
    "                               #'snRNA',\n",
    "                               #'snoRNA',\n",
    "                               #'scRNA',\n",
    "                               'rRNA',\n",
    "                               'tRNA',\n",
    "                               'processed_transcript',\n",
    "                               ]\n",
    "offtarget_features_pr = gff_pr.query('Feature in @offtarget_features')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d1834b24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>190278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>21947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding_LoF</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IG_C_gene</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TR_C_gene</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   biotype   count\n",
       "0           protein_coding  190278\n",
       "1  nonsense_mediated_decay   21947\n",
       "2       protein_coding_LoF      43\n",
       "3                IG_C_gene      22\n",
       "4                TR_C_gene       6\n",
       "5                IG_V_gene       2"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3'-UTR biotype stats\n",
    "# Считаем именно уникальные вхождения, так как один 3'-UTR может быть разбить на несколько интервалов в GFF (разделен интронами).\n",
    "# В особенности это выражено у nonsense_mediated_decay транскриптов (нам они не пригодятся, но тем не менее).\n",
    "three_utrs_pr.drop_duplicates(subset='ID')['biotype'].value_counts().to_frame().reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3b10ec4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          biotype  count\n",
       "0  protein_coding  19757"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Keep only MANE_select / MANE_Select|Ensembl_canonical\n",
    "three_utrs_pr_filtered = three_utrs_pr[three_utrs_pr['tag'].str.contains('MANE_Select', na=False, regex=True)]\n",
    "\n",
    "# Stats\n",
    "# MANE_select genes count = 19437 (release_1.5) from summary file: https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.5/\n",
    "three_utrs_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20ce9d9",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Prepare SV data</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2a71e19",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">gnomAD</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d8c1ed5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/2433410726.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/2433410726.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')\n"
     ]
    }
   ],
   "source": [
    "# Keep target SVs (DEL)\n",
    "sv_targets = ['DEL']\n",
    "sv_filter = ['PASS']\n",
    "target_sv_pr = sv_pr.query('svtype in @sv_targets and FILTER in @sv_filter')\n",
    "\n",
    "# Add columns\n",
    "target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
    "target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "970eda5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>627321.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean</td>\n",
       "      <td>0.0045937636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>std</td>\n",
       "      <td>0.0415649697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>min</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10%</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25%</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>50%</td>\n",
       "      <td>0.0000240000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>75%</td>\n",
       "      <td>0.0001110000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>90%</td>\n",
       "      <td>0.0009820000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>95%</td>\n",
       "      <td>0.0033650000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>99%</td>\n",
       "      <td>0.1228398010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>99.9%</td>\n",
       "      <td>0.6803870177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>max</td>\n",
       "      <td>1.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index                AF\n",
       "0   count 627321.0000000000\n",
       "1    mean      0.0045937636\n",
       "2     std      0.0415649697\n",
       "3     min      0.0000080000\n",
       "4     10%      0.0000080000\n",
       "5     25%      0.0000080000\n",
       "6     50%      0.0000240000\n",
       "7     75%      0.0001110000\n",
       "8     90%      0.0009820000\n",
       "9     95%      0.0033650000\n",
       "10    99%      0.1228398010\n",
       "11  99.9%      0.6803870177\n",
       "12    max      1.0000000000"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SV percentiles by AF\n",
    "target_sv_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "66228d5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sv_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>627321.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean</td>\n",
       "      <td>8529.8884988706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>std</td>\n",
       "      <td>139720.7127188754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>min</td>\n",
       "      <td>51.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10%</td>\n",
       "      <td>76.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25%</td>\n",
       "      <td>137.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>50%</td>\n",
       "      <td>702.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>75%</td>\n",
       "      <td>4073.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>90%</td>\n",
       "      <td>10309.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>95%</td>\n",
       "      <td>23359.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>99%</td>\n",
       "      <td>142071.0000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>99.9%</td>\n",
       "      <td>568683.9200000279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>max</td>\n",
       "      <td>75976919.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index              sv_len\n",
       "0   count   627321.0000000000\n",
       "1    mean     8529.8884988706\n",
       "2     std   139720.7127188754\n",
       "3     min       51.0000000000\n",
       "4     10%       76.0000000000\n",
       "5     25%      137.0000000000\n",
       "6     50%      702.0000000000\n",
       "7     75%     4073.0000000000\n",
       "8     90%    10309.0000000000\n",
       "9     95%    23359.0000000000\n",
       "10    99%   142071.0000000005\n",
       "11  99.9%   568683.9200000279\n",
       "12    max 75976919.0000000000"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SV percentiles by SV length\n",
    "target_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "79e87fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"fig = px.histogram( \\n    x=target_sv_pr['sv_len'],\\n    nbins=300,\\n    log_y=True,\\n    title='gnomAD SV DEL lenght distribution (log)'\\n)\\nfig.show()\""
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''fig = px.histogram( \n",
    "    x=target_sv_pr['sv_len'],\n",
    "    nbins=300,\n",
    "    log_y=True,\n",
    "    title='gnomAD SV DEL lenght distribution (log)'\n",
    ")\n",
    "fig.show()'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0daf451b",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">RCMG BTK SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "615a2160",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_90049/1031382406.py:36: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_filepath), compression='gzip')\n"
     ]
    }
   ],
   "source": [
    "rcmg_btk_sv_filepath = rcmg_btk_sv_dir / 'rcmg_btk_sv.csv.gz'\n",
    "\n",
    "if not rcmg_btk_sv_filepath.exists():\n",
    "    rcmg_bkt_sv_df = pd.DataFrame() # concated filtered aggregated\n",
    "    for rcmg_btk_sv_file in rcmg_btk_sv_filepaths:\n",
    "        print(rcmg_btk_sv_file)\n",
    "        current_rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_file), compression='gzip')\n",
    "\n",
    "        # Add columns\n",
    "        current_rcmg_bkt_sv_df['AF'] = current_rcmg_bkt_sv_df['AD'] / current_rcmg_bkt_sv_df['DP']\n",
    "        current_rcmg_bkt_sv_df['SVLEN'] = abs(current_rcmg_bkt_sv_df['END'] - current_rcmg_bkt_sv_df['POS'])\n",
    "\n",
    "        # Drop columns\n",
    "        current_rcmg_bkt_sv_df = current_rcmg_bkt_sv_df.drop(columns=['REF', 'ALT'])\n",
    "\n",
    "        # Filter\n",
    "        current_rcmg_bkt_sv_df_filtered = current_rcmg_bkt_sv_df.query('caller == \"manta\"')\n",
    "\n",
    "        # Aggregate\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered.groupby(['Chromosome', 'POS', 'END', 'SVTYPE', 'caller', 'SVLEN'], as_index=False).agg({\n",
    "            'SID': [list, 'count'],\n",
    "            'DP': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AD': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AF': [list, 'min', 'median', 'mean', 'max']\n",
    "        })\n",
    "\n",
    "        # Rename columns\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated.columns = ['_'.join(col).strip('_') for col in current_rcmg_bkt_sv_df_filtered_aggregated.columns]\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.rename(columns={'POS': 'Start', 'END': 'End'})\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.reset_index(drop=True)\n",
    "\n",
    "        # Concat\n",
    "        rcmg_bkt_sv_df = pd.concat([rcmg_bkt_sv_df, current_rcmg_bkt_sv_df_filtered_aggregated], ignore_index=False)\n",
    "    rcmg_bkt_sv_df.to_csv(str(rcmg_btk_sv_filepath), index=False, compression='gzip')\n",
    "else:\n",
    "    rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_filepath), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d296b61f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>caller</th>\n",
       "      <th>SVLEN</th>\n",
       "      <th>SID_list</th>\n",
       "      <th>SID_count</th>\n",
       "      <th>DP_list</th>\n",
       "      <th>DP_min</th>\n",
       "      <th>DP_median</th>\n",
       "      <th>DP_mean</th>\n",
       "      <th>DP_max</th>\n",
       "      <th>AD_list</th>\n",
       "      <th>AD_min</th>\n",
       "      <th>AD_median</th>\n",
       "      <th>AD_mean</th>\n",
       "      <th>AD_max</th>\n",
       "      <th>AF_list</th>\n",
       "      <th>AF_min</th>\n",
       "      <th>AF_median</th>\n",
       "      <th>AF_mean</th>\n",
       "      <th>AF_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3028568</th>\n",
       "      <td>16</td>\n",
       "      <td>89587758</td>\n",
       "      <td>89587887</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>129</td>\n",
       "      <td>['000007121080', '000007090090', '000007146070', '000007092940']</td>\n",
       "      <td>4</td>\n",
       "      <td>[15, 10, 7, 21]</td>\n",
       "      <td>7</td>\n",
       "      <td>12.5000000000</td>\n",
       "      <td>13.2500000000</td>\n",
       "      <td>21</td>\n",
       "      <td>[10, 7, 7, 15]</td>\n",
       "      <td>7</td>\n",
       "      <td>8.5000000000</td>\n",
       "      <td>9.7500000000</td>\n",
       "      <td>15</td>\n",
       "      <td>[0.6666666666666666, 0.7, 1.0, 0.7142857142857143]</td>\n",
       "      <td>0.6666666667</td>\n",
       "      <td>0.7071428571</td>\n",
       "      <td>0.7702380952</td>\n",
       "      <td>1.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3757127</th>\n",
       "      <td>X</td>\n",
       "      <td>151763379</td>\n",
       "      <td>151763512</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>133</td>\n",
       "      <td>['000027027560', '000027027610', '000007425420', '000027023500']</td>\n",
       "      <td>4</td>\n",
       "      <td>[4, 3, 3, 8]</td>\n",
       "      <td>3</td>\n",
       "      <td>3.5000000000</td>\n",
       "      <td>4.5000000000</td>\n",
       "      <td>8</td>\n",
       "      <td>[4, 2, 2, 4]</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0000000000</td>\n",
       "      <td>3.0000000000</td>\n",
       "      <td>4</td>\n",
       "      <td>[1.0, 0.6666666666666666, 0.6666666666666666, 0.5]</td>\n",
       "      <td>0.5000000000</td>\n",
       "      <td>0.6666666667</td>\n",
       "      <td>0.7083333333</td>\n",
       "      <td>1.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2688245</th>\n",
       "      <td>13</td>\n",
       "      <td>17086759</td>\n",
       "      <td>17146613</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>59854</td>\n",
       "      <td>['000007070820', '000007405730', '000029004370', '000029001870']</td>\n",
       "      <td>4</td>\n",
       "      <td>[10, 10, 19, 13]</td>\n",
       "      <td>10</td>\n",
       "      <td>11.5000000000</td>\n",
       "      <td>13.0000000000</td>\n",
       "      <td>19</td>\n",
       "      <td>[5, 7, 10, 10]</td>\n",
       "      <td>5</td>\n",
       "      <td>8.5000000000</td>\n",
       "      <td>8.0000000000</td>\n",
       "      <td>10</td>\n",
       "      <td>[0.5, 0.7, 0.5263157894736842, 0.7692307692307693]</td>\n",
       "      <td>0.5000000000</td>\n",
       "      <td>0.6131578947</td>\n",
       "      <td>0.6238866397</td>\n",
       "      <td>0.7692307692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2688247</th>\n",
       "      <td>13</td>\n",
       "      <td>17086759</td>\n",
       "      <td>17146615</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>59856</td>\n",
       "      <td>['000007073710', '000007177110', '000007184240', '000007416750']</td>\n",
       "      <td>4</td>\n",
       "      <td>[23, 17, 14, 16]</td>\n",
       "      <td>14</td>\n",
       "      <td>16.5000000000</td>\n",
       "      <td>17.5000000000</td>\n",
       "      <td>23</td>\n",
       "      <td>[13, 11, 8, 12]</td>\n",
       "      <td>8</td>\n",
       "      <td>11.5000000000</td>\n",
       "      <td>11.0000000000</td>\n",
       "      <td>13</td>\n",
       "      <td>[0.5652173913043478, 0.6470588235294118, 0.5714285714285714, 0.75]</td>\n",
       "      <td>0.5652173913</td>\n",
       "      <td>0.6092436975</td>\n",
       "      <td>0.6334261966</td>\n",
       "      <td>0.7500000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1070691</th>\n",
       "      <td>5</td>\n",
       "      <td>46659146</td>\n",
       "      <td>46708885</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>49739</td>\n",
       "      <td>['000007075930', '000007147100', '000007168650', '000007186100']</td>\n",
       "      <td>4</td>\n",
       "      <td>[30, 33, 24, 32]</td>\n",
       "      <td>24</td>\n",
       "      <td>31.0000000000</td>\n",
       "      <td>29.7500000000</td>\n",
       "      <td>33</td>\n",
       "      <td>[28, 31, 17, 28]</td>\n",
       "      <td>17</td>\n",
       "      <td>28.0000000000</td>\n",
       "      <td>26.0000000000</td>\n",
       "      <td>31</td>\n",
       "      <td>[0.9333333333333333, 0.9393939393939394, 0.7083333333333334, 0.875]</td>\n",
       "      <td>0.7083333333</td>\n",
       "      <td>0.9041666667</td>\n",
       "      <td>0.8640151515</td>\n",
       "      <td>0.9393939394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441101</th>\n",
       "      <td>6</td>\n",
       "      <td>160507372</td>\n",
       "      <td>160659872</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>152500</td>\n",
       "      <td>['000007151860']</td>\n",
       "      <td>1</td>\n",
       "      <td>[63]</td>\n",
       "      <td>63</td>\n",
       "      <td>63.0000000000</td>\n",
       "      <td>63.0000000000</td>\n",
       "      <td>63</td>\n",
       "      <td>[20]</td>\n",
       "      <td>20</td>\n",
       "      <td>20.0000000000</td>\n",
       "      <td>20.0000000000</td>\n",
       "      <td>20</td>\n",
       "      <td>[0.31746031746031744]</td>\n",
       "      <td>0.3174603175</td>\n",
       "      <td>0.3174603175</td>\n",
       "      <td>0.3174603175</td>\n",
       "      <td>0.3174603175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441104</th>\n",
       "      <td>6</td>\n",
       "      <td>160507372</td>\n",
       "      <td>160659882</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>152510</td>\n",
       "      <td>['000007139390']</td>\n",
       "      <td>1</td>\n",
       "      <td>[53]</td>\n",
       "      <td>53</td>\n",
       "      <td>53.0000000000</td>\n",
       "      <td>53.0000000000</td>\n",
       "      <td>53</td>\n",
       "      <td>[17]</td>\n",
       "      <td>17</td>\n",
       "      <td>17.0000000000</td>\n",
       "      <td>17.0000000000</td>\n",
       "      <td>17</td>\n",
       "      <td>[0.32075471698113206]</td>\n",
       "      <td>0.3207547170</td>\n",
       "      <td>0.3207547170</td>\n",
       "      <td>0.3207547170</td>\n",
       "      <td>0.3207547170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441106</th>\n",
       "      <td>6</td>\n",
       "      <td>160507372</td>\n",
       "      <td>160659895</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>152523</td>\n",
       "      <td>['000007408380']</td>\n",
       "      <td>1</td>\n",
       "      <td>[54]</td>\n",
       "      <td>54</td>\n",
       "      <td>54.0000000000</td>\n",
       "      <td>54.0000000000</td>\n",
       "      <td>54</td>\n",
       "      <td>[11]</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0000000000</td>\n",
       "      <td>11.0000000000</td>\n",
       "      <td>11</td>\n",
       "      <td>[0.2037037037037037]</td>\n",
       "      <td>0.2037037037</td>\n",
       "      <td>0.2037037037</td>\n",
       "      <td>0.2037037037</td>\n",
       "      <td>0.2037037037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1441107</th>\n",
       "      <td>6</td>\n",
       "      <td>160507373</td>\n",
       "      <td>160659825</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>152452</td>\n",
       "      <td>['000007181600']</td>\n",
       "      <td>1</td>\n",
       "      <td>[57]</td>\n",
       "      <td>57</td>\n",
       "      <td>57.0000000000</td>\n",
       "      <td>57.0000000000</td>\n",
       "      <td>57</td>\n",
       "      <td>[27]</td>\n",
       "      <td>27</td>\n",
       "      <td>27.0000000000</td>\n",
       "      <td>27.0000000000</td>\n",
       "      <td>27</td>\n",
       "      <td>[0.47368421052631576]</td>\n",
       "      <td>0.4736842105</td>\n",
       "      <td>0.4736842105</td>\n",
       "      <td>0.4736842105</td>\n",
       "      <td>0.4736842105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3810646</th>\n",
       "      <td>M</td>\n",
       "      <td>7951</td>\n",
       "      <td>15904</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>7953</td>\n",
       "      <td>[7121740]</td>\n",
       "      <td>1</td>\n",
       "      <td>[13520]</td>\n",
       "      <td>13520</td>\n",
       "      <td>13520.0000000000</td>\n",
       "      <td>13520.0000000000</td>\n",
       "      <td>13520</td>\n",
       "      <td>[3777]</td>\n",
       "      <td>3777</td>\n",
       "      <td>3777.0000000000</td>\n",
       "      <td>3777.0000000000</td>\n",
       "      <td>3777</td>\n",
       "      <td>[0.2793639053254438]</td>\n",
       "      <td>0.2793639053</td>\n",
       "      <td>0.2793639053</td>\n",
       "      <td>0.2793639053</td>\n",
       "      <td>0.2793639053</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3592218 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome      Start        End SVTYPE caller   SVLEN  \\\n",
       "3028568         16   89587758   89587887    DEL  manta     129   \n",
       "3757127          X  151763379  151763512    DEL  manta     133   \n",
       "2688245         13   17086759   17146613    DEL  manta   59854   \n",
       "2688247         13   17086759   17146615    DEL  manta   59856   \n",
       "1070691          5   46659146   46708885    DEL  manta   49739   \n",
       "...            ...        ...        ...    ...    ...     ...   \n",
       "1441101          6  160507372  160659872    DEL  manta  152500   \n",
       "1441104          6  160507372  160659882    DEL  manta  152510   \n",
       "1441106          6  160507372  160659895    DEL  manta  152523   \n",
       "1441107          6  160507373  160659825    DEL  manta  152452   \n",
       "3810646          M       7951      15904    DEL  manta    7953   \n",
       "\n",
       "                                                                 SID_list  \\\n",
       "3028568  ['000007121080', '000007090090', '000007146070', '000007092940']   \n",
       "3757127  ['000027027560', '000027027610', '000007425420', '000027023500']   \n",
       "2688245  ['000007070820', '000007405730', '000029004370', '000029001870']   \n",
       "2688247  ['000007073710', '000007177110', '000007184240', '000007416750']   \n",
       "1070691  ['000007075930', '000007147100', '000007168650', '000007186100']   \n",
       "...                                                                   ...   \n",
       "1441101                                                  ['000007151860']   \n",
       "1441104                                                  ['000007139390']   \n",
       "1441106                                                  ['000007408380']   \n",
       "1441107                                                  ['000007181600']   \n",
       "3810646                                                         [7121740]   \n",
       "\n",
       "         SID_count           DP_list  DP_min        DP_median  \\\n",
       "3028568          4   [15, 10, 7, 21]       7    12.5000000000   \n",
       "3757127          4      [4, 3, 3, 8]       3     3.5000000000   \n",
       "2688245          4  [10, 10, 19, 13]      10    11.5000000000   \n",
       "2688247          4  [23, 17, 14, 16]      14    16.5000000000   \n",
       "1070691          4  [30, 33, 24, 32]      24    31.0000000000   \n",
       "...            ...               ...     ...              ...   \n",
       "1441101          1              [63]      63    63.0000000000   \n",
       "1441104          1              [53]      53    53.0000000000   \n",
       "1441106          1              [54]      54    54.0000000000   \n",
       "1441107          1              [57]      57    57.0000000000   \n",
       "3810646          1           [13520]   13520 13520.0000000000   \n",
       "\n",
       "                 DP_mean  DP_max           AD_list  AD_min       AD_median  \\\n",
       "3028568    13.2500000000      21    [10, 7, 7, 15]       7    8.5000000000   \n",
       "3757127     4.5000000000       8      [4, 2, 2, 4]       2    3.0000000000   \n",
       "2688245    13.0000000000      19    [5, 7, 10, 10]       5    8.5000000000   \n",
       "2688247    17.5000000000      23   [13, 11, 8, 12]       8   11.5000000000   \n",
       "1070691    29.7500000000      33  [28, 31, 17, 28]      17   28.0000000000   \n",
       "...                  ...     ...               ...     ...             ...   \n",
       "1441101    63.0000000000      63              [20]      20   20.0000000000   \n",
       "1441104    53.0000000000      53              [17]      17   17.0000000000   \n",
       "1441106    54.0000000000      54              [11]      11   11.0000000000   \n",
       "1441107    57.0000000000      57              [27]      27   27.0000000000   \n",
       "3810646 13520.0000000000   13520            [3777]    3777 3777.0000000000   \n",
       "\n",
       "                AD_mean  AD_max  \\\n",
       "3028568    9.7500000000      15   \n",
       "3757127    3.0000000000       4   \n",
       "2688245    8.0000000000      10   \n",
       "2688247   11.0000000000      13   \n",
       "1070691   26.0000000000      31   \n",
       "...                 ...     ...   \n",
       "1441101   20.0000000000      20   \n",
       "1441104   17.0000000000      17   \n",
       "1441106   11.0000000000      11   \n",
       "1441107   27.0000000000      27   \n",
       "3810646 3777.0000000000    3777   \n",
       "\n",
       "                                                                     AF_list  \\\n",
       "3028568                   [0.6666666666666666, 0.7, 1.0, 0.7142857142857143]   \n",
       "3757127                   [1.0, 0.6666666666666666, 0.6666666666666666, 0.5]   \n",
       "2688245                   [0.5, 0.7, 0.5263157894736842, 0.7692307692307693]   \n",
       "2688247   [0.5652173913043478, 0.6470588235294118, 0.5714285714285714, 0.75]   \n",
       "1070691  [0.9333333333333333, 0.9393939393939394, 0.7083333333333334, 0.875]   \n",
       "...                                                                      ...   \n",
       "1441101                                                [0.31746031746031744]   \n",
       "1441104                                                [0.32075471698113206]   \n",
       "1441106                                                 [0.2037037037037037]   \n",
       "1441107                                                [0.47368421052631576]   \n",
       "3810646                                                 [0.2793639053254438]   \n",
       "\n",
       "              AF_min    AF_median      AF_mean       AF_max  \n",
       "3028568 0.6666666667 0.7071428571 0.7702380952 1.0000000000  \n",
       "3757127 0.5000000000 0.6666666667 0.7083333333 1.0000000000  \n",
       "2688245 0.5000000000 0.6131578947 0.6238866397 0.7692307692  \n",
       "2688247 0.5652173913 0.6092436975 0.6334261966 0.7500000000  \n",
       "1070691 0.7083333333 0.9041666667 0.8640151515 0.9393939394  \n",
       "...              ...          ...          ...          ...  \n",
       "1441101 0.3174603175 0.3174603175 0.3174603175 0.3174603175  \n",
       "1441104 0.3207547170 0.3207547170 0.3207547170 0.3207547170  \n",
       "1441106 0.2037037037 0.2037037037 0.2037037037 0.2037037037  \n",
       "1441107 0.4736842105 0.4736842105 0.4736842105 0.4736842105  \n",
       "3810646 0.2793639053 0.2793639053 0.2793639053 0.2793639053  \n",
       "\n",
       "[3592218 rows x 23 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rcmg_bkt_sv_df.query('SID_count < 5').sort_values(by='SID_count', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e0ca86",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Join GFF3 x gnomAD SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "45b14ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join GFF x SV\n",
    "features_sv_joined_pr = three_utrs_pr_filtered.join_overlaps(target_sv_pr, suffix='_sv', report_overlap_column='join_len')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a6885c5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add additional columns (join_start, join_end, join_id)\n",
    "features_sv_joined_pr['join_start'] = np.maximum(features_sv_joined_pr['Start'], features_sv_joined_pr['Start_sv'])\n",
    "features_sv_joined_pr['join_end'] = np.minimum(features_sv_joined_pr['End'], features_sv_joined_pr['End_sv'])\n",
    "features_sv_joined_pr['join_id'] = 'join_' + features_sv_joined_pr['Chromosome'].astype('str') + ':' + features_sv_joined_pr['join_start'].astype('str') + '-' + features_sv_joined_pr['join_end'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "33ec54b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>join_type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>feature_in_sv</td>\n",
       "      <td>17788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>3610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>2129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sv_left_free</td>\n",
       "      <td>1975</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       join_type  count\n",
       "0  feature_in_sv  17788\n",
       "1  sv_in_feature   3610\n",
       "2  sv_right_free   2129\n",
       "3   sv_left_free   1975"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define joint type\n",
    "features_sv_joined_pr['join_type'] = features_sv_joined_pr.apply(define_join_type, axis=1)\n",
    "features_sv_joined_pr['join_type'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "33ae52f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count joints by SV\n",
    "joints_count_by_sv = features_sv_joined_pr['sv_id'].value_counts().to_frame().reset_index()\n",
    "joints_count_by_sv = joints_count_by_sv.rename(columns={'count': 'joins_count'})\n",
    "\n",
    "# Add joints count data\n",
    "features_sv_joined_pr = features_sv_joined_pr.merge(joints_count_by_sv, how='left')\n",
    "#len(features_sv_joined_pr)\n",
    "\n",
    "# Filter joins by join type\n",
    "targey_joint_types = ['sv_in_feature', 'sv_right_free', 'sv_left_free']\n",
    "features_sv_joined_pr_filtered = features_sv_joined_pr.query('join_type in @targey_joint_types and ((Strand == \"+\" and join_type == \"sv_right_free\") or (Strand == \"-\" and join_type == \"sv_left_free\") or join_type == \"sv_in_feature\") and joins_count == 1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b449576",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join type stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4a69d46c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>join_type</th>\n",
       "      <th>sv_in_feature</th>\n",
       "      <th>sv_left_free</th>\n",
       "      <th>sv_right_free</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3521.0000000000</td>\n",
       "      <td>962.0000000000</td>\n",
       "      <td>1124.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>508.7157057654</td>\n",
       "      <td>7150.2079002079</td>\n",
       "      <td>7269.1227758007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>928.4987400891</td>\n",
       "      <td>18421.3678622167</td>\n",
       "      <td>19884.5691740781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>51.0000000000</td>\n",
       "      <td>52.0000000000</td>\n",
       "      <td>53.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10%</th>\n",
       "      <td>61.0000000000</td>\n",
       "      <td>656.6000000000</td>\n",
       "      <td>628.3000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>83.0000000000</td>\n",
       "      <td>1627.2500000000</td>\n",
       "      <td>1584.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>167.0000000000</td>\n",
       "      <td>4131.5000000000</td>\n",
       "      <td>3872.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>570.0000000000</td>\n",
       "      <td>6911.5000000000</td>\n",
       "      <td>7001.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90%</th>\n",
       "      <td>1226.0000000000</td>\n",
       "      <td>11954.3000000000</td>\n",
       "      <td>12049.1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95%</th>\n",
       "      <td>1965.0000000000</td>\n",
       "      <td>22855.2500000000</td>\n",
       "      <td>20279.4000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99%</th>\n",
       "      <td>4783.2000000000</td>\n",
       "      <td>68047.7300000000</td>\n",
       "      <td>78191.7400000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99.9%</th>\n",
       "      <td>9268.8000000000</td>\n",
       "      <td>275754.6970000100</td>\n",
       "      <td>267968.8130000200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>14078.0000000000</td>\n",
       "      <td>371108.0000000000</td>\n",
       "      <td>455766.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "join_type    sv_in_feature      sv_left_free     sv_right_free\n",
       "count      3521.0000000000    962.0000000000   1124.0000000000\n",
       "mean        508.7157057654   7150.2079002079   7269.1227758007\n",
       "std         928.4987400891  18421.3678622167  19884.5691740781\n",
       "min          51.0000000000     52.0000000000     53.0000000000\n",
       "10%          61.0000000000    656.6000000000    628.3000000000\n",
       "25%          83.0000000000   1627.2500000000   1584.0000000000\n",
       "50%         167.0000000000   4131.5000000000   3872.0000000000\n",
       "75%         570.0000000000   6911.5000000000   7001.0000000000\n",
       "90%        1226.0000000000  11954.3000000000  12049.1000000000\n",
       "95%        1965.0000000000  22855.2500000000  20279.4000000000\n",
       "99%        4783.2000000000  68047.7300000000  78191.7400000000\n",
       "99.9%      9268.8000000000 275754.6970000100 267968.8130000200\n",
       "max       14078.0000000000 371108.0000000000 455766.0000000000"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stats by join type\n",
    "features_sv_joined_pr_filtered.groupby('join_type')['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cc037975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          biotype  count\n",
       "0  protein_coding   5607"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_joined_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4135ee",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Offtarget features exclusion</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ce637a",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">features_sv_joined_pr_filtered copy</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ab7f6e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Offtarget stats\n",
    "features_sv_joined_pr_filtered_cp = features_sv_joined_pr_filtered.copy()\n",
    "features_sv_joined_pr_filtered_cp[['Start_tmp', 'End_tmp']] = features_sv_joined_pr_filtered_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_joined_pr_filtered_cp[['Start', 'End']] = features_sv_joined_pr_filtered_cp[['Start_sv', 'End_sv']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157fc717",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Filter offtargets</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "1dc11ee7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4154"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Exclude offtargets (SV x offtarget overlaps)\n",
    "features_sv_no_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.overlap(offtarget_features_pr, invert=True, strand_behavior='ignore')\n",
    "\n",
    "features_sv_no_offtereget_joined_pr[['Start', 'End']] = features_sv_no_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_joined_pr = features_sv_no_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "len(features_sv_no_offtereget_joined_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "623083fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before offtargets filtering\n",
      "       join_type  count\n",
      "0  sv_in_feature   3521\n",
      "1  sv_right_free   1124\n",
      "2   sv_left_free    962\n",
      "\n",
      "After offtargets filtering\n",
      "       join_type  count\n",
      "0  sv_in_feature   2834\n",
      "1  sv_right_free    702\n",
      "2   sv_left_free    618\n"
     ]
    }
   ],
   "source": [
    "# Joint type stats after offtargets exclution\n",
    "print('Before offtargets filtering')\n",
    "print(features_sv_joined_pr_filtered['join_type'].value_counts().to_frame().reset_index())\n",
    "print('\\nAfter offtargets filtering')\n",
    "print(features_sv_no_offtereget_joined_pr['join_type'].value_counts().to_frame().reset_index())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ddbe504",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Offtarget stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2bd478da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# features_sv_offtereget_joined_pr - только для отслеживания пересечений с offtarget. Фильтрованный от offtarget pr - features_sv_no_offtereget_joined_pr (через overlap, потому что join находя пересечения с offtarget (через left) также оставляет оригинальную строку).DS_Storefeatures_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len')\n",
    "features_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len', strand_behavior='ignore')\n",
    "\n",
    "features_sv_offtereget_joined_pr[['Start', 'End']] = features_sv_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_offtereget_joined_pr = features_sv_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_offtereget_joined_pr['offtarget_id'] = 'offtarget_' + features_sv_offtereget_joined_pr['Chromosome'].astype('str') + ':' + features_sv_offtereget_joined_pr['Start_offtarget'].astype('str') + '-' + features_sv_offtereget_joined_pr['End_offtarget'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c516d79f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>join_type</th>\n",
       "      <th>Feature_offtarget</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">sv_in_feature</th>\n",
       "      <th>CDS</th>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>processed_transcript</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">sv_left_free</th>\n",
       "      <th>CDS</th>\n",
       "      <td>1458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>1015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">sv_right_free</th>\n",
       "      <th>CDS</th>\n",
       "      <td>1903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>416</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    count\n",
       "join_type     Feature_offtarget          \n",
       "sv_in_feature CDS                     164\n",
       "              five_prime_UTR           19\n",
       "              ncRNA_gene              656\n",
       "              processed_transcript      4\n",
       "sv_left_free  CDS                    1458\n",
       "              five_prime_UTR         1015\n",
       "              ncRNA_gene              320\n",
       "sv_right_free CDS                    1903\n",
       "              five_prime_UTR          813\n",
       "              ncRNA_gene              416"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features offtargets stats\n",
    "# Суммарные значения данной статистики будут выше чем разница всех джоинов - оффтартет джоинов, так как один и тот же SV (напр. DEL) могут пересекать разные изоформы одного и того же транскрипта.\n",
    "features_sv_offtereget_joined_pr.groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])\n",
    "#features_sv_offtereget_joined_pr.drop_duplicates(subset='sv_id').groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d8bf654",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">miRNA join</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1e4a88",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">features_sv_no_offtereget_joined_pr copy</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e88e7aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Offtarget stats\n",
    "features_sv_no_offtereget_joined_pr_cp = features_sv_no_offtereget_joined_pr.copy()\n",
    "features_sv_no_offtereget_joined_pr_cp[['Start_tmp', 'End_tmp']] = features_sv_no_offtereget_joined_pr_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_no_offtereget_joined_pr_cp[['Start', 'End']] = features_sv_no_offtereget_joined_pr_cp[['join_start', 'join_end']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99af8a69",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1c871563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join TarBase\n",
    "features_sv_no_offtereget_tarbase_joined_pr = features_sv_no_offtereget_joined_pr_cp.join_overlaps(tarbase_pr, suffix='_tarbase', strand_behavior='ignore', report_overlap_column='tarbase_join_len')\n",
    "\n",
    "features_sv_no_offtereget_tarbase_joined_pr[['Start', 'End']] = features_sv_no_offtereget_tarbase_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_tarbase_joined_pr = features_sv_no_offtereget_tarbase_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_no_offtereget_tarbase_joined_pr['tarbase_id'] = 'tarbase_' + features_sv_no_offtereget_tarbase_joined_pr['Chromosome'].astype('str') + ':' + features_sv_no_offtereget_tarbase_joined_pr['Start_tarbase'].astype('str') + '-' + features_sv_no_offtereget_tarbase_joined_pr['End_tarbase'].astype('str') + ':' + features_sv_no_offtereget_tarbase_joined_pr['mirna_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6984057b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>90375.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean</td>\n",
       "      <td>0.0002522849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>std</td>\n",
       "      <td>0.0032934593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>min</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10%</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25%</td>\n",
       "      <td>0.0000080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>50%</td>\n",
       "      <td>0.0000160000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>75%</td>\n",
       "      <td>0.0000560000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>90%</td>\n",
       "      <td>0.0002540000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>95%</td>\n",
       "      <td>0.0008310000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>99%</td>\n",
       "      <td>0.0025530001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>99.9%</td>\n",
       "      <td>0.0397749990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>max</td>\n",
       "      <td>0.1723610014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index               AF\n",
       "0   count 90375.0000000000\n",
       "1    mean     0.0002522849\n",
       "2     std     0.0032934593\n",
       "3     min     0.0000080000\n",
       "4     10%     0.0000080000\n",
       "5     25%     0.0000080000\n",
       "6     50%     0.0000160000\n",
       "7     75%     0.0000560000\n",
       "8     90%     0.0002540000\n",
       "9     95%     0.0008310000\n",
       "10    99%     0.0025530001\n",
       "11  99.9%     0.0397749990\n",
       "12    max     0.1723610014"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "653cb4b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_location\n",
       "3UTR    90373\n",
       "CDS         2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['gene_location'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ae28975b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "interaction_group\n",
       "primary      67005\n",
       "secondary    23370\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['interaction_group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e230a118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "regulation\n",
       "Negative    90375\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['regulation'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "084ddca5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YOD1-202</td>\n",
       "      <td>982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SDE2-201</td>\n",
       "      <td>976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZBTB10-203</td>\n",
       "      <td>953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PTEN-220</td>\n",
       "      <td>911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KATNAL1-201</td>\n",
       "      <td>773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>ZBTB10-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>MS4A1-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1538</th>\n",
       "      <td>MMP16-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539</th>\n",
       "      <td>CYTH2-203</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1540</th>\n",
       "      <td>SPRY3-204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1541 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_name  count\n",
       "0           YOD1-202    982\n",
       "1           SDE2-201    976\n",
       "2         ZBTB10-203    953\n",
       "3           PTEN-220    911\n",
       "4        KATNAL1-201    773\n",
       "...              ...    ...\n",
       "1536      ZBTB10-201      1\n",
       "1537       MS4A1-201      1\n",
       "1538       MMP16-201      1\n",
       "1539       CYTH2-203      1\n",
       "1540       SPRY3-204      1\n",
       "\n",
       "[1541 rows x 2 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['transcript_name'].value_counts().to_frame().reset_index().to_csv(output_dir / 'mirna_genes.csv', index=False)\n",
    "features_sv_no_offtereget_tarbase_joined_pr['transcript_name'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9eaea44e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "experimental_method\n",
       "HITS-CLIP                    56089\n",
       "PAR-CLIP                     33412\n",
       "Chimeric fragments             390\n",
       "qCLASH                         257\n",
       "Luciferase Reporter Assay      191\n",
       "CLASH                           28\n",
       "IMPACT-Seq                       5\n",
       "3LIFE                            3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['experimental_method'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "53825b57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "join_type\n",
       "sv_in_feature    40369\n",
       "sv_right_free    29495\n",
       "sv_left_free     20511\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['join_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e67fb804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "join_id\n",
       "join_10:87966367-87971930     1098\n",
       "join_1:207043848-207048293     971\n",
       "join_8:80519631-80525632       785\n",
       "join_3:86067499-86074429       750\n",
       "join_13:30202629-30207106      725\n",
       "                              ... \n",
       "join_14:57469299-57470464        1\n",
       "join_6:56056873-56056976         1\n",
       "join_6:43615332-43616937         1\n",
       "join_6:41599219-41599319         1\n",
       "join_X:155777999-155782459       1\n",
       "Name: count, Length: 2290, dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['join_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f9903c45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>Alias</th>\n",
       "      <th>external_name</th>\n",
       "      <th>logic_name</th>\n",
       "      <th>Name</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>version</th>\n",
       "      <th>Parent</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>constitutive</th>\n",
       "      <th>ensembl_end_phase</th>\n",
       "      <th>ensembl_phase</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>Start_sv</th>\n",
       "      <th>End_sv</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>samples</th>\n",
       "      <th>MULTIALLELIC</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>CHR2</th>\n",
       "      <th>CPX_INTERVALS</th>\n",
       "      <th>CPX_TYPE</th>\n",
       "      <th>END</th>\n",
       "      <th>END2</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>LOW_CONFIDENCE_REPETITIVE_LARGE_DUP</th>\n",
       "      <th>MEMBERS</th>\n",
       "      <th>NCR</th>\n",
       "      <th>OUTLIER_SAMPLE_ENRICHED_LENIENT</th>\n",
       "      <th>PAR</th>\n",
       "      <th>PCRMINUS_NCR</th>\n",
       "      <th>PCRPLUS_NCR</th>\n",
       "      <th>PESR_GT_OVERDISPERSION</th>\n",
       "      <th>POS2</th>\n",
       "      <th>PREDICTED_BREAKEND_EXONIC</th>\n",
       "      <th>PREDICTED_COPY_GAIN</th>\n",
       "      <th>PREDICTED_DUP_PARTIAL</th>\n",
       "      <th>PREDICTED_INTERGENIC</th>\n",
       "      <th>PREDICTED_INTRAGENIC_EXON_DUP</th>\n",
       "      <th>PREDICTED_INTRONIC</th>\n",
       "      <th>PREDICTED_INV_SPAN</th>\n",
       "      <th>PREDICTED_LOF</th>\n",
       "      <th>PREDICTED_MSV_EXON_OVERLAP</th>\n",
       "      <th>PREDICTED_NEAREST_TSS</th>\n",
       "      <th>PREDICTED_NONCODING_BREAKPOINT</th>\n",
       "      <th>PREDICTED_NONCODING_SPAN</th>\n",
       "      <th>PREDICTED_PARTIAL_DISPERSED_DUP</th>\n",
       "      <th>PREDICTED_PARTIAL_EXON_DUP</th>\n",
       "      <th>PREDICTED_PROMOTER</th>\n",
       "      <th>PREDICTED_TSS_DUP</th>\n",
       "      <th>PREDICTED_UTR</th>\n",
       "      <th>RESOLVED_POSTHOC</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>SVLEN</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>UNRESOLVED_TYPE</th>\n",
       "      <th>AN</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>N_BI_GENOS</th>\n",
       "      <th>N_HOMREF</th>\n",
       "      <th>N_HET</th>\n",
       "      <th>N_HOMALT</th>\n",
       "      <th>FREQ_HOMREF</th>\n",
       "      <th>FREQ_HET</th>\n",
       "      <th>FREQ_HOMALT</th>\n",
       "      <th>CN_NUMBER</th>\n",
       "      <th>CN_COUNT</th>\n",
       "      <th>CN_STATUS</th>\n",
       "      <th>CN_FREQ</th>\n",
       "      <th>CN_NONREF_COUNT</th>\n",
       "      <th>CN_NONREF_FREQ</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>sv_len</th>\n",
       "      <th>sv_id</th>\n",
       "      <th>join_len</th>\n",
       "      <th>join_start</th>\n",
       "      <th>join_end</th>\n",
       "      <th>join_id</th>\n",
       "      <th>join_type</th>\n",
       "      <th>joins_count</th>\n",
       "      <th>species</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id_tarbase</th>\n",
       "      <th>gene_location</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_id_tarbase</th>\n",
       "      <th>Start_tarbase</th>\n",
       "      <th>End_tarbase</th>\n",
       "      <th>Strand_tarbase</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>regulation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>article_pubmed_id</th>\n",
       "      <th>confidence</th>\n",
       "      <th>interaction_group</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>microt_score</th>\n",
       "      <th>comment</th>\n",
       "      <th>tarbase_join_len</th>\n",
       "      <th>tarbase_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>MIMAT0000062</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7a-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7b-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7c-5p</td>\n",
       "      <td>MIMAT0000064</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7c-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7f-5p</td>\n",
       "      <td>MIMAT0000067</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7f-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7g-5p</td>\n",
       "      <td>MIMAT0000414</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7g-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4140</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055047</td>\n",
       "      <td>155055455</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_c97db132</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155055455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,RD,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0012172300</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0007600070</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95749.0000000000</td>\n",
       "      <td>19.0000000000</td>\n",
       "      <td>0.0001984350</td>\n",
       "      <td>62944.0000000000</td>\n",
       "      <td>62925.0000000000</td>\n",
       "      <td>19.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9996979833</td>\n",
       "      <td>0.0003018560</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>408</td>\n",
       "      <td>sv_X:155055047-155055455</td>\n",
       "      <td>408</td>\n",
       "      <td>155055047</td>\n",
       "      <td>155055455</td>\n",
       "      <td>join_X:155055047-155055455</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7i-5p</td>\n",
       "      <td>MIMAT0000415</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155055264</td>\n",
       "      <td>155055276</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Primary Motor Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30455455.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4800000000</td>\n",
       "      <td>Gestation week 15</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155055264-155055276:hsa-let-7i-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_b6884c3c</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155057049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024497800</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0014355700</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1338</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95654.0000000000</td>\n",
       "      <td>68.0000000000</td>\n",
       "      <td>0.0007108960</td>\n",
       "      <td>62900.0000000000</td>\n",
       "      <td>62833.0000000000</td>\n",
       "      <td>66.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.9989349842</td>\n",
       "      <td>0.0010492801</td>\n",
       "      <td>0.0000158983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1339</td>\n",
       "      <td>sv_X:155055710-155057049</td>\n",
       "      <td>1339</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>join_X:155055710-155057049</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-1275</td>\n",
       "      <td>MIMAT0005929</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155055701</td>\n",
       "      <td>155055713</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Pleura</td>\n",
       "      <td>BC-3</td>\n",
       "      <td>22100165.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>B lymphocytes</td>\n",
       "      <td>0.6800000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>tarbase_X:155055701-155055713:hsa-miR-1275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_b6884c3c</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155057049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024497800</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0014355700</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1338</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95654.0000000000</td>\n",
       "      <td>68.0000000000</td>\n",
       "      <td>0.0007108960</td>\n",
       "      <td>62900.0000000000</td>\n",
       "      <td>62833.0000000000</td>\n",
       "      <td>66.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.9989349842</td>\n",
       "      <td>0.0010492801</td>\n",
       "      <td>0.0000158983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1339</td>\n",
       "      <td>sv_X:155055710-155057049</td>\n",
       "      <td>1339</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>join_X:155055710-155057049</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-32-5p</td>\n",
       "      <td>MIMAT0000090</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155056053</td>\n",
       "      <td>155056065</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4200000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155056053-155056065:hsa-miR-32-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_b6884c3c</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155057049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024497800</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0014355700</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1338</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95654.0000000000</td>\n",
       "      <td>68.0000000000</td>\n",
       "      <td>0.0007108960</td>\n",
       "      <td>62900.0000000000</td>\n",
       "      <td>62833.0000000000</td>\n",
       "      <td>66.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.9989349842</td>\n",
       "      <td>0.0010492801</td>\n",
       "      <td>0.0000158983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1339</td>\n",
       "      <td>sv_X:155055710-155057049</td>\n",
       "      <td>1339</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>join_X:155055710-155057049</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-92b-3p</td>\n",
       "      <td>MIMAT0003218</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155056053</td>\n",
       "      <td>155056065</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5000000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155056053-155056065:hsa-miR-92b-3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4152</th>\n",
       "      <td>X</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155774738</td>\n",
       "      <td>155782459</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000695325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRY3-204</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000695325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7721</td>\n",
       "      <td>feature_X:155774738-155782459</td>\n",
       "      <td>155777999</td>\n",
       "      <td>155789000</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_857103bd</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>depth</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155789000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RD</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0060325800</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAMP7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11000</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125298.0000000000</td>\n",
       "      <td>2.0000000000</td>\n",
       "      <td>0.0000160000</td>\n",
       "      <td>62649.0000000000</td>\n",
       "      <td>62647.0000000000</td>\n",
       "      <td>2.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999679923</td>\n",
       "      <td>0.0000319239</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>11001</td>\n",
       "      <td>sv_X:155777999-155789000</td>\n",
       "      <td>4460</td>\n",
       "      <td>155777999</td>\n",
       "      <td>155782459</td>\n",
       "      <td>join_X:155777999-155782459</td>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-33a-5p</td>\n",
       "      <td>MIMAT0000091</td>\n",
       "      <td>SPRY3</td>\n",
       "      <td>ENSG00000168939</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>SPRY3-204</td>\n",
       "      <td>ENST00000695325</td>\n",
       "      <td>155781719</td>\n",
       "      <td>155781731</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4700000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155781719-155781731:hsa-miR-33a-5p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90375 rows × 123 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome          Source          Feature      Start        End Score  \\\n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "...         ...             ...              ...        ...        ...   ...   \n",
       "4140          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4141          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4141          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4141          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4152          X          havana  three_prime_UTR  155774738  155782459     .   \n",
       "\n",
       "     Strand Frame                          ID Alias external_name logic_name  \\\n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "...     ...   ...                         ...   ...           ...        ...   \n",
       "4140      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4141      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4141      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4141      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4152      +     .  transcript:ENST00000695325   NaN           NaN        NaN   \n",
       "\n",
       "             Name         biotype description gene_id version  \\\n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "...           ...             ...         ...     ...     ...   \n",
       "4140   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4141   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4141   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4141   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4152    SPRY3-204  protein_coding         NaN     NaN     NaN   \n",
       "\n",
       "                          Parent transcript_id constitutive ensembl_end_phase  \\\n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "...                          ...           ...          ...               ...   \n",
       "4140  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4141  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4141  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4141  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4152  transcript:ENST00000695325           NaN          NaN               NaN   \n",
       "\n",
       "     ensembl_phase exon_id rank  \\\n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "...            ...     ...  ...   \n",
       "4140           NaN     NaN  NaN   \n",
       "4141           NaN     NaN  NaN   \n",
       "4141           NaN     NaN  NaN   \n",
       "4141           NaN     NaN  NaN   \n",
       "4152           NaN     NaN  NaN   \n",
       "\n",
       "                                                              tag  \\\n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "...                                                           ...   \n",
       "4140  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4141  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4141  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4141  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4152  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "\n",
       "                transcript_support_level ccdsid protein_id  feature_len  \\\n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "...                                  ...    ...        ...          ...   \n",
       "4140  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4141  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4141  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4141  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4152                                 NaN    NaN        NaN         7721   \n",
       "\n",
       "                         feature_id   Start_sv     End_sv  \\\n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "...                             ...        ...        ...   \n",
       "4140  feature_X:155054672-155060304  155055047  155055455   \n",
       "4141  feature_X:155054672-155060304  155055710  155057049   \n",
       "4141  feature_X:155054672-155060304  155055710  155057049   \n",
       "4141  feature_X:155054672-155060304  155055710  155057049   \n",
       "4152  feature_X:155774738-155782459  155777999  155789000   \n",
       "\n",
       "                                name svtype  samples  MULTIALLELIC ALGORITHMS  \\\n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "...                              ...    ...      ...           ...        ...   \n",
       "4140  gnomAD-SV_v3_DEL_chrX_c97db132    DEL      NaN         False       wham   \n",
       "4141  gnomAD-SV_v3_DEL_chrX_b6884c3c    DEL      NaN         False      manta   \n",
       "4141  gnomAD-SV_v3_DEL_chrX_b6884c3c    DEL      NaN         False      manta   \n",
       "4141  gnomAD-SV_v3_DEL_chrX_b6884c3c    DEL      NaN         False      manta   \n",
       "4152  gnomAD-SV_v3_DEL_chrX_857103bd    DEL      NaN         False      depth   \n",
       "\n",
       "      BOTHSIDES_SUPPORT  CHR2 CPX_INTERVALS CPX_TYPE        END  END2  \\\n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "...                 ...   ...           ...      ...        ...   ...   \n",
       "4140              False  chrX           NaN      NaN  155055455   NaN   \n",
       "4141              False  chrX           NaN      NaN  155057049   NaN   \n",
       "4141              False  chrX           NaN      NaN  155057049   NaN   \n",
       "4141              False  chrX           NaN      NaN  155057049   NaN   \n",
       "4152              False  chrX           NaN      NaN  155789000   NaN   \n",
       "\n",
       "      EVIDENCE  LOW_CONFIDENCE_REPETITIVE_LARGE_DUP MEMBERS          NCR  \\\n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "...        ...                                  ...     ...          ...   \n",
       "4140  PE,RD,SR                                False     NaN 0.0012172300   \n",
       "4141     PE,SR                                False     NaN 0.0024497800   \n",
       "4141     PE,SR                                False     NaN 0.0024497800   \n",
       "4141     PE,SR                                False     NaN 0.0024497800   \n",
       "4152        RD                                False     NaN 0.0060325800   \n",
       "\n",
       "      OUTLIER_SAMPLE_ENRICHED_LENIENT    PAR  PCRMINUS_NCR  PCRPLUS_NCR  \\\n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "...                               ...    ...           ...          ...   \n",
       "4140                            False  False           NaN 0.0007600070   \n",
       "4141                            False  False           NaN 0.0014355700   \n",
       "4141                            False  False           NaN 0.0014355700   \n",
       "4141                            False  False           NaN 0.0014355700   \n",
       "4152                            False   True           NaN 0.0000000000   \n",
       "\n",
       "      PESR_GT_OVERDISPERSION  POS2 PREDICTED_BREAKEND_EXONIC  \\\n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "...                      ...   ...                       ...   \n",
       "4140                   False   NaN                       NaN   \n",
       "4141                   False   NaN                       NaN   \n",
       "4141                   False   NaN                       NaN   \n",
       "4141                   False   NaN                       NaN   \n",
       "4152                   False   NaN                       NaN   \n",
       "\n",
       "     PREDICTED_COPY_GAIN PREDICTED_DUP_PARTIAL  PREDICTED_INTERGENIC  \\\n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "...                  ...                   ...                   ...   \n",
       "4140                 NaN                   NaN                 False   \n",
       "4141                 NaN                   NaN                 False   \n",
       "4141                 NaN                   NaN                 False   \n",
       "4141                 NaN                   NaN                 False   \n",
       "4152                 NaN                   NaN                  True   \n",
       "\n",
       "     PREDICTED_INTRAGENIC_EXON_DUP PREDICTED_INTRONIC PREDICTED_INV_SPAN  \\\n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "...                            ...                ...                ...   \n",
       "4140                           NaN                NaN                NaN   \n",
       "4141                           NaN                NaN                NaN   \n",
       "4141                           NaN                NaN                NaN   \n",
       "4141                           NaN                NaN                NaN   \n",
       "4152                           NaN                NaN                NaN   \n",
       "\n",
       "     PREDICTED_LOF PREDICTED_MSV_EXON_OVERLAP PREDICTED_NEAREST_TSS  \\\n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "...            ...                        ...                   ...   \n",
       "4140           NaN                        NaN                   NaN   \n",
       "4141           NaN                        NaN                   NaN   \n",
       "4141           NaN                        NaN                   NaN   \n",
       "4141           NaN                        NaN                   NaN   \n",
       "4152           NaN                        NaN                 VAMP7   \n",
       "\n",
       "      PREDICTED_NONCODING_BREAKPOINT  PREDICTED_NONCODING_SPAN  \\\n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "...                              ...                       ...   \n",
       "4140                             NaN                       NaN   \n",
       "4141                             NaN                       NaN   \n",
       "4141                             NaN                       NaN   \n",
       "4141                             NaN                       NaN   \n",
       "4152                             NaN                       NaN   \n",
       "\n",
       "     PREDICTED_PARTIAL_DISPERSED_DUP PREDICTED_PARTIAL_EXON_DUP  \\\n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "...                              ...                        ...   \n",
       "4140                             NaN                        NaN   \n",
       "4141                             NaN                        NaN   \n",
       "4141                             NaN                        NaN   \n",
       "4141                             NaN                        NaN   \n",
       "4152                             NaN                        NaN   \n",
       "\n",
       "     PREDICTED_PROMOTER PREDICTED_TSS_DUP PREDICTED_UTR  RESOLVED_POSTHOC  \\\n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "...                 ...               ...           ...               ...   \n",
       "4140                NaN               NaN        FUNDC2             False   \n",
       "4141                NaN               NaN        FUNDC2             False   \n",
       "4141                NaN               NaN        FUNDC2             False   \n",
       "4141                NaN               NaN        FUNDC2             False   \n",
       "4152                NaN               NaN           NaN             False   \n",
       "\n",
       "     SOURCE  SVLEN SVTYPE UNRESOLVED_TYPE                AN            AC  \\\n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "...     ...    ...    ...             ...               ...           ...   \n",
       "4140    NaN    407    DEL             NaN  95749.0000000000 19.0000000000   \n",
       "4141    NaN   1338    DEL             NaN  95654.0000000000 68.0000000000   \n",
       "4141    NaN   1338    DEL             NaN  95654.0000000000 68.0000000000   \n",
       "4141    NaN   1338    DEL             NaN  95654.0000000000 68.0000000000   \n",
       "4152    NaN  11000    DEL             NaN 125298.0000000000  2.0000000000   \n",
       "\n",
       "               AF       N_BI_GENOS         N_HOMREF         N_HET  \\\n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "...           ...              ...              ...           ...   \n",
       "4140 0.0001984350 62944.0000000000 62925.0000000000 19.0000000000   \n",
       "4141 0.0007108960 62900.0000000000 62833.0000000000 66.0000000000   \n",
       "4141 0.0007108960 62900.0000000000 62833.0000000000 66.0000000000   \n",
       "4141 0.0007108960 62900.0000000000 62833.0000000000 66.0000000000   \n",
       "4152 0.0000160000 62649.0000000000 62647.0000000000  2.0000000000   \n",
       "\n",
       "         N_HOMALT  FREQ_HOMREF     FREQ_HET  FREQ_HOMALT  CN_NUMBER CN_COUNT  \\\n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "...           ...          ...          ...          ...        ...      ...   \n",
       "4140 0.0000000000 0.9996979833 0.0003018560 0.0000000000        NaN      NaN   \n",
       "4141 1.0000000000 0.9989349842 0.0010492801 0.0000158983        NaN      NaN   \n",
       "4141 1.0000000000 0.9989349842 0.0010492801 0.0000158983        NaN      NaN   \n",
       "4141 1.0000000000 0.9989349842 0.0010492801 0.0000158983        NaN      NaN   \n",
       "4152 0.0000000000 0.9999679923 0.0000319239 0.0000000000        NaN      NaN   \n",
       "\n",
       "     CN_STATUS CN_FREQ  CN_NONREF_COUNT  CN_NONREF_FREQ FILTER  sv_len  \\\n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "...        ...     ...              ...             ...    ...     ...   \n",
       "4140       NaN     NaN              NaN             NaN   PASS     408   \n",
       "4141       NaN     NaN              NaN             NaN   PASS    1339   \n",
       "4141       NaN     NaN              NaN             NaN   PASS    1339   \n",
       "4141       NaN     NaN              NaN             NaN   PASS    1339   \n",
       "4152       NaN     NaN              NaN             NaN   PASS   11001   \n",
       "\n",
       "                         sv_id  join_len  join_start   join_end  \\\n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "...                        ...       ...         ...        ...   \n",
       "4140  sv_X:155055047-155055455       408   155055047  155055455   \n",
       "4141  sv_X:155055710-155057049      1339   155055710  155057049   \n",
       "4141  sv_X:155055710-155057049      1339   155055710  155057049   \n",
       "4141  sv_X:155055710-155057049      1339   155055710  155057049   \n",
       "4152  sv_X:155777999-155789000      4460   155777999  155782459   \n",
       "\n",
       "                         join_id      join_type  joins_count       species  \\\n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "...                          ...            ...          ...           ...   \n",
       "4140  join_X:155055047-155055455  sv_in_feature            1  Homo sapiens   \n",
       "4141  join_X:155055710-155057049  sv_in_feature            1  Homo sapiens   \n",
       "4141  join_X:155055710-155057049  sv_in_feature            1  Homo sapiens   \n",
       "4141  join_X:155055710-155057049  sv_in_feature            1  Homo sapiens   \n",
       "4152  join_X:155777999-155782459  sv_right_free            1  Homo sapiens   \n",
       "\n",
       "          mirna_name      mirna_id gene_name  gene_id_tarbase gene_location  \\\n",
       "4      hsa-let-7a-5p  MIMAT0000062   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7b-5p  MIMAT0000063   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7c-5p  MIMAT0000064   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7f-5p  MIMAT0000067   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7g-5p  MIMAT0000414   TMEM88B  ENSG00000205116          3UTR   \n",
       "...              ...           ...       ...              ...           ...   \n",
       "4140   hsa-let-7i-5p  MIMAT0000415    FUNDC2  ENSG00000165775          3UTR   \n",
       "4141    hsa-miR-1275  MIMAT0005929    FUNDC2  ENSG00000165775          3UTR   \n",
       "4141   hsa-miR-32-5p  MIMAT0000090    FUNDC2  ENSG00000165775          3UTR   \n",
       "4141  hsa-miR-92b-3p  MIMAT0003218    FUNDC2  ENSG00000165775          3UTR   \n",
       "4152  hsa-miR-33a-5p  MIMAT0000091     SPRY3  ENSG00000168939          3UTR   \n",
       "\n",
       "     transcript_name transcript_id_tarbase  Start_tarbase  End_tarbase  \\\n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "...              ...                   ...            ...          ...   \n",
       "4140      FUNDC2-201       ENST00000369498      155055264    155055276   \n",
       "4141      FUNDC2-201       ENST00000369498      155055701    155055713   \n",
       "4141      FUNDC2-201       ENST00000369498      155056053    155056065   \n",
       "4141      FUNDC2-201       ENST00000369498      155056053    155056065   \n",
       "4152       SPRY3-204       ENST00000695325      155781719    155781731   \n",
       "\n",
       "     Strand_tarbase experimental_method regulation  \\\n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "...             ...                 ...        ...   \n",
       "4140              +           HITS-CLIP   Negative   \n",
       "4141              +            PAR-CLIP   Negative   \n",
       "4141              +           HITS-CLIP   Negative   \n",
       "4141              +           HITS-CLIP   Negative   \n",
       "4152              +           HITS-CLIP   Negative   \n",
       "\n",
       "                            tissue cell_line   article_pubmed_id  confidence  \\\n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "...                            ...       ...                 ...         ...   \n",
       "4140  Brain - Primary Motor Cortex       NaN 30455455.0000000000           1   \n",
       "4141                        Pleura      BC-3 22100165.0000000000           1   \n",
       "4141          Brain - Motor cortex       NaN 24389009.0000000000           1   \n",
       "4141          Brain - Motor cortex       NaN 24389009.0000000000           1   \n",
       "4152          Brain - Motor cortex       NaN 24389009.0000000000           1   \n",
       "\n",
       "     interaction_group      cell_type  microt_score            comment  \\\n",
       "4              primary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "...                ...            ...           ...                ...   \n",
       "4140         secondary            NaN  0.4800000000  Gestation week 15   \n",
       "4141           primary  B lymphocytes  0.6800000000                NaN   \n",
       "4141           primary            NaN  0.4200000000    Male - 44 years   \n",
       "4141         secondary            NaN  0.5000000000    Male - 44 years   \n",
       "4152           primary            NaN  0.4700000000    Male - 44 years   \n",
       "\n",
       "      tarbase_join_len                                    tarbase_id  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7a-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7b-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7c-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7f-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7g-5p  \n",
       "...                ...                                           ...  \n",
       "4140                12   tarbase_X:155055264-155055276:hsa-let-7i-5p  \n",
       "4141                 3    tarbase_X:155055701-155055713:hsa-miR-1275  \n",
       "4141                12   tarbase_X:155056053-155056065:hsa-miR-32-5p  \n",
       "4141                12  tarbase_X:155056053-155056065:hsa-miR-92b-3p  \n",
       "4152                12  tarbase_X:155781719-155781731:hsa-miR-33a-5p  \n",
       "\n",
       "[90375 rows x 123 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(features_sv_no_offtereget_tarbase_joined_pr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1dec50d",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">GFF output for IGV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "8ac3dc61",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr_for_igv = features_sv_no_offtereget_tarbase_joined_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e4946b2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>Alias</th>\n",
       "      <th>external_name</th>\n",
       "      <th>logic_name</th>\n",
       "      <th>Name</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>version</th>\n",
       "      <th>Parent</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>constitutive</th>\n",
       "      <th>ensembl_end_phase</th>\n",
       "      <th>ensembl_phase</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>Start_sv</th>\n",
       "      <th>End_sv</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>samples</th>\n",
       "      <th>MULTIALLELIC</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>CHR2</th>\n",
       "      <th>CPX_INTERVALS</th>\n",
       "      <th>CPX_TYPE</th>\n",
       "      <th>END</th>\n",
       "      <th>END2</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>LOW_CONFIDENCE_REPETITIVE_LARGE_DUP</th>\n",
       "      <th>MEMBERS</th>\n",
       "      <th>NCR</th>\n",
       "      <th>OUTLIER_SAMPLE_ENRICHED_LENIENT</th>\n",
       "      <th>PAR</th>\n",
       "      <th>PCRMINUS_NCR</th>\n",
       "      <th>PCRPLUS_NCR</th>\n",
       "      <th>PESR_GT_OVERDISPERSION</th>\n",
       "      <th>POS2</th>\n",
       "      <th>PREDICTED_BREAKEND_EXONIC</th>\n",
       "      <th>PREDICTED_COPY_GAIN</th>\n",
       "      <th>PREDICTED_DUP_PARTIAL</th>\n",
       "      <th>PREDICTED_INTERGENIC</th>\n",
       "      <th>PREDICTED_INTRAGENIC_EXON_DUP</th>\n",
       "      <th>PREDICTED_INTRONIC</th>\n",
       "      <th>PREDICTED_INV_SPAN</th>\n",
       "      <th>PREDICTED_LOF</th>\n",
       "      <th>PREDICTED_MSV_EXON_OVERLAP</th>\n",
       "      <th>PREDICTED_NEAREST_TSS</th>\n",
       "      <th>PREDICTED_NONCODING_BREAKPOINT</th>\n",
       "      <th>PREDICTED_NONCODING_SPAN</th>\n",
       "      <th>PREDICTED_PARTIAL_DISPERSED_DUP</th>\n",
       "      <th>PREDICTED_PARTIAL_EXON_DUP</th>\n",
       "      <th>PREDICTED_PROMOTER</th>\n",
       "      <th>PREDICTED_TSS_DUP</th>\n",
       "      <th>PREDICTED_UTR</th>\n",
       "      <th>RESOLVED_POSTHOC</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>SVLEN</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>UNRESOLVED_TYPE</th>\n",
       "      <th>AN</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>N_BI_GENOS</th>\n",
       "      <th>N_HOMREF</th>\n",
       "      <th>N_HET</th>\n",
       "      <th>N_HOMALT</th>\n",
       "      <th>FREQ_HOMREF</th>\n",
       "      <th>FREQ_HET</th>\n",
       "      <th>FREQ_HOMALT</th>\n",
       "      <th>CN_NUMBER</th>\n",
       "      <th>CN_COUNT</th>\n",
       "      <th>CN_STATUS</th>\n",
       "      <th>CN_FREQ</th>\n",
       "      <th>CN_NONREF_COUNT</th>\n",
       "      <th>CN_NONREF_FREQ</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>sv_len</th>\n",
       "      <th>sv_id</th>\n",
       "      <th>join_len</th>\n",
       "      <th>join_start</th>\n",
       "      <th>join_end</th>\n",
       "      <th>join_id</th>\n",
       "      <th>join_type</th>\n",
       "      <th>joins_count</th>\n",
       "      <th>species</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id_tarbase</th>\n",
       "      <th>gene_location</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_id_tarbase</th>\n",
       "      <th>Start_tarbase</th>\n",
       "      <th>End_tarbase</th>\n",
       "      <th>Strand_tarbase</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>regulation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>article_pubmed_id</th>\n",
       "      <th>confidence</th>\n",
       "      <th>interaction_group</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>microt_score</th>\n",
       "      <th>comment</th>\n",
       "      <th>tarbase_join_len</th>\n",
       "      <th>tarbase_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>MIMAT0000062</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7a-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7b-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7c-5p</td>\n",
       "      <td>MIMAT0000064</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7c-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7f-5p</td>\n",
       "      <td>MIMAT0000067</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7f-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1427787</td>\n",
       "      <td>1430255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000378821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2468</td>\n",
       "      <td>feature_1:1427787-1430255</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chr1_08f4439e</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1429179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0019674799</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0019242300</td>\n",
       "      <td>0.0024584001</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>141</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125770.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000080000</td>\n",
       "      <td>62885.0000000000</td>\n",
       "      <td>62884.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999840260</td>\n",
       "      <td>0.0000159020</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>142</td>\n",
       "      <td>sv_1:1429037-1429179</td>\n",
       "      <td>142</td>\n",
       "      <td>1429037</td>\n",
       "      <td>1429179</td>\n",
       "      <td>join_1:1429037-1429179</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7g-5p</td>\n",
       "      <td>MIMAT0000414</td>\n",
       "      <td>TMEM88B</td>\n",
       "      <td>ENSG00000205116</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>TMEM88B-201</td>\n",
       "      <td>ENST00000378821</td>\n",
       "      <td>1429042</td>\n",
       "      <td>1429054</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Cingulate gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1429042-1429054:hsa-let-7g-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4140</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055047</td>\n",
       "      <td>155055455</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_c97db132</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155055455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,RD,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0012172300</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0007600070</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>407</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95749.0000000000</td>\n",
       "      <td>19.0000000000</td>\n",
       "      <td>0.0001984350</td>\n",
       "      <td>62944.0000000000</td>\n",
       "      <td>62925.0000000000</td>\n",
       "      <td>19.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9996979833</td>\n",
       "      <td>0.0003018560</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>408</td>\n",
       "      <td>sv_X:155055047-155055455</td>\n",
       "      <td>408</td>\n",
       "      <td>155055047</td>\n",
       "      <td>155055455</td>\n",
       "      <td>join_X:155055047-155055455</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7i-5p</td>\n",
       "      <td>MIMAT0000415</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155055264</td>\n",
       "      <td>155055276</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Primary Motor Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30455455.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4800000000</td>\n",
       "      <td>Gestation week 15</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155055264-155055276:hsa-let-7i-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_b6884c3c</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155057049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024497800</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0014355700</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1338</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95654.0000000000</td>\n",
       "      <td>68.0000000000</td>\n",
       "      <td>0.0007108960</td>\n",
       "      <td>62900.0000000000</td>\n",
       "      <td>62833.0000000000</td>\n",
       "      <td>66.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.9989349842</td>\n",
       "      <td>0.0010492801</td>\n",
       "      <td>0.0000158983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1339</td>\n",
       "      <td>sv_X:155055710-155057049</td>\n",
       "      <td>1339</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>join_X:155055710-155057049</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-1275</td>\n",
       "      <td>MIMAT0005929</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155055701</td>\n",
       "      <td>155055713</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Pleura</td>\n",
       "      <td>BC-3</td>\n",
       "      <td>22100165.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>B lymphocytes</td>\n",
       "      <td>0.6800000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>tarbase_X:155055701-155055713:hsa-miR-1275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_b6884c3c</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155057049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024497800</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0014355700</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1338</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95654.0000000000</td>\n",
       "      <td>68.0000000000</td>\n",
       "      <td>0.0007108960</td>\n",
       "      <td>62900.0000000000</td>\n",
       "      <td>62833.0000000000</td>\n",
       "      <td>66.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.9989349842</td>\n",
       "      <td>0.0010492801</td>\n",
       "      <td>0.0000158983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1339</td>\n",
       "      <td>sv_X:155055710-155057049</td>\n",
       "      <td>1339</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>join_X:155055710-155057049</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-32-5p</td>\n",
       "      <td>MIMAT0000090</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155056053</td>\n",
       "      <td>155056065</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4200000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155056053-155056065:hsa-miR-32-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_b6884c3c</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155057049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0024497800</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0014355700</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1338</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95654.0000000000</td>\n",
       "      <td>68.0000000000</td>\n",
       "      <td>0.0007108960</td>\n",
       "      <td>62900.0000000000</td>\n",
       "      <td>62833.0000000000</td>\n",
       "      <td>66.0000000000</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>0.9989349842</td>\n",
       "      <td>0.0010492801</td>\n",
       "      <td>0.0000158983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1339</td>\n",
       "      <td>sv_X:155055710-155057049</td>\n",
       "      <td>1339</td>\n",
       "      <td>155055710</td>\n",
       "      <td>155057049</td>\n",
       "      <td>join_X:155055710-155057049</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-92b-3p</td>\n",
       "      <td>MIMAT0003218</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155056053</td>\n",
       "      <td>155056065</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5000000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155056053-155056065:hsa-miR-92b-3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4152</th>\n",
       "      <td>X</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155774738</td>\n",
       "      <td>155782459</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000695325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRY3-204</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000695325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7721</td>\n",
       "      <td>feature_X:155774738-155782459</td>\n",
       "      <td>155777999</td>\n",
       "      <td>155789000</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrX_857103bd</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>depth</td>\n",
       "      <td>False</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155789000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RD</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0060325800</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAMP7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11000</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>125298.0000000000</td>\n",
       "      <td>2.0000000000</td>\n",
       "      <td>0.0000160000</td>\n",
       "      <td>62649.0000000000</td>\n",
       "      <td>62647.0000000000</td>\n",
       "      <td>2.0000000000</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>0.9999679923</td>\n",
       "      <td>0.0000319239</td>\n",
       "      <td>0.0000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>11001</td>\n",
       "      <td>sv_X:155777999-155789000</td>\n",
       "      <td>4460</td>\n",
       "      <td>155777999</td>\n",
       "      <td>155782459</td>\n",
       "      <td>join_X:155777999-155782459</td>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-33a-5p</td>\n",
       "      <td>MIMAT0000091</td>\n",
       "      <td>SPRY3</td>\n",
       "      <td>ENSG00000168939</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>SPRY3-204</td>\n",
       "      <td>ENST00000695325</td>\n",
       "      <td>155781719</td>\n",
       "      <td>155781731</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4700000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155781719-155781731:hsa-miR-33a-5p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90375 rows × 123 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome          Source          Feature      Start        End Score  \\\n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "4             1  ensembl_havana  three_prime_UTR    1427787    1430255     .   \n",
       "...         ...             ...              ...        ...        ...   ...   \n",
       "4140          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4141          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4141          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4141          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "4152          X          havana  three_prime_UTR  155774738  155782459     .   \n",
       "\n",
       "     Strand Frame                          ID Alias external_name logic_name  \\\n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "4         +     .  transcript:ENST00000378821   NaN           NaN        NaN   \n",
       "...     ...   ...                         ...   ...           ...        ...   \n",
       "4140      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4141      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4141      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4141      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "4152      +     .  transcript:ENST00000695325   NaN           NaN        NaN   \n",
       "\n",
       "             Name         biotype description gene_id version  \\\n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "4     TMEM88B-201  protein_coding         NaN     NaN     NaN   \n",
       "...           ...             ...         ...     ...     ...   \n",
       "4140   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4141   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4141   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4141   FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "4152    SPRY3-204  protein_coding         NaN     NaN     NaN   \n",
       "\n",
       "                          Parent transcript_id constitutive ensembl_end_phase  \\\n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "4     transcript:ENST00000378821           NaN          NaN               NaN   \n",
       "...                          ...           ...          ...               ...   \n",
       "4140  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4141  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4141  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4141  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "4152  transcript:ENST00000695325           NaN          NaN               NaN   \n",
       "\n",
       "     ensembl_phase exon_id rank  \\\n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "4              NaN     NaN  NaN   \n",
       "...            ...     ...  ...   \n",
       "4140           NaN     NaN  NaN   \n",
       "4141           NaN     NaN  NaN   \n",
       "4141           NaN     NaN  NaN   \n",
       "4141           NaN     NaN  NaN   \n",
       "4152           NaN     NaN  NaN   \n",
       "\n",
       "                                                              tag  \\\n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "...                                                           ...   \n",
       "4140  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4141  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4141  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4141  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "4152  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "\n",
       "                transcript_support_level ccdsid protein_id  feature_len  \\\n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "4     2 (assigned to previous version 3)    NaN        NaN         2468   \n",
       "...                                  ...    ...        ...          ...   \n",
       "4140  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4141  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4141  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4141  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "4152                                 NaN    NaN        NaN         7721   \n",
       "\n",
       "                         feature_id   Start_sv     End_sv  \\\n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "4         feature_1:1427787-1430255    1429037    1429179   \n",
       "...                             ...        ...        ...   \n",
       "4140  feature_X:155054672-155060304  155055047  155055455   \n",
       "4141  feature_X:155054672-155060304  155055710  155057049   \n",
       "4141  feature_X:155054672-155060304  155055710  155057049   \n",
       "4141  feature_X:155054672-155060304  155055710  155057049   \n",
       "4152  feature_X:155774738-155782459  155777999  155789000   \n",
       "\n",
       "                                name svtype  samples  MULTIALLELIC ALGORITHMS  \\\n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "4     gnomAD-SV_v3_DEL_chr1_08f4439e    DEL      NaN         False      manta   \n",
       "...                              ...    ...      ...           ...        ...   \n",
       "4140  gnomAD-SV_v3_DEL_chrX_c97db132    DEL      NaN         False       wham   \n",
       "4141  gnomAD-SV_v3_DEL_chrX_b6884c3c    DEL      NaN         False      manta   \n",
       "4141  gnomAD-SV_v3_DEL_chrX_b6884c3c    DEL      NaN         False      manta   \n",
       "4141  gnomAD-SV_v3_DEL_chrX_b6884c3c    DEL      NaN         False      manta   \n",
       "4152  gnomAD-SV_v3_DEL_chrX_857103bd    DEL      NaN         False      depth   \n",
       "\n",
       "      BOTHSIDES_SUPPORT  CHR2 CPX_INTERVALS CPX_TYPE        END  END2  \\\n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "4                  True  chr1           NaN      NaN    1429179   NaN   \n",
       "...                 ...   ...           ...      ...        ...   ...   \n",
       "4140              False  chrX           NaN      NaN  155055455   NaN   \n",
       "4141              False  chrX           NaN      NaN  155057049   NaN   \n",
       "4141              False  chrX           NaN      NaN  155057049   NaN   \n",
       "4141              False  chrX           NaN      NaN  155057049   NaN   \n",
       "4152              False  chrX           NaN      NaN  155789000   NaN   \n",
       "\n",
       "      EVIDENCE  LOW_CONFIDENCE_REPETITIVE_LARGE_DUP MEMBERS          NCR  \\\n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "4           SR                                False     NaN 0.0019674799   \n",
       "...        ...                                  ...     ...          ...   \n",
       "4140  PE,RD,SR                                False     NaN 0.0012172300   \n",
       "4141     PE,SR                                False     NaN 0.0024497800   \n",
       "4141     PE,SR                                False     NaN 0.0024497800   \n",
       "4141     PE,SR                                False     NaN 0.0024497800   \n",
       "4152        RD                                False     NaN 0.0060325800   \n",
       "\n",
       "      OUTLIER_SAMPLE_ENRICHED_LENIENT    PAR  PCRMINUS_NCR  PCRPLUS_NCR  \\\n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "4                               False  False  0.0019242300 0.0024584001   \n",
       "...                               ...    ...           ...          ...   \n",
       "4140                            False  False           NaN 0.0007600070   \n",
       "4141                            False  False           NaN 0.0014355700   \n",
       "4141                            False  False           NaN 0.0014355700   \n",
       "4141                            False  False           NaN 0.0014355700   \n",
       "4152                            False   True           NaN 0.0000000000   \n",
       "\n",
       "      PESR_GT_OVERDISPERSION  POS2 PREDICTED_BREAKEND_EXONIC  \\\n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "4                      False   NaN                       NaN   \n",
       "...                      ...   ...                       ...   \n",
       "4140                   False   NaN                       NaN   \n",
       "4141                   False   NaN                       NaN   \n",
       "4141                   False   NaN                       NaN   \n",
       "4141                   False   NaN                       NaN   \n",
       "4152                   False   NaN                       NaN   \n",
       "\n",
       "     PREDICTED_COPY_GAIN PREDICTED_DUP_PARTIAL  PREDICTED_INTERGENIC  \\\n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "4                    NaN                   NaN                 False   \n",
       "...                  ...                   ...                   ...   \n",
       "4140                 NaN                   NaN                 False   \n",
       "4141                 NaN                   NaN                 False   \n",
       "4141                 NaN                   NaN                 False   \n",
       "4141                 NaN                   NaN                 False   \n",
       "4152                 NaN                   NaN                  True   \n",
       "\n",
       "     PREDICTED_INTRAGENIC_EXON_DUP PREDICTED_INTRONIC PREDICTED_INV_SPAN  \\\n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "4                              NaN                NaN                NaN   \n",
       "...                            ...                ...                ...   \n",
       "4140                           NaN                NaN                NaN   \n",
       "4141                           NaN                NaN                NaN   \n",
       "4141                           NaN                NaN                NaN   \n",
       "4141                           NaN                NaN                NaN   \n",
       "4152                           NaN                NaN                NaN   \n",
       "\n",
       "     PREDICTED_LOF PREDICTED_MSV_EXON_OVERLAP PREDICTED_NEAREST_TSS  \\\n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "4              NaN                        NaN                   NaN   \n",
       "...            ...                        ...                   ...   \n",
       "4140           NaN                        NaN                   NaN   \n",
       "4141           NaN                        NaN                   NaN   \n",
       "4141           NaN                        NaN                   NaN   \n",
       "4141           NaN                        NaN                   NaN   \n",
       "4152           NaN                        NaN                 VAMP7   \n",
       "\n",
       "      PREDICTED_NONCODING_BREAKPOINT  PREDICTED_NONCODING_SPAN  \\\n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "...                              ...                       ...   \n",
       "4140                             NaN                       NaN   \n",
       "4141                             NaN                       NaN   \n",
       "4141                             NaN                       NaN   \n",
       "4141                             NaN                       NaN   \n",
       "4152                             NaN                       NaN   \n",
       "\n",
       "     PREDICTED_PARTIAL_DISPERSED_DUP PREDICTED_PARTIAL_EXON_DUP  \\\n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "4                                NaN                        NaN   \n",
       "...                              ...                        ...   \n",
       "4140                             NaN                        NaN   \n",
       "4141                             NaN                        NaN   \n",
       "4141                             NaN                        NaN   \n",
       "4141                             NaN                        NaN   \n",
       "4152                             NaN                        NaN   \n",
       "\n",
       "     PREDICTED_PROMOTER PREDICTED_TSS_DUP PREDICTED_UTR  RESOLVED_POSTHOC  \\\n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "4                   NaN               NaN       TMEM88B             False   \n",
       "...                 ...               ...           ...               ...   \n",
       "4140                NaN               NaN        FUNDC2             False   \n",
       "4141                NaN               NaN        FUNDC2             False   \n",
       "4141                NaN               NaN        FUNDC2             False   \n",
       "4141                NaN               NaN        FUNDC2             False   \n",
       "4152                NaN               NaN           NaN             False   \n",
       "\n",
       "     SOURCE  SVLEN SVTYPE UNRESOLVED_TYPE                AN            AC  \\\n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "4       NaN    141    DEL             NaN 125770.0000000000  1.0000000000   \n",
       "...     ...    ...    ...             ...               ...           ...   \n",
       "4140    NaN    407    DEL             NaN  95749.0000000000 19.0000000000   \n",
       "4141    NaN   1338    DEL             NaN  95654.0000000000 68.0000000000   \n",
       "4141    NaN   1338    DEL             NaN  95654.0000000000 68.0000000000   \n",
       "4141    NaN   1338    DEL             NaN  95654.0000000000 68.0000000000   \n",
       "4152    NaN  11000    DEL             NaN 125298.0000000000  2.0000000000   \n",
       "\n",
       "               AF       N_BI_GENOS         N_HOMREF         N_HET  \\\n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "4    0.0000080000 62885.0000000000 62884.0000000000  1.0000000000   \n",
       "...           ...              ...              ...           ...   \n",
       "4140 0.0001984350 62944.0000000000 62925.0000000000 19.0000000000   \n",
       "4141 0.0007108960 62900.0000000000 62833.0000000000 66.0000000000   \n",
       "4141 0.0007108960 62900.0000000000 62833.0000000000 66.0000000000   \n",
       "4141 0.0007108960 62900.0000000000 62833.0000000000 66.0000000000   \n",
       "4152 0.0000160000 62649.0000000000 62647.0000000000  2.0000000000   \n",
       "\n",
       "         N_HOMALT  FREQ_HOMREF     FREQ_HET  FREQ_HOMALT  CN_NUMBER CN_COUNT  \\\n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "4    0.0000000000 0.9999840260 0.0000159020 0.0000000000        NaN      NaN   \n",
       "...           ...          ...          ...          ...        ...      ...   \n",
       "4140 0.0000000000 0.9996979833 0.0003018560 0.0000000000        NaN      NaN   \n",
       "4141 1.0000000000 0.9989349842 0.0010492801 0.0000158983        NaN      NaN   \n",
       "4141 1.0000000000 0.9989349842 0.0010492801 0.0000158983        NaN      NaN   \n",
       "4141 1.0000000000 0.9989349842 0.0010492801 0.0000158983        NaN      NaN   \n",
       "4152 0.0000000000 0.9999679923 0.0000319239 0.0000000000        NaN      NaN   \n",
       "\n",
       "     CN_STATUS CN_FREQ  CN_NONREF_COUNT  CN_NONREF_FREQ FILTER  sv_len  \\\n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "4          NaN     NaN              NaN             NaN   PASS     142   \n",
       "...        ...     ...              ...             ...    ...     ...   \n",
       "4140       NaN     NaN              NaN             NaN   PASS     408   \n",
       "4141       NaN     NaN              NaN             NaN   PASS    1339   \n",
       "4141       NaN     NaN              NaN             NaN   PASS    1339   \n",
       "4141       NaN     NaN              NaN             NaN   PASS    1339   \n",
       "4152       NaN     NaN              NaN             NaN   PASS   11001   \n",
       "\n",
       "                         sv_id  join_len  join_start   join_end  \\\n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "4         sv_1:1429037-1429179       142     1429037    1429179   \n",
       "...                        ...       ...         ...        ...   \n",
       "4140  sv_X:155055047-155055455       408   155055047  155055455   \n",
       "4141  sv_X:155055710-155057049      1339   155055710  155057049   \n",
       "4141  sv_X:155055710-155057049      1339   155055710  155057049   \n",
       "4141  sv_X:155055710-155057049      1339   155055710  155057049   \n",
       "4152  sv_X:155777999-155789000      4460   155777999  155782459   \n",
       "\n",
       "                         join_id      join_type  joins_count       species  \\\n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "4         join_1:1429037-1429179  sv_in_feature            1  Homo sapiens   \n",
       "...                          ...            ...          ...           ...   \n",
       "4140  join_X:155055047-155055455  sv_in_feature            1  Homo sapiens   \n",
       "4141  join_X:155055710-155057049  sv_in_feature            1  Homo sapiens   \n",
       "4141  join_X:155055710-155057049  sv_in_feature            1  Homo sapiens   \n",
       "4141  join_X:155055710-155057049  sv_in_feature            1  Homo sapiens   \n",
       "4152  join_X:155777999-155782459  sv_right_free            1  Homo sapiens   \n",
       "\n",
       "          mirna_name      mirna_id gene_name  gene_id_tarbase gene_location  \\\n",
       "4      hsa-let-7a-5p  MIMAT0000062   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7b-5p  MIMAT0000063   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7c-5p  MIMAT0000064   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7f-5p  MIMAT0000067   TMEM88B  ENSG00000205116          3UTR   \n",
       "4      hsa-let-7g-5p  MIMAT0000414   TMEM88B  ENSG00000205116          3UTR   \n",
       "...              ...           ...       ...              ...           ...   \n",
       "4140   hsa-let-7i-5p  MIMAT0000415    FUNDC2  ENSG00000165775          3UTR   \n",
       "4141    hsa-miR-1275  MIMAT0005929    FUNDC2  ENSG00000165775          3UTR   \n",
       "4141   hsa-miR-32-5p  MIMAT0000090    FUNDC2  ENSG00000165775          3UTR   \n",
       "4141  hsa-miR-92b-3p  MIMAT0003218    FUNDC2  ENSG00000165775          3UTR   \n",
       "4152  hsa-miR-33a-5p  MIMAT0000091     SPRY3  ENSG00000168939          3UTR   \n",
       "\n",
       "     transcript_name transcript_id_tarbase  Start_tarbase  End_tarbase  \\\n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "4        TMEM88B-201       ENST00000378821        1429042      1429054   \n",
       "...              ...                   ...            ...          ...   \n",
       "4140      FUNDC2-201       ENST00000369498      155055264    155055276   \n",
       "4141      FUNDC2-201       ENST00000369498      155055701    155055713   \n",
       "4141      FUNDC2-201       ENST00000369498      155056053    155056065   \n",
       "4141      FUNDC2-201       ENST00000369498      155056053    155056065   \n",
       "4152       SPRY3-204       ENST00000695325      155781719    155781731   \n",
       "\n",
       "     Strand_tarbase experimental_method regulation  \\\n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "4                 +           HITS-CLIP   Negative   \n",
       "...             ...                 ...        ...   \n",
       "4140              +           HITS-CLIP   Negative   \n",
       "4141              +            PAR-CLIP   Negative   \n",
       "4141              +           HITS-CLIP   Negative   \n",
       "4141              +           HITS-CLIP   Negative   \n",
       "4152              +           HITS-CLIP   Negative   \n",
       "\n",
       "                            tissue cell_line   article_pubmed_id  confidence  \\\n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "4          Brain - Cingulate gyrus       NaN 24389009.0000000000           1   \n",
       "...                            ...       ...                 ...         ...   \n",
       "4140  Brain - Primary Motor Cortex       NaN 30455455.0000000000           1   \n",
       "4141                        Pleura      BC-3 22100165.0000000000           1   \n",
       "4141          Brain - Motor cortex       NaN 24389009.0000000000           1   \n",
       "4141          Brain - Motor cortex       NaN 24389009.0000000000           1   \n",
       "4152          Brain - Motor cortex       NaN 24389009.0000000000           1   \n",
       "\n",
       "     interaction_group      cell_type  microt_score            comment  \\\n",
       "4              primary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "4            secondary            NaN           NaN    Male - 44 years   \n",
       "...                ...            ...           ...                ...   \n",
       "4140         secondary            NaN  0.4800000000  Gestation week 15   \n",
       "4141           primary  B lymphocytes  0.6800000000                NaN   \n",
       "4141           primary            NaN  0.4200000000    Male - 44 years   \n",
       "4141         secondary            NaN  0.5000000000    Male - 44 years   \n",
       "4152           primary            NaN  0.4700000000    Male - 44 years   \n",
       "\n",
       "      tarbase_join_len                                    tarbase_id  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7a-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7b-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7c-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7f-5p  \n",
       "4                   12       tarbase_1:1429042-1429054:hsa-let-7g-5p  \n",
       "...                ...                                           ...  \n",
       "4140                12   tarbase_X:155055264-155055276:hsa-let-7i-5p  \n",
       "4141                 3    tarbase_X:155055701-155055713:hsa-miR-1275  \n",
       "4141                12   tarbase_X:155056053-155056065:hsa-miR-32-5p  \n",
       "4141                12  tarbase_X:155056053-155056065:hsa-miR-92b-3p  \n",
       "4152                12  tarbase_X:155781719-155781731:hsa-miR-33a-5p  \n",
       "\n",
       "[90375 rows x 123 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(pr_for_igv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1769e350",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/1204189949.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  output_features_pr['Source'] = 'GFF'\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_75092/1204189949.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  output_features_pr['color'] = 'blue'\n"
     ]
    }
   ],
   "source": [
    "# Add GFF\n",
    "output_features_pr = pr_for_igv[['Chromosome', 'Feature', 'Start', 'End', 'Strand', 'Name', 'biotype', 'tag', 'ID']]\n",
    "output_features_pr = output_features_pr.drop_duplicates()\n",
    "output_features_pr['Source'] = 'GFF'\n",
    "output_features_pr['color'] = 'blue'\n",
    "\n",
    "# Add gnomAD SV\n",
    "output_sv_pr = pr_for_igv[['Chromosome', 'Start_sv', 'End_sv', 'svtype', 'name', 'sv_id', 'AF', 'N_HET', 'N_HOMALT']]\n",
    "output_sv_pr = output_sv_pr.drop_duplicates()\n",
    "output_sv_pr['Source'] = 'gnomAD_SV'\n",
    "output_sv_pr['color'] = 'red'\n",
    "output_sv_pr = pr.PyRanges(output_sv_pr.rename(columns={'Start_sv': 'Start', 'End_sv': 'End', 'svtype': 'Feature', 'name': 'Name', 'sv_id': 'ID'}))\n",
    "\n",
    "# Add joins data\n",
    "output_joins_pr = pr_for_igv[['Chromosome', 'join_start', 'join_end', 'join_type', 'join_id']]\n",
    "output_joins_pr = output_joins_pr.drop_duplicates()\n",
    "output_joins_pr['Source'] = 'join'\n",
    "output_joins_pr['color'] = 'orange'\n",
    "output_joins_pr = pr.PyRanges(output_joins_pr.rename(columns={'join_start': 'Start', 'join_end': 'End', 'join_type': 'Feature', 'join_id': 'ID'}))\n",
    "\n",
    "# Add miRNA TarBase\n",
    "output_tarbase_pr = pr_for_igv[['Chromosome', 'Start_tarbase', 'End_tarbase', 'Strand_tarbase', 'mirna_name', 'interaction_group', 'mirna_id', 'tarbase_id']]\n",
    "output_tarbase_pr = output_tarbase_pr.drop_duplicates(subset=['mirna_id', 'tarbase_id'])\n",
    "output_tarbase_pr['Source'] = 'tarbase'\n",
    "output_tarbase_pr['color'] = 'green'\n",
    "output_tarbase_pr = pr.PyRanges(output_tarbase_pr.rename(columns={'Start_tarbase': 'Start', 'End_tarbase': 'End', 'Strand_tarbase': 'Strand', 'interaction_group': 'Feature', 'mirna_name': 'Name', 'tarbase_id': 'ID'}))\n",
    "\n",
    "# Add offtarget GFF (features) data\n",
    "output_offtargets_pr = features_sv_offtereget_joined_pr[['Chromosome', 'Start_offtarget', 'End_offtarget', 'Feature_offtarget', 'offtarget_id']]\n",
    "output_offtargets_pr = output_offtargets_pr.query('Start_offtarget == Start_offtarget')\n",
    "output_offtargets_pr = output_offtargets_pr.drop_duplicates()\n",
    "output_offtargets_pr['Source'] = 'GFF_offtarget'\n",
    "output_offtargets_pr['color'] = '#d9d9d9'\n",
    "output_offtargets_pr = pr.PyRanges(output_offtargets_pr.rename(columns={'Start_offtarget': 'Start', 'End_offtarget': 'End', 'Feature_offtarget': 'Feature', 'offtarget_id': 'ID'}))\n",
    "\n",
    "output_igv_pr = pd.concat([output_features_pr, output_sv_pr, output_joins_pr, output_tarbase_pr, output_offtargets_pr], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5b16cad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tarbase</td>\n",
       "      <td>41904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GFF_offtarget</td>\n",
       "      <td>2921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gnomAD_SV</td>\n",
       "      <td>2294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>join</td>\n",
       "      <td>2290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFF</td>\n",
       "      <td>1669</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Source  count\n",
       "0        tarbase  41904\n",
       "1  GFF_offtarget   2921\n",
       "2      gnomAD_SV   2294\n",
       "3           join   2290\n",
       "4            GFF   1669"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stats\n",
    "output_igv_pr['Source'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f2cb8479",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill NA and save\n",
    "output_igv_pr = output_igv_pr.reset_index(drop=True)\n",
    "#output_igv_pr = pr.PyRanges(output_igv_pr.fillna('.'))\n",
    "\n",
    "output_igv_pr.to_gff3(output_dir / 'regions_for_igv.gff3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541f18c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a29f19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "668dfa8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "662de559",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e9d4e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1fd440d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
