{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import pandas as pd, numpy as np, pyranges as pr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## Fetch data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.chr.gff3.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/gnomad/gnomad.v4.1.sv.sites.bed.gz' already there; not retrieving.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.chr.gff3.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e446fd43",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_pr = pr.read_gff3(str(gff_filepath))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_67448/387753410.py:3: DtypeWarning: Columns (10,11,24,25,26,28,29,30,31,32,33,36,37,38,39,40,42,57,58,59) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gnomad_sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=list(range(len(target_columns))), names=target_columns, comment='#')\n"
     ]
    }
   ],
   "source": [
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ']\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End']\n",
    "gnomad_sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=list(range(len(target_columns))), names=target_columns, comment='#')\n",
    "gnomad_sv_df['Chromosome'] = gnomad_sv_df['Chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "666c6691",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>samples</th>\n",
       "      <th>MULTIALLELIC</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>CHR2</th>\n",
       "      <th>...</th>\n",
       "      <th>N_HOMALT</th>\n",
       "      <th>FREQ_HOMREF</th>\n",
       "      <th>FREQ_HET</th>\n",
       "      <th>FREQ_HOMALT</th>\n",
       "      <th>CN_NUMBER</th>\n",
       "      <th>CN_COUNT</th>\n",
       "      <th>CN_STATUS</th>\n",
       "      <th>CN_FREQ</th>\n",
       "      <th>CN_NONREF_COUNT</th>\n",
       "      <th>CN_NONREF_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9999</td>\n",
       "      <td>295666</td>\n",
       "      <td>gnomAD-SV_v3_DUP_chr1_01c2781c</td>\n",
       "      <td>DUP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>depth</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996005</td>\n",
       "      <td>0.003995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10433</td>\n",
       "      <td>10434</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_1a45f73a</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr2</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.772378</td>\n",
       "      <td>0.226975</td>\n",
       "      <td>0.000646</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10439</td>\n",
       "      <td>10440</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_3fa36917</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.991599</td>\n",
       "      <td>0.008401</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10449</td>\n",
       "      <td>10450</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_933a2971</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr18</td>\n",
       "      <td>...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.354435</td>\n",
       "      <td>0.643333</td>\n",
       "      <td>0.002232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10463</td>\n",
       "      <td>10464</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_7bbf34b5</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr12</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.926078</td>\n",
       "      <td>0.073875</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154481</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56878257</td>\n",
       "      <td>56878816</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrY_2d25b4a6</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999726</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154482</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56878270</td>\n",
       "      <td>56878718</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrY_ab4f307f</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.582192</td>\n",
       "      <td>0.390411</td>\n",
       "      <td>0.027397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154483</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56878973</td>\n",
       "      <td>56879563</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrY_0db85cf8</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999896</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154484</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56880080</td>\n",
       "      <td>56880081</td>\n",
       "      <td>gnomAD-SV_v3_BND_chrY_0151e635</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999967</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154485</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56884435</td>\n",
       "      <td>56884467</td>\n",
       "      <td>gnomAD-SV_v3_INS_chrY_c14eea89</td>\n",
       "      <td>INS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999934</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2154486 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End                            name svtype  \\\n",
       "0             chr1      9999    295666  gnomAD-SV_v3_DUP_chr1_01c2781c    DUP   \n",
       "1             chr1     10433     10434  gnomAD-SV_v3_BND_chr1_1a45f73a    BND   \n",
       "2             chr1     10439     10440  gnomAD-SV_v3_BND_chr1_3fa36917    BND   \n",
       "3             chr1     10449     10450  gnomAD-SV_v3_BND_chr1_933a2971    BND   \n",
       "4             chr1     10463     10464  gnomAD-SV_v3_BND_chr1_7bbf34b5    BND   \n",
       "...            ...       ...       ...                             ...    ...   \n",
       "2154481       chrY  56878257  56878816  gnomAD-SV_v3_DEL_chrY_2d25b4a6    DEL   \n",
       "2154482       chrY  56878270  56878718  gnomAD-SV_v3_DEL_chrY_ab4f307f    DEL   \n",
       "2154483       chrY  56878973  56879563  gnomAD-SV_v3_DEL_chrY_0db85cf8    DEL   \n",
       "2154484       chrY  56880080  56880081  gnomAD-SV_v3_BND_chrY_0151e635    BND   \n",
       "2154485       chrY  56884435  56884467  gnomAD-SV_v3_INS_chrY_c14eea89    INS   \n",
       "\n",
       "         samples  MULTIALLELIC ALGORITHMS  BOTHSIDES_SUPPORT   CHR2  ...  \\\n",
       "0            NaN         False      depth              False   chr1  ...   \n",
       "1            NaN         False      manta               True   chr2  ...   \n",
       "2            NaN         False      manta               True   chr3  ...   \n",
       "3            NaN         False      manta               True  chr18  ...   \n",
       "4            NaN         False      manta               True  chr12  ...   \n",
       "...          ...           ...        ...                ...    ...  ...   \n",
       "2154481      NaN         False       wham              False   chrY  ...   \n",
       "2154482      NaN         False       wham              False   chrY  ...   \n",
       "2154483      NaN         False       wham              False   chrY  ...   \n",
       "2154484      NaN         False       wham              False   chrY  ...   \n",
       "2154485      NaN         False      manta              False   chrY  ...   \n",
       "\n",
       "        N_HOMALT FREQ_HOMREF  FREQ_HET  FREQ_HOMALT CN_NUMBER  CN_COUNT  \\\n",
       "0            0.0    0.996005  0.003995     0.000000       NaN       NaN   \n",
       "1           24.0    0.772378  0.226975     0.000646       NaN       NaN   \n",
       "2            0.0    0.991599  0.008401     0.000000       NaN       NaN   \n",
       "3           75.0    0.354435  0.643333     0.002232       NaN       NaN   \n",
       "4            2.0    0.926078  0.073875     0.000047       NaN       NaN   \n",
       "...          ...         ...       ...          ...       ...       ...   \n",
       "2154481      0.0    0.999726  0.000274     0.000000       NaN       NaN   \n",
       "2154482      4.0    0.582192  0.390411     0.027397       NaN       NaN   \n",
       "2154483      0.0    0.999896  0.000104     0.000000       NaN       NaN   \n",
       "2154484      0.0    0.999967  0.000033     0.000000       NaN       NaN   \n",
       "2154485      0.0    0.999934  0.000066     0.000000       NaN       NaN   \n",
       "\n",
       "        CN_STATUS  CN_FREQ  CN_NONREF_COUNT  CN_NONREF_FREQ  \n",
       "0             NaN      NaN              NaN             NaN  \n",
       "1             NaN      NaN              NaN             NaN  \n",
       "2             NaN      NaN              NaN             NaN  \n",
       "3             NaN      NaN              NaN             NaN  \n",
       "4             NaN      NaN              NaN             NaN  \n",
       "...           ...      ...              ...             ...  \n",
       "2154481       NaN      NaN              NaN             NaN  \n",
       "2154482       NaN      NaN              NaN             NaN  \n",
       "2154483       NaN      NaN              NaN             NaN  \n",
       "2154484       NaN      NaN              NaN             NaN  \n",
       "2154485       NaN      NaN              NaN             NaN  \n",
       "\n",
       "[2154486 rows x 62 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad_sv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0b01e50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad_sv_pr = pr.PyRanges(gnomad_sv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ca56850",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>Alias</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>constitutive</th>\n",
       "      <th>ensembl_end_phase</th>\n",
       "      <th>ensembl_phase</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>protein_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>0</td>\n",
       "      <td>248956422</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chromosome:1</td>\n",
       "      <td>CM000663.2,chr1,NC_000001.11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10468</td>\n",
       "      <td>11240</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10649</td>\n",
       "      <td>10657</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10654</td>\n",
       "      <td>10657</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10677</td>\n",
       "      <td>10687</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7491838</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215803</td>\n",
       "      <td>57215813</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7491839</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215831</td>\n",
       "      <td>57215840</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7491840</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832</td>\n",
       "      <td>57215842</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7491841</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215861</td>\n",
       "      <td>57215867</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7491842</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215861</td>\n",
       "      <td>57215870</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7491843 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-------------------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature           | Start     | +24   |\n",
       "| (category)   | (object)   | (object)          | (int64)   | ...   |\n",
       "|--------------+------------+-------------------+-----------+-------|\n",
       "| 1            | GRCh38     | chromosome        | 0         | ...   |\n",
       "| 1            | .          | biological_region | 10468     | ...   |\n",
       "| 1            | .          | biological_region | 10649     | ...   |\n",
       "| 1            | .          | biological_region | 10654     | ...   |\n",
       "| ...          | ...        | ...               | ...       | ...   |\n",
       "| Y            | .          | biological_region | 57215831  | ...   |\n",
       "| Y            | .          | biological_region | 57215832  | ...   |\n",
       "| Y            | .          | biological_region | 57215861  | ...   |\n",
       "| Y            | .          | biological_region | 57215861  | ...   |\n",
       "+--------------+------------+-------------------+-----------+-------+\n",
       "Unstranded PyRanges object has 7,491,843 rows and 28 columns from 25 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "Considered unstranded due to these Strand values: '.'\n",
       "24 hidden columns: End, Score, Strand, Frame, ID, Alias, external_name, ... (+ 17 more.)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a976904",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>samples</th>\n",
       "      <th>MULTIALLELIC</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>CHR2</th>\n",
       "      <th>...</th>\n",
       "      <th>N_HOMALT</th>\n",
       "      <th>FREQ_HOMREF</th>\n",
       "      <th>FREQ_HET</th>\n",
       "      <th>FREQ_HOMALT</th>\n",
       "      <th>CN_NUMBER</th>\n",
       "      <th>CN_COUNT</th>\n",
       "      <th>CN_STATUS</th>\n",
       "      <th>CN_FREQ</th>\n",
       "      <th>CN_NONREF_COUNT</th>\n",
       "      <th>CN_NONREF_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9999</td>\n",
       "      <td>295666</td>\n",
       "      <td>gnomAD-SV_v3_DUP_chr1_01c2781c</td>\n",
       "      <td>DUP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>depth</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996005</td>\n",
       "      <td>0.003995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10433</td>\n",
       "      <td>10434</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_1a45f73a</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr2</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.772378</td>\n",
       "      <td>0.226975</td>\n",
       "      <td>0.000646</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10439</td>\n",
       "      <td>10440</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_3fa36917</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.991599</td>\n",
       "      <td>0.008401</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10449</td>\n",
       "      <td>10450</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_933a2971</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr18</td>\n",
       "      <td>...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.354435</td>\n",
       "      <td>0.643333</td>\n",
       "      <td>0.002232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10463</td>\n",
       "      <td>10464</td>\n",
       "      <td>gnomAD-SV_v3_BND_chr1_7bbf34b5</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>True</td>\n",
       "      <td>chr12</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.926078</td>\n",
       "      <td>0.073875</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154481</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56878257</td>\n",
       "      <td>56878816</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrY_2d25b4a6</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999726</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154482</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56878270</td>\n",
       "      <td>56878718</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrY_ab4f307f</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.582192</td>\n",
       "      <td>0.390411</td>\n",
       "      <td>0.027397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154483</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56878973</td>\n",
       "      <td>56879563</td>\n",
       "      <td>gnomAD-SV_v3_DEL_chrY_0db85cf8</td>\n",
       "      <td>DEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999896</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154484</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56880080</td>\n",
       "      <td>56880081</td>\n",
       "      <td>gnomAD-SV_v3_BND_chrY_0151e635</td>\n",
       "      <td>BND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>wham</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999967</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154485</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56884435</td>\n",
       "      <td>56884467</td>\n",
       "      <td>gnomAD-SV_v3_INS_chrY_c14eea89</td>\n",
       "      <td>INS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>manta</td>\n",
       "      <td>False</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999934</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2154486 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End                            name svtype  \\\n",
       "0             chr1      9999    295666  gnomAD-SV_v3_DUP_chr1_01c2781c    DUP   \n",
       "1             chr1     10433     10434  gnomAD-SV_v3_BND_chr1_1a45f73a    BND   \n",
       "2             chr1     10439     10440  gnomAD-SV_v3_BND_chr1_3fa36917    BND   \n",
       "3             chr1     10449     10450  gnomAD-SV_v3_BND_chr1_933a2971    BND   \n",
       "4             chr1     10463     10464  gnomAD-SV_v3_BND_chr1_7bbf34b5    BND   \n",
       "...            ...       ...       ...                             ...    ...   \n",
       "2154481       chrY  56878257  56878816  gnomAD-SV_v3_DEL_chrY_2d25b4a6    DEL   \n",
       "2154482       chrY  56878270  56878718  gnomAD-SV_v3_DEL_chrY_ab4f307f    DEL   \n",
       "2154483       chrY  56878973  56879563  gnomAD-SV_v3_DEL_chrY_0db85cf8    DEL   \n",
       "2154484       chrY  56880080  56880081  gnomAD-SV_v3_BND_chrY_0151e635    BND   \n",
       "2154485       chrY  56884435  56884467  gnomAD-SV_v3_INS_chrY_c14eea89    INS   \n",
       "\n",
       "         samples  MULTIALLELIC ALGORITHMS  BOTHSIDES_SUPPORT   CHR2  ...  \\\n",
       "0            NaN         False      depth              False   chr1  ...   \n",
       "1            NaN         False      manta               True   chr2  ...   \n",
       "2            NaN         False      manta               True   chr3  ...   \n",
       "3            NaN         False      manta               True  chr18  ...   \n",
       "4            NaN         False      manta               True  chr12  ...   \n",
       "...          ...           ...        ...                ...    ...  ...   \n",
       "2154481      NaN         False       wham              False   chrY  ...   \n",
       "2154482      NaN         False       wham              False   chrY  ...   \n",
       "2154483      NaN         False       wham              False   chrY  ...   \n",
       "2154484      NaN         False       wham              False   chrY  ...   \n",
       "2154485      NaN         False      manta              False   chrY  ...   \n",
       "\n",
       "        N_HOMALT FREQ_HOMREF  FREQ_HET  FREQ_HOMALT CN_NUMBER  CN_COUNT  \\\n",
       "0            0.0    0.996005  0.003995     0.000000       NaN       NaN   \n",
       "1           24.0    0.772378  0.226975     0.000646       NaN       NaN   \n",
       "2            0.0    0.991599  0.008401     0.000000       NaN       NaN   \n",
       "3           75.0    0.354435  0.643333     0.002232       NaN       NaN   \n",
       "4            2.0    0.926078  0.073875     0.000047       NaN       NaN   \n",
       "...          ...         ...       ...          ...       ...       ...   \n",
       "2154481      0.0    0.999726  0.000274     0.000000       NaN       NaN   \n",
       "2154482      4.0    0.582192  0.390411     0.027397       NaN       NaN   \n",
       "2154483      0.0    0.999896  0.000104     0.000000       NaN       NaN   \n",
       "2154484      0.0    0.999967  0.000033     0.000000       NaN       NaN   \n",
       "2154485      0.0    0.999934  0.000066     0.000000       NaN       NaN   \n",
       "\n",
       "        CN_STATUS  CN_FREQ  CN_NONREF_COUNT  CN_NONREF_FREQ  \n",
       "0             NaN      NaN              NaN             NaN  \n",
       "1             NaN      NaN              NaN             NaN  \n",
       "2             NaN      NaN              NaN             NaN  \n",
       "3             NaN      NaN              NaN             NaN  \n",
       "4             NaN      NaN              NaN             NaN  \n",
       "...           ...      ...              ...             ...  \n",
       "2154481       NaN      NaN              NaN             NaN  \n",
       "2154482       NaN      NaN              NaN             NaN  \n",
       "2154483       NaN      NaN              NaN             NaN  \n",
       "2154484       NaN      NaN              NaN             NaN  \n",
       "2154485       NaN      NaN              NaN             NaN  \n",
       "\n",
       "[2154486 rows x 62 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad_sv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5fa95e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "260a2444",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c1ed5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa670cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3f5b08",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
