{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import pandas as pd, numpy as np, pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a644ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.float_format', '{:.10f}'.format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## Fetch data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.chr.gff3.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/gnomad/gnomad.v4.1.sv.sites.bed.gz' already there; not retrieving.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.chr.gff3.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e446fd43",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b315f3",
   "metadata": {},
   "source": [
    "### Create PR/DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_pr = pr.read_gff3(str(gff_filepath))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_67448/3615737592.py:3: DtypeWarning: Columns (10,11,24,25,26,28,29,30,31,32,33,36,37,38,39,40,42,57,58,59) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gnomad_sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=list(range(len(target_columns))), names=target_columns, comment='#')\n"
     ]
    }
   ],
   "source": [
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ']\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End']\n",
    "gnomad_sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=list(range(len(target_columns))), names=target_columns, comment='#')\n",
    "gnomad_sv_df['Chromosome'] = gnomad_sv_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "gnomad_sv_pr = pr.PyRanges(gnomad_sv_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b001f54",
   "metadata": {},
   "source": [
    "### EDA (Exploratory Data Analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "788e1cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Feature\n",
       "exon                      3673949\n",
       "CDS                       2284258\n",
       "five_prime_UTR             426597\n",
       "three_prime_UTR            340566\n",
       "mRNA                       233574\n",
       "lnc_RNA                    223403\n",
       "biological_region          180392\n",
       "ncRNA_gene                  41946\n",
       "transcript                  28799\n",
       "gene                        21547\n",
       "pseudogenic_transcript      15201\n",
       "pseudogene                  15198\n",
       "snRNA                        1906\n",
       "miRNA                        1879\n",
       "unconfirmed_transcript       1108\n",
       "snoRNA                        942\n",
       "V_gene_segment                253\n",
       "J_gene_segment                 97\n",
       "scRNA                          49\n",
       "rRNA                           49\n",
       "D_gene_segment                 42\n",
       "C_gene_segment                 29\n",
       "chromosome                     25\n",
       "tRNA                           22\n",
       "processed_transcript           12\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_pr.Feature.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a9e7f9eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "biotype\n",
       "protein_coding                        231543\n",
       "lncRNA                                224032\n",
       "retained_intron                        34239\n",
       "protein_coding_CDS_not_defined         26573\n",
       "nonsense_mediated_decay                21949\n",
       "processed_pseudogene                   18975\n",
       "misc_RNA                                4414\n",
       "unprocessed_pseudogene                  3898\n",
       "snRNA                                   3802\n",
       "miRNA                                   3758\n",
       "transcribed_unprocessed_pseudogene      3176\n",
       "transcribed_processed_pseudogene        2298\n",
       "TEC                                     2127\n",
       "snoRNA                                  1884\n",
       "rRNA_pseudogene                          994\n",
       "transcribed_unitary_pseudogene           402\n",
       "IG_V_pseudogene                          374\n",
       "IG_V_gene                                292\n",
       "TR_V_gene                                214\n",
       "unitary_pseudogene                       178\n",
       "TR_J_gene                                158\n",
       "non_stop_decay                           105\n",
       "scaRNA                                    98\n",
       "rRNA                                      94\n",
       "protein_coding_LoF                        74\n",
       "IG_D_gene                                 74\n",
       "TR_V_pseudogene                           66\n",
       "Mt_tRNA                                   44\n",
       "artifact                                  38\n",
       "IG_C_gene                                 37\n",
       "IG_J_gene                                 36\n",
       "IG_C_pseudogene                           18\n",
       "ribozyme                                  16\n",
       "TR_C_gene                                 12\n",
       "processed_transcript                      12\n",
       "TR_D_gene                                 10\n",
       "sRNA                                      10\n",
       "TR_J_pseudogene                            8\n",
       "vault_RNA                                  8\n",
       "IG_J_pseudogene                            6\n",
       "translated_processed_pseudogene            4\n",
       "Mt_rRNA                                    4\n",
       "IG_pseudogene                              2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_pr.biotype.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "011dad84",
   "metadata": {},
   "source": [
    "### Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "79f6e1ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_pr = gff_pr[gff_pr.Feature == 'mRNA']\n",
    "three_utrs_pr = gff_pr[gff_pr.Feature == 'three_prime_UTR']\n",
    "\n",
    "mrna_df = mrna_pr.df\n",
    "three_utrs_df = three_utrs_pr.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1834b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation of 3'-UTRs with mRNA data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level'] # Replace the data in these empty 3'-UTRs columns with mRNA data\n",
    "mrna_df_short = mrna_df[target_columns]\n",
    "three_utrs_df_annotated = three_utrs_df.merge(mrna_df_short, how='left', left_on='Parent', right_on='ID', suffixes=['_3utr', '']) \n",
    "three_utrs_df_annotated = three_utrs_df_annotated.drop(['Score', 'Frame'], axis=1) # Remove columns with \".\"\n",
    "three_utrs_df_annotated = three_utrs_df_annotated.dropna(axis=1, how='all') # Remove empty columns\n",
    "three_utrs_df_annotated = pr.PyRanges(three_utrs_df_annotated)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "60cfd4c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Parent_3utr</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>biotype</th>\n",
       "      <th>Parent</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>70008</td>\n",
       "      <td>71585</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript:ENST00000641515</td>\n",
       "      <td>transcript:ENST00000641515</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000186092</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonica...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944153</td>\n",
       "      <td>944572</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript:ENST00000968544</td>\n",
       "      <td>transcript:ENST00000968544</td>\n",
       "      <td>SAMD11-218</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000187634</td>\n",
       "      <td>gencode_basic,gencode_primary</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944153</td>\n",
       "      <td>944574</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript:ENST00000616016</td>\n",
       "      <td>transcript:ENST00000616016</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000187634</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonica...</td>\n",
       "      <td>5 (assigned to previous version 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944153</td>\n",
       "      <td>944574</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript:ENST00000618323</td>\n",
       "      <td>transcript:ENST00000618323</td>\n",
       "      <td>SAMD11-213</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000187634</td>\n",
       "      <td>gencode_basic</td>\n",
       "      <td>5 (assigned to previous version 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944153</td>\n",
       "      <td>944574</td>\n",
       "      <td>+</td>\n",
       "      <td>transcript:ENST00000968542</td>\n",
       "      <td>transcript:ENST00000968542</td>\n",
       "      <td>SAMD11-216</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000187634</td>\n",
       "      <td>gencode_basic,gencode_primary</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340561</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>24763068</td>\n",
       "      <td>24764925</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript:ENST00000446723</td>\n",
       "      <td>transcript:ENST00000446723</td>\n",
       "      <td>DAZ3-203</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000187191</td>\n",
       "      <td>gencode_basic,gencode_primary</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340562</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>24768898</td>\n",
       "      <td>24768933</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript:ENST00000446723</td>\n",
       "      <td>transcript:ENST00000446723</td>\n",
       "      <td>DAZ3-203</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000187191</td>\n",
       "      <td>gencode_basic,gencode_primary</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340563</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25030900</td>\n",
       "      <td>25031222</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript:ENST00000382287</td>\n",
       "      <td>transcript:ENST00000382287</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000185894</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonica...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340564</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25031316</td>\n",
       "      <td>25031441</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript:ENST00000382287</td>\n",
       "      <td>transcript:ENST00000382287</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000185894</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonica...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340565</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25037991</td>\n",
       "      <td>25038097</td>\n",
       "      <td>-</td>\n",
       "      <td>transcript:ENST00000382287</td>\n",
       "      <td>transcript:ENST00000382287</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>gene:ENSG00000185894</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonica...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>340566 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+----------------+-----------------+-----------+-------+\n",
       "| Chromosome   | Source         | Feature         | Start     | +9    |\n",
       "| (category)   | (object)       | (object)        | (int64)   | ...   |\n",
       "|--------------+----------------+-----------------+-----------+-------|\n",
       "| 1            | havana         | three_prime_UTR | 70008     | ...   |\n",
       "| 1            | havana_tagene  | three_prime_UTR | 944153    | ...   |\n",
       "| 1            | havana         | three_prime_UTR | 944153    | ...   |\n",
       "| 1            | havana         | three_prime_UTR | 944153    | ...   |\n",
       "| ...          | ...            | ...             | ...       | ...   |\n",
       "| Y            | ensembl_havana | three_prime_UTR | 24768898  | ...   |\n",
       "| Y            | havana         | three_prime_UTR | 25030900  | ...   |\n",
       "| Y            | havana         | three_prime_UTR | 25031316  | ...   |\n",
       "| Y            | havana         | three_prime_UTR | 25037991  | ...   |\n",
       "+--------------+----------------+-----------------+-----------+-------+\n",
       "Unstranded PyRanges object has 340,566 rows and 13 columns from 24 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "9 hidden columns: End, Strand, Parent_3utr, ID, Name, biotype, Parent, tag, ... (+ 1 more.)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "three_utrs_df_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ee4682",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a38826d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c1ed5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa670cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3f5b08",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
