{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9bfa1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, subprocess\n",
    "from pathlib import Path, PurePosixPath\n",
    "import gzip\n",
    "import json\n",
    "import pandas as pd, numpy as np, pyranges as pr\n",
    "import plotly.express as px\n",
    "import clickhouse_connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a644ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 500)\n",
    "pd.set_option('display.float_format', '{:.10f}'.format)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4374a7",
   "metadata": {},
   "source": [
    "# UTR3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b041df15",
   "metadata": {},
   "source": [
    "## Docs / notes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189c8ada",
   "metadata": {},
   "source": [
    "**Docs:**  \n",
    "* [PyRanges v1.x GitHub](https://github.com/pyranges/pyranges_1.x) / [PyRanges v1.x Docs](https://pyranges1.readthedocs.io/en/latest/index.html)\n",
    "\n",
    "**Notes:**  \n",
    " - !!! Нужно иметь ввиду, что позиции ФИЧ в GFF, например 3'-UTR, могут быть разделены на несколько интервалов, так как разделены, например, интронами.\n",
    "\n",
    "**Sources**\n",
    " - [miRBase](https://www.mirbase.org): hsa.gff - human\n",
    " - [multiMiR](http://multimir.org)\n",
    " - [TarBase](https://dianalab.e-ce.uth.gr/tarbasev9/downloads)\n",
    " - [miRTarBase](https://awi.cuhk.edu.cn/~miRTarBase/miRTarBase_2025/php/index.php)\n",
    " - [MiRanda](http://mirtoolsgallery.tech/mirtoolsgallery/node/1055): ineractions prediction\n",
    "\n",
    "**Tasks:**  \n",
    "1. Привести все к одной нумерации позиций: GFF - 1-based, BED - 0-based\n",
    "2. \n",
    "3. Проверить правильность offtarget_id. Сейчас это по факту позиции SV.\n",
    "\n",
    "**Commands:**  \n",
    "```{bash}\n",
    "mamba env export -n utr3.venv > environment.yml\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d6c257",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6c4f710",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_path = Path.cwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "30337bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create tree\n",
    "refs_dir = main_path / 'data/refs'\n",
    "gnomad_dir = main_path / 'data/gnomad'\n",
    "rcmg_btk_sv_dir = main_path / 'data/rcmg_btk_sv'\n",
    "clinvar_dir = main_path / 'data/clinvar'\n",
    "mirna_dir = main_path / 'data/mirna'\n",
    "mirbase_dir = mirna_dir / 'mirbase'\n",
    "tarbase_dir = mirna_dir / 'tarbase'\n",
    "domino_dir = main_path / 'data/domino'\n",
    "other_dir = main_path / 'data/other'\n",
    "output_dir = main_path / 'data/output'\n",
    "\n",
    "Path(refs_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(gnomad_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(rcmg_btk_sv_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(clinvar_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(mirbase_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(tarbase_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(domino_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(other_dir).mkdir(parents=True, exist_ok=True)\n",
    "Path(output_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1cbd55",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d3857b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_file(link, output_dir):\n",
    "    command = f'wget --no-clobber -P {output_dir} {link}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    filename = PurePosixPath(link).name\n",
    "    return output_dir / filename\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def gunzip_file(input_filepath):\n",
    "    output_filepath = input_filepath.with_suffix('')\n",
    "    command = f'gunzip -c {input_filepath} > {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def sort_gff(input_filepath):\n",
    "    output_filepath = input_filepath.with_stem(input_filepath.stem + \".sorted\")\n",
    "    command = f'igvtools sort {input_filepath} {output_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "    return output_filepath\n",
    "\n",
    "def index_gff(input_filepath):\n",
    "    command = f'igvtools index {input_filepath}'\n",
    "    subprocess.run(command, shell=True)\n",
    "\n",
    "def define_join_type(row):\n",
    "    feature_start = row['Start']\n",
    "    feature_end = row['End']\n",
    "    sv_start = row['Start_sv']\n",
    "    sv_end = row['End_sv']\n",
    "\n",
    "    join_type = np.nan\n",
    "    if feature_start <= sv_start and feature_end >= sv_end: # SV полностью в feature\n",
    "        join_type = 'sv_in_feature'\n",
    "    elif sv_start <= feature_start and sv_end >= feature_end: # feature полностью в sv\n",
    "        join_type = 'feature_in_sv'\n",
    "    elif sv_start < feature_start and sv_end <= sv_end:\n",
    "        join_type = 'sv_left_free'\n",
    "    elif sv_start > feature_start and sv_end >= sv_end:\n",
    "        join_type = 'sv_right_free'\n",
    "    elif sv_start == feature_start and sv_end == sv_end:\n",
    "        join_type = 'full_join'\n",
    "\n",
    "    return join_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5be17b6",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Fetch data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "ddc710c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.gff3.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.dna.toplevel.fa.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/gnomad/gnomad.v4.1.sv.sites.bed.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/clinvar/variant_summary.txt.gz' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/mirna/mirbase/hsa.gff3' already there; not retrieving.\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/mirna/tarbase/Homo_sapiens_TarBase-v9.tsv.gz' already there; not retrieving.\n",
      "\n",
      "--2026-02-04 17:11:25--  https://domino.iob.ch/score_all_final_19.02.19.txt\n",
      "Resolving domino.iob.ch (domino.iob.ch)... 2a00:1450:4010:c01::79, 74.125.131.121\n",
      "Connecting to domino.iob.ch (domino.iob.ch)|2a00:1450:4010:c01::79|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2379134 (2.3M) [text/plain]\n",
      "Saving to: '/Users/andrejnekrasov/pro/my/utr3/data/domino/score_all_final_19.02.19.txt'\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  2% 1.03M 2s\n",
      "    50K .......... .......... .......... .......... ..........  4% 2.28M 2s\n",
      "   100K .......... .......... .......... .......... ..........  6% 3.76M 1s\n",
      "   150K .......... .......... .......... .......... ..........  8% 4.40M 1s\n",
      "   200K .......... .......... .......... .......... .......... 10% 5.68M 1s\n",
      "   250K .......... .......... .......... .......... .......... 12% 9.40M 1s\n",
      "   300K .......... .......... .......... .......... .......... 15% 7.06M 1s\n",
      "   350K .......... .......... .......... .......... .......... 17% 7.67M 1s\n",
      "   400K .......... .......... .......... .......... .......... 19% 34.6M 1s\n",
      "   450K .......... .......... .......... .......... .......... 21% 8.12M 0s\n",
      "   500K .......... .......... .......... .......... .......... 23% 19.6M 0s\n",
      "   550K .......... .......... .......... .......... .......... 25% 26.1M 0s\n",
      "   600K .......... .......... .......... .......... .......... 27% 8.05M 0s\n",
      "   650K .......... .......... .......... .......... .......... 30% 12.1M 0s\n",
      "   700K .......... .......... .......... .......... .......... 32% 14.3M 0s\n",
      "   750K .......... .......... .......... .......... .......... 34% 26.1M 0s\n",
      "   800K .......... .......... .......... .......... .......... 36% 19.6M 0s\n",
      "   850K .......... .......... .......... .......... .......... 38% 35.6M 0s\n",
      "   900K .......... .......... .......... .......... .......... 40% 10.9M 0s\n",
      "   950K .......... .......... .......... .......... .......... 43% 9.05M 0s\n",
      "  1000K .......... .......... .......... .......... .......... 45% 5.91M 0s\n",
      "  1050K .......... .......... .......... .......... .......... 47% 26.0M 0s\n",
      "  1100K .......... .......... .......... .......... .......... 49% 46.7M 0s\n",
      "  1150K .......... .......... .......... .......... .......... 51% 9.84M 0s\n",
      "  1200K .......... .......... .......... .......... .......... 53% 9.52M 0s\n",
      "  1250K .......... .......... .......... .......... .......... 55% 9.37M 0s\n",
      "  1300K .......... .......... .......... .......... .......... 58% 57.0M 0s\n",
      "  1350K .......... .......... .......... .......... .......... 60% 24.7M 0s\n",
      "  1400K .......... .......... .......... .......... .......... 62% 35.6M 0s\n",
      "  1450K .......... .......... .......... .......... .......... 64% 15.1M 0s\n",
      "  1500K .......... .......... .......... .......... .......... 66% 24.2M 0s\n",
      "  1550K .......... .......... .......... .......... .......... 68% 54.0M 0s\n",
      "  1600K .......... .......... .......... .......... .......... 71% 13.2M 0s\n",
      "  1650K .......... .......... .......... .......... .......... 73% 89.6M 0s\n",
      "  1700K .......... .......... .......... .......... .......... 75% 16.1M 0s\n",
      "  1750K .......... .......... .......... .......... .......... 77% 7.50M 0s\n",
      "  1800K .......... .......... .......... .......... .......... 79% 23.3M 0s\n",
      "  1850K .......... .......... .......... .......... .......... 81% 73.2M 0s\n",
      "  1900K .......... .......... .......... .......... .......... 83% 10.6M 0s\n",
      "  1950K .......... .......... .......... .......... .......... 86% 27.6M 0s\n",
      "  2000K .......... .......... .......... .......... .......... 88% 17.4M 0s\n",
      "  2050K .......... .......... .......... .......... .......... 90% 15.0M 0s\n",
      "  2100K .......... .......... .......... .......... .......... 92% 15.8M 0s\n",
      "  2150K .......... .......... .......... .......... .......... 94% 12.4M 0s\n",
      "  2200K .......... .......... .......... .......... .......... 96% 15.9M 0s\n",
      "  2250K .......... .......... .......... .......... .......... 98% 21.7M 0s\n",
      "  2300K .......... .......... ...                             100%  304M=0.2s\n",
      "\n",
      "2026-02-04 17:11:26 (9.49 MB/s) - '/Users/andrejnekrasov/pro/my/utr3/data/domino/score_all_final_19.02.19.txt' saved [2379134/2379134]\n",
      "\n",
      "File '/Users/andrejnekrasov/pro/my/utr3/data/other/so.json' already there; not retrieving.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gff_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/gff3/homo_sapiens/Homo_sapiens.GRCh38.115.gff3.gz', refs_dir)\n",
    "fasta_filepath = fetch_file('https://ftp.ensembl.org/pub/release-115/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz', refs_dir)\n",
    "gnomad_sv_filepath = fetch_file('https://storage.googleapis.com/gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.bed.gz', gnomad_dir)\n",
    "clinvar_filepath = fetch_file('https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz', clinvar_dir)\n",
    "mirbase_filepath = fetch_file('https://www.mirbase.org/download/hsa.gff3', mirbase_dir)\n",
    "tarbase_filepath = fetch_file('https://dianalab.e-ce.uth.gr/tarbasev9/data/Homo_sapiens_TarBase-v9.tsv.gz', tarbase_dir)\n",
    "domino_filepath = fetch_file('https://domino.iob.ch/score_all_final_19.02.19.txt', domino_dir)\n",
    "\n",
    "so_terms_filepath = fetch_file('https://raw.githubusercontent.com/The-Sequence-Ontology/SO-Ontologies/refs/heads/master/Ontology_Files/so.json', other_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e76bd71b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'ConnectTimeoutError(<HTTPConnection(host='192.168.86.236', port=8123) at 0x16f3aa810>, 'Connection to 192.168.86.236 timed out. (connect timeout=10)')': /?\n",
      "WARNING:clickhouse_connect.driver.httpclient:Unexpected Http Driver Exception\n"
     ]
    }
   ],
   "source": [
    "# RCMG BTK SV\n",
    "chromosomes = list(range(1, 23)) + ['X', 'Y', 'M']\n",
    "rcmg_btk_sv_filepaths = []\n",
    "\n",
    "try:\n",
    "    client = clickhouse_connect.get_client(host='192.168.86.236', user='rdr', password='Dru_Todd3')\n",
    "    db_name = \"ngs\"\n",
    "    client.database = db_name\n",
    "except:\n",
    "    for chrom in chromosomes:\n",
    "        output_filename =  f'rcmg_btk_sv_chr{chrom}.csv.gz'\n",
    "        output_filepath = rcmg_btk_sv_dir / output_filename\n",
    "        rcmg_btk_sv_filepaths.append(output_filepath)\n",
    "\n",
    "        if not output_filepath.exists():\n",
    "            rcmg_bkt_sv_df = client.query_df(f\"SELECT * from ngs.GENOME_hg38_2_sv_chr{chrom} WHERE SVTYPE='DEL';\")\n",
    "\n",
    "            rcmg_bkt_sv_df['Chromosome'] = str(chrom)\n",
    "            rcmg_bkt_sv_df.to_csv(output_filepath, index=False, compression='gzip')\n",
    "    try:\n",
    "        client.close()\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61697b2",
   "metadata": {},
   "source": [
    "## <span style=\"color:#00ff00;\">Main</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20681675",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sorting /Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.gff3  -> /Users/andrejnekrasov/pro/my/utr3/data/refs/Homo_sapiens.GRCh38.115.sorted.gff3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: A restricted method in java.lang.System has been called\n",
      "WARNING: java.lang.System::load has been called by org.xerial.snappy.SnappyLoader in module snappy.java (file:/Users/andrejnekrasov/mambaforge/envs/utr3.venv/share/igvtools-2.17.3-0/lib/snappy-java-1.1.10.5.jar)\n",
      "WARNING: Use --enable-native-access=snappy.java to avoid a warning for callers in this module\n",
      "WARNING: Restricted methods will be blocked in a future release unless native access is enabled\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "gff_gunzipped_filepath = gunzip_file(gff_filepath)\n",
    "gff_sorted_filepath = sort_gff(gff_gunzipped_filepath)\n",
    "index_gff(gff_sorted_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b315f3",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Prepare data</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "007545ae",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Prepare GFF3 data</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4515e81b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          biotype  count\n",
       "0  protein_coding  19757"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GFF\n",
    "gff_pr = pr.read_gff3(str(gff_filepath))\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Add additional metrics\n",
    "gff_pr['feature_len'] = (gff_pr['End'] - gff_pr['Start']).astype('int')\n",
    "gff_pr['feature_id'] = 'feature_' + gff_pr['Chromosome'].astype('str') + ':' + gff_pr['Start'].astype('str') + '-' + gff_pr['End'].astype('str')\n",
    "\n",
    "# Self annotation of GFF with mRNA (parent) data\n",
    "target_columns = ['ID', 'Name', 'biotype', 'Parent', 'tag', 'transcript_support_level', 'feature_len', 'feature_id'] # Replace the data in these empty columns with mRNA data\n",
    "annotation_gff_pr = gff_pr[target_columns].dropna(subset='ID')\n",
    "\n",
    "annotation_suffix = '_annotation'\n",
    "annotated_gff_pr = gff_pr.merge(annotation_gff_pr, how='left', left_on='Parent', right_on='ID', suffixes=['', annotation_suffix])\n",
    "\n",
    "annotated_columns = [f'{column}{annotation_suffix}' for column in target_columns]\n",
    "fill_dict = dict(zip(target_columns, annotated_columns))\n",
    "for target, annotation in fill_dict.items():\n",
    "    annotated_gff_pr[target] = annotated_gff_pr[target].fillna(annotated_gff_pr[annotation])\n",
    "annotated_gff_pr = annotated_gff_pr.drop(annotated_columns, axis=1)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Create specific PRs (mRNAm, 3'-UTR, offtargets)\n",
    "mrna_pr = annotated_gff_pr.query('Feature == \"mRNA\"')\n",
    "three_utrs_pr = annotated_gff_pr.query('Feature == \"three_prime_UTR\"')\n",
    "\n",
    "offtarget_features = ['five_prime_UTR',\n",
    "                               'CDS',\n",
    "                               #'lnc_RNA',\n",
    "                               'ncRNA_gene',\n",
    "                               #'snRNA',\n",
    "                               #'snoRNA',\n",
    "                               #'scRNA',\n",
    "                               'rRNA',\n",
    "                               'tRNA',\n",
    "                               'processed_transcript',\n",
    "                               ]\n",
    "offtarget_features_pr = gff_pr.query('Feature in @offtarget_features')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# 3'-UTR biotype stats\n",
    "# Считаем именно уникальные вхождения, так как один 3'-UTR может быть разбить на несколько интервалов в GFF (разделен интронами).\n",
    "# В особенности это выражено у nonsense_mediated_decay транскриптов (нам они не пригодятся, но тем не менее).\n",
    "three_utrs_pr.drop_duplicates(subset='ID')['biotype'].value_counts().to_frame().reset_index()\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Keep only MANE_select / MANE_Select|Ensembl_canonical\n",
    "three_utrs_pr_filtered = three_utrs_pr[three_utrs_pr['tag'].str.contains('MANE_Select', na=False, regex=True)]\n",
    "\n",
    "# Stats\n",
    "# MANE_select genes count = 19437 (release_1.5) from summary file: https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.5/\n",
    "three_utrs_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ad65e24",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">gnomAD SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7a665c68",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/3419294281.py:10: DtypeWarning: Columns (10,11,24,25,26,28,29,30,31,32,33,36,37,38,39,40,42,57,58,59) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=target_columns_indexes, names=target_columns, comment='#')\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/3419294281.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/3419294281.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sv_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>627321.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean</td>\n",
       "      <td>8529.8884988706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>std</td>\n",
       "      <td>139720.7127188754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>min</td>\n",
       "      <td>51.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10%</td>\n",
       "      <td>76.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>25%</td>\n",
       "      <td>137.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>50%</td>\n",
       "      <td>702.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>75%</td>\n",
       "      <td>4073.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>90%</td>\n",
       "      <td>10309.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>95%</td>\n",
       "      <td>23359.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>99%</td>\n",
       "      <td>142071.0000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>99.9%</td>\n",
       "      <td>568683.9200000279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>max</td>\n",
       "      <td>75976919.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index              sv_len\n",
       "0   count   627321.0000000000\n",
       "1    mean     8529.8884988706\n",
       "2     std   139720.7127188754\n",
       "3     min       51.0000000000\n",
       "4     10%       76.0000000000\n",
       "5     25%      137.0000000000\n",
       "6     50%      702.0000000000\n",
       "7     75%     4073.0000000000\n",
       "8     90%    10309.0000000000\n",
       "9     95%    23359.0000000000\n",
       "10    99%   142071.0000000005\n",
       "11  99.9%   568683.9200000279\n",
       "12    max 75976919.0000000000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gnomAD SV\n",
    "# Keep only the necessary columns (63 of 600...) in SV\n",
    "with gzip.open(str(gnomad_sv_filepath), 'rt') as sv_file:\n",
    "    sv_header = sv_file.readline().strip().split('\\t')\n",
    "\n",
    "target_columns = ['#chrom', 'start', 'end', 'name', 'svtype', 'samples', 'MULTIALLELIC', 'ALGORITHMS', 'BOTHSIDES_SUPPORT', 'CHR2', 'CPX_INTERVALS', 'CPX_TYPE', 'END', 'END2', 'EVIDENCE', 'LOW_CONFIDENCE_REPETITIVE_LARGE_DUP', 'MEMBERS', 'NCR', 'OUTLIER_SAMPLE_ENRICHED_LENIENT', 'PAR', 'PCRMINUS_NCR', 'PCRPLUS_NCR', 'PESR_GT_OVERDISPERSION', 'POS2', 'PREDICTED_BREAKEND_EXONIC', 'PREDICTED_COPY_GAIN', 'PREDICTED_DUP_PARTIAL', 'PREDICTED_INTERGENIC', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_INTRONIC', 'PREDICTED_INV_SPAN', 'PREDICTED_LOF', 'PREDICTED_MSV_EXON_OVERLAP', 'PREDICTED_NEAREST_TSS', 'PREDICTED_NONCODING_BREAKPOINT', 'PREDICTED_NONCODING_SPAN', 'PREDICTED_PARTIAL_DISPERSED_DUP', 'PREDICTED_PARTIAL_EXON_DUP', 'PREDICTED_PROMOTER', 'PREDICTED_TSS_DUP', 'PREDICTED_UTR', 'RESOLVED_POSTHOC', 'SOURCE', 'SVLEN', 'SVTYPE', 'UNRESOLVED_TYPE', 'AN', 'AC', 'AF', 'N_BI_GENOS', 'N_HOMREF', 'N_HET', 'N_HOMALT', 'FREQ_HOMREF', 'FREQ_HET', 'FREQ_HOMALT', 'CN_NUMBER', 'CN_COUNT', 'CN_STATUS', 'CN_FREQ', 'CN_NONREF_COUNT', 'CN_NONREF_FREQ', 'FILTER']\n",
    "target_columns_indexes = [sv_header.index(i) for i in target_columns]\n",
    "target_columns[0:3] = ['Chromosome', 'Start', 'End'] # Rename columns\n",
    "\n",
    "sv_df = pd.read_csv(str(gnomad_sv_filepath), sep='\\t', usecols=target_columns_indexes, names=target_columns, comment='#')\n",
    "sv_df['Chromosome'] = sv_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "sv_pr = pr.PyRanges(sv_df)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Keep target SVs (DEL)\n",
    "sv_targets = ['DEL']\n",
    "sv_filter = ['PASS']\n",
    "target_sv_pr = sv_pr.query('svtype in @sv_targets and FILTER in @sv_filter')\n",
    "\n",
    "# Add columns\n",
    "target_sv_pr['sv_len'] = target_sv_pr.End - target_sv_pr.Start\n",
    "target_sv_pr['sv_id'] = 'sv_' + target_sv_pr['Chromosome'].astype('str') + ':' + target_sv_pr['Start'].astype('str') + '-' + target_sv_pr['End'].astype('str')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# SV percentiles by AF\n",
    "target_sv_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()\n",
    "\n",
    "# SV percentiles by SV length\n",
    "target_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb02969",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">ClinVar</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "381eb36e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/1514172660.py:2: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinvar_df = pd.read_csv(clinvar_filepath, sep='\\t', compression='gzip')\n"
     ]
    }
   ],
   "source": [
    "# ClinVar\n",
    "clinvar_df = pd.read_csv(clinvar_filepath, sep='\\t', compression='gzip')\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "clinvar_df = clinvar_df.replace(['na', '-'], np.nan)\n",
    "clinvar_df = clinvar_df.rename(columns={'Stop': 'End'})\n",
    "clinvar_df_filtered = clinvar_df.query('Assembly == \"GRCh38\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "693042b0",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">RCMG BTK SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "id": "84b6188b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/4041134283.py:39: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_filepath), compression='gzip')\n"
     ]
    }
   ],
   "source": [
    "# RCMG BTK SV\n",
    "rcmg_btk_sv_filepath = rcmg_btk_sv_dir / 'rcmg_btk_sv.csv.gz'\n",
    "\n",
    "if not rcmg_btk_sv_filepath.exists():\n",
    "    rcmg_bkt_sv_df = pd.DataFrame() # concated filtered aggregated\n",
    "    for rcmg_btk_sv_file in rcmg_btk_sv_filepaths:\n",
    "        print(rcmg_btk_sv_file)\n",
    "        current_rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_file), compression='gzip')\n",
    "\n",
    "        # Add columns\n",
    "        current_rcmg_bkt_sv_df['AF'] = current_rcmg_bkt_sv_df['AD'] / current_rcmg_bkt_sv_df['DP']\n",
    "        current_rcmg_bkt_sv_df['sv_len'] = abs(current_rcmg_bkt_sv_df['END'] - current_rcmg_bkt_sv_df['POS'])\n",
    "        current_rcmg_bkt_sv_df['sv_id'] = 'sv_' + current_rcmg_bkt_sv_df['Chromosome'].astype('str') + ':' + current_rcmg_bkt_sv_df['POS'].astype('str') + '-' + current_rcmg_bkt_sv_df['END'].astype('str')\n",
    "\n",
    "        # Drop columns\n",
    "        current_rcmg_bkt_sv_df = current_rcmg_bkt_sv_df.drop(columns=['REF', 'ALT'])\n",
    "\n",
    "        # Filter\n",
    "        current_rcmg_bkt_sv_df_filtered = current_rcmg_bkt_sv_df.query('caller == \"manta\"')\n",
    "\n",
    "        # Aggregate\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered.groupby(['Chromosome', 'POS', 'END', 'SVTYPE', 'caller', 'sv_len', 'sv_id'], as_index=False).agg({\n",
    "            'SID': [list, 'count'],\n",
    "            'DP': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AD': [list, 'min', 'median', 'mean', 'max'],\n",
    "            'AF': [list, 'min', 'median', 'mean', 'max']\n",
    "        })\n",
    "\n",
    "        # Rename columns\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated.columns = ['_'.join(col).strip('_') for col in current_rcmg_bkt_sv_df_filtered_aggregated.columns]\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.rename(columns={'POS': 'Start', 'END': 'End'})\n",
    "        current_rcmg_bkt_sv_df_filtered_aggregated = current_rcmg_bkt_sv_df_filtered_aggregated.reset_index(drop=True)\n",
    "\n",
    "        # Concat\n",
    "        rcmg_bkt_sv_df['Chromosome'] = rcmg_bkt_sv_df['Chromosome'].astype('str')\n",
    "        rcmg_bkt_sv_df = pd.concat([rcmg_bkt_sv_df, current_rcmg_bkt_sv_df_filtered_aggregated], ignore_index=False)\n",
    "    rcmg_bkt_sv_df.to_csv(str(rcmg_btk_sv_filepath), index=False, compression='gzip')\n",
    "else:\n",
    "    rcmg_bkt_sv_df = pd.read_csv(str(rcmg_btk_sv_filepath), compression='gzip')\n",
    "    rcmg_bkt_sv_df['Chromosome'] = rcmg_bkt_sv_df['Chromosome'].astype('str')\n",
    "\n",
    "rcmg_btk_sv_pr = pr.PyRanges(rcmg_bkt_sv_df)\n",
    "rcmg_btk_sv_pr_filtered = rcmg_btk_sv_pr.query('SID_count < 30')\n",
    "rcmg_btk_sv_pr_filtered = rcmg_btk_sv_pr_filtered.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "1dc5b242",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     3810648.0000000000\n",
       "mean       910374.0638663031\n",
       "std       7319054.6422773376\n",
       "min            49.0000000000\n",
       "10%           631.0000000000\n",
       "25%          2244.0000000000\n",
       "50%         23863.5000000000\n",
       "75%        256919.2500000000\n",
       "90%       1169973.3000000003\n",
       "95%       2045701.0000000000\n",
       "99%      20410049.0000000000\n",
       "99.9%    66126493.7060000002\n",
       "max     242323073.0000000000\n",
       "Name: sv_len, dtype: float64"
      ]
     },
     "execution_count": 380,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rcmg_btk_sv_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d32aa6",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">miRBase</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "id": "9dfaa41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# miRBase\n",
    "mirbase_pr = pr.read_gff3(str(mirbase_filepath))\n",
    "mirbase_pr['Chromosome'] = mirbase_pr['Chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "718e8ef5",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">TarBase</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "id": "7a86eb49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/1318660838.py:2: DtypeWarning: Columns (5,14,15,19,21) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  tarbase_df = pd.read_csv(str(tarbase_filepath), sep='\\t', compression='gzip')\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/1318660838.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tarbase_pr[['Start', 'End']] = tarbase_pr[['Start', 'End']].astype('int64')\n"
     ]
    }
   ],
   "source": [
    "# TarBase\n",
    "tarbase_df = pd.read_csv(str(tarbase_filepath), sep='\\t', compression='gzip')\n",
    "tarbase_df = tarbase_df.rename(columns={'chromosome': 'Chromosome', 'start': 'Start', 'end': 'End', 'strand': 'Strand'})\n",
    "tarbase_df['Chromosome'] = tarbase_df['Chromosome'].str.replace('chr', '')\n",
    "\n",
    "tarbase_pr = pr.PyRanges(tarbase_df)\n",
    "tarbase_pr = tarbase_pr.dropna(subset=['Chromosome', 'Start', 'End'], how='any')\n",
    "tarbase_pr[['Start', 'End']] = tarbase_pr[['Start', 'End']].astype('int64')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f18e8a51",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Sequence Ontology terms</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "d0a4271f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SO terms\n",
    "with open(so_terms_filepath, 'r') as so_terms_file:\n",
    "    so_terms = json.load(so_terms_file)\n",
    "\n",
    "so_terms_filtered = {\n",
    "    'lbl': [],\n",
    "    'definition': []\n",
    "}\n",
    "for i in so_terms['graphs'][0]['nodes']:\n",
    "    lbl = i.get('lbl', np.nan)\n",
    "    definition = i.get('meta', {}).get('definition', {}).get('val', np.nan)\n",
    "    so_terms_filtered['lbl'].append(lbl)\n",
    "    so_terms_filtered['definition'].append(definition)\n",
    "\n",
    "so_terms_df = pd.DataFrame.from_dict(so_terms_filtered)\n",
    "so_terms_df['lbl'] = so_terms_df['lbl'].str.capitalize().replace('_', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b001f54",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">EDA (Exploratory Data Analysis)</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "id": "788e1cc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF features stats\n",
    "gff_features = gff_pr.Feature.value_counts().to_frame().reset_index()\n",
    "gff_features['tmp_id'] = gff_features['Feature'].str.capitalize().replace('_', '')\n",
    "gff_features = gff_features.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_features = gff_features.drop(['lbl', 'tmp_id'], axis=1)\n",
    "#pd.DataFrame(gff_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "id": "a9e7f9eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GFF biotypes stats\n",
    "gff_biotypes = gff_pr.biotype.value_counts().to_frame().reset_index()\n",
    "gff_biotypes['tmp_id'] = gff_biotypes['biotype'].str.capitalize().replace('_', '')\n",
    "gff_biotypes = gff_biotypes.merge(so_terms_df, how='left', left_on='tmp_id', right_on='lbl')\n",
    "gff_biotypes = gff_biotypes.drop(['lbl', 'tmp_id'], axis=1)\n",
    "#pd.DataFrame(gff_biotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "id": "8cfc34fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sv_pr.svtype.value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "id": "3a2175ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_location\n",
       "CDS     2502793\n",
       "3UTR    2056551\n",
       "5UTR        175\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 387,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tarbase_pr['gene_location'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e0ca86",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Join GFF3 x SV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64921092",
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mПри выполнении кода в текущей ячейке или предыдущей ячейке ядро аварийно завершило работу. \n",
      "\u001b[1;31mПроверьте код в ячейках, чтобы определить возможную причину сбоя. \n",
      "\u001b[1;31mЩелкните <a href='https://aka.ms/vscodeJupyterKernelCrash'>здесь</a>, чтобы получить дополнительные сведения. \n",
      "\u001b[1;31mПодробнее см. в <a href='command:jupyter.viewOutput'>журнале Jupyter</a>."
     ]
    }
   ],
   "source": [
    "sv_for_join_pr = rcmg_btk_sv_pr_filtered # target_sv_pr / rcmg_btk_sv_pr_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b14ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join GFF x SV\n",
    "features_sv_joined_pr = three_utrs_pr_filtered.join_overlaps(sv_for_join_pr, suffix='_sv', report_overlap_column='join_len', strand_behavior='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6885c5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add additional columns (join_start, join_end, join_id)\n",
    "features_sv_joined_pr['join_start'] = np.maximum(features_sv_joined_pr['Start'], features_sv_joined_pr['Start_sv'])\n",
    "features_sv_joined_pr['join_end'] = np.minimum(features_sv_joined_pr['End'], features_sv_joined_pr['End_sv'])\n",
    "features_sv_joined_pr['join_id'] = 'join_' + features_sv_joined_pr['Chromosome'].astype('str') + ':' + features_sv_joined_pr['join_start'].astype('str') + '-' + features_sv_joined_pr['join_end'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ec54b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>join_type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>feature_in_sv</td>\n",
       "      <td>536683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>14756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sv_left_free</td>\n",
       "      <td>3114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       join_type   count\n",
       "0  feature_in_sv  536683\n",
       "1  sv_right_free   14756\n",
       "2   sv_left_free    3114\n",
       "3  sv_in_feature    1686"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define joint type\n",
    "features_sv_joined_pr['join_type'] = features_sv_joined_pr.apply(define_join_type, axis=1)\n",
    "features_sv_joined_pr['join_type'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ae52f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count joints by SV\n",
    "joints_count_by_sv = features_sv_joined_pr['sv_id'].value_counts().to_frame().reset_index()\n",
    "joints_count_by_sv = joints_count_by_sv.rename(columns={'count': 'joins_count'})\n",
    "\n",
    "# Add joints count data\n",
    "features_sv_joined_pr = features_sv_joined_pr.merge(joints_count_by_sv, how='left')\n",
    "#len(features_sv_joined_pr)\n",
    "\n",
    "# Filter joins by join type\n",
    "targey_joint_types = ['sv_in_feature', 'sv_right_free', 'sv_left_free']\n",
    "features_sv_joined_pr_filtered = features_sv_joined_pr.query('join_type in @targey_joint_types and ((Strand == \"+\" and join_type == \"sv_right_free\") or (Strand == \"-\" and join_type == \"sv_left_free\") or join_type == \"sv_in_feature\") and joins_count == 1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b449576",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join type stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a69d46c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>join_type</th>\n",
       "      <th>sv_in_feature</th>\n",
       "      <th>sv_left_free</th>\n",
       "      <th>sv_right_free</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1668.0000000000</td>\n",
       "      <td>360.0000000000</td>\n",
       "      <td>766.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>375.0821342926</td>\n",
       "      <td>5948.2527777778</td>\n",
       "      <td>5229.9751958225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>826.7939945275</td>\n",
       "      <td>13683.9939506017</td>\n",
       "      <td>7526.0590659882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>50.0000000000</td>\n",
       "      <td>77.0000000000</td>\n",
       "      <td>60.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10%</th>\n",
       "      <td>55.0000000000</td>\n",
       "      <td>1064.5000000000</td>\n",
       "      <td>1823.5000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>65.0000000000</td>\n",
       "      <td>1706.7500000000</td>\n",
       "      <td>4495.2500000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>106.5000000000</td>\n",
       "      <td>4332.0000000000</td>\n",
       "      <td>4567.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>289.0000000000</td>\n",
       "      <td>4964.5000000000</td>\n",
       "      <td>4620.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90%</th>\n",
       "      <td>826.3000000000</td>\n",
       "      <td>9786.3000000000</td>\n",
       "      <td>6082.0000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95%</th>\n",
       "      <td>1721.4500000000</td>\n",
       "      <td>14858.4000000000</td>\n",
       "      <td>9293.7500000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99%</th>\n",
       "      <td>3746.3000000000</td>\n",
       "      <td>39100.8000000000</td>\n",
       "      <td>36879.5500000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99.9%</th>\n",
       "      <td>8462.0100000001</td>\n",
       "      <td>175049.1620000026</td>\n",
       "      <td>91751.6250000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>12047.0000000000</td>\n",
       "      <td>221964.0000000000</td>\n",
       "      <td>135414.0000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "join_type    sv_in_feature      sv_left_free     sv_right_free\n",
       "count      1668.0000000000    360.0000000000    766.0000000000\n",
       "mean        375.0821342926   5948.2527777778   5229.9751958225\n",
       "std         826.7939945275  13683.9939506017   7526.0590659882\n",
       "min          50.0000000000     77.0000000000     60.0000000000\n",
       "10%          55.0000000000   1064.5000000000   1823.5000000000\n",
       "25%          65.0000000000   1706.7500000000   4495.2500000000\n",
       "50%         106.5000000000   4332.0000000000   4567.0000000000\n",
       "75%         289.0000000000   4964.5000000000   4620.0000000000\n",
       "90%         826.3000000000   9786.3000000000   6082.0000000000\n",
       "95%        1721.4500000000  14858.4000000000   9293.7500000000\n",
       "99%        3746.3000000000  39100.8000000000  36879.5500000000\n",
       "99.9%      8462.0100000001 175049.1620000026  91751.6250000073\n",
       "max       12047.0000000000 221964.0000000000 135414.0000000000"
      ]
     },
     "execution_count": 357,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stats by join type\n",
    "features_sv_joined_pr_filtered.groupby('join_type')['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc037975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2794</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          biotype  count\n",
       "0  protein_coding   2794"
      ]
     },
     "execution_count": 358,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_joined_pr_filtered['biotype'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4135ee",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">Offtarget features exclusion</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ce637a",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">features_sv_joined_pr_filtered copy</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab7f6e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Offtarget stats\n",
    "features_sv_joined_pr_filtered_cp = features_sv_joined_pr_filtered.copy()\n",
    "features_sv_joined_pr_filtered_cp[['Start_tmp', 'End_tmp']] = features_sv_joined_pr_filtered_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_joined_pr_filtered_cp[['Start', 'End']] = features_sv_joined_pr_filtered_cp[['Start_sv', 'End_sv']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157fc717",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Filter offtargets</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dc11ee7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2299"
      ]
     },
     "execution_count": 360,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Exclude offtargets (SV x offtarget overlaps)\n",
    "features_sv_no_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.overlap(offtarget_features_pr, invert=True, strand_behavior='ignore')\n",
    "\n",
    "features_sv_no_offtereget_joined_pr[['Start', 'End']] = features_sv_no_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_joined_pr = features_sv_no_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "len(features_sv_no_offtereget_joined_pr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "623083fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before offtargets filtering\n",
      "       join_type  count\n",
      "0  sv_in_feature   1668\n",
      "1  sv_right_free    766\n",
      "2   sv_left_free    360\n",
      "\n",
      "After offtargets filtering\n",
      "       join_type  count\n",
      "0  sv_in_feature   1352\n",
      "1  sv_right_free    660\n",
      "2   sv_left_free    287\n"
     ]
    }
   ],
   "source": [
    "# Joint type stats after offtargets exclution\n",
    "print('Before offtargets filtering')\n",
    "print(features_sv_joined_pr_filtered['join_type'].value_counts().to_frame().reset_index())\n",
    "print('\\nAfter offtargets filtering')\n",
    "print(features_sv_no_offtereget_joined_pr['join_type'].value_counts().to_frame().reset_index())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ddbe504",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Offtarget stats</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd478da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# features_sv_offtereget_joined_pr - только для отслеживания пересечений с offtarget. Фильтрованный от offtarget pr - features_sv_no_offtereget_joined_pr (через overlap, потому что join находя пересечения с offtarget (через left) также оставляет оригинальную строку).DS_Storefeatures_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len')\n",
    "features_sv_offtereget_joined_pr = features_sv_joined_pr_filtered_cp.join_overlaps(offtarget_features_pr, join_type='left', suffix='_offtarget', report_overlap_column='offtarget_join_len', strand_behavior='ignore')\n",
    "\n",
    "features_sv_offtereget_joined_pr[['Start', 'End']] = features_sv_offtereget_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_offtereget_joined_pr = features_sv_offtereget_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_offtereget_joined_pr['offtarget_id'] = 'offtarget_' + features_sv_offtereget_joined_pr['Chromosome'].astype('str') + ':' + features_sv_offtereget_joined_pr['Start_offtarget'].astype('str') + '-' + features_sv_offtereget_joined_pr['End_offtarget'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c516d79f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>join_type</th>\n",
       "      <th>Feature_offtarget</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">sv_in_feature</th>\n",
       "      <th>CDS</th>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">sv_left_free</th>\n",
       "      <th>CDS</th>\n",
       "      <td>688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">sv_right_free</th>\n",
       "      <th>CDS</th>\n",
       "      <td>607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA_gene</th>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 count\n",
       "join_type     Feature_offtarget       \n",
       "sv_in_feature CDS                  129\n",
       "              five_prime_UTR        27\n",
       "              ncRNA_gene           300\n",
       "sv_left_free  CDS                  688\n",
       "              five_prime_UTR       213\n",
       "              ncRNA_gene            71\n",
       "sv_right_free CDS                  607\n",
       "              five_prime_UTR       161\n",
       "              ncRNA_gene           113"
      ]
     },
     "execution_count": 363,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Features offtargets stats\n",
    "# Суммарные значения данной статистики будут выше чем разница всех джоинов - оффтартет джоинов, так как один и тот же SV (напр. DEL) могут пересекать разные изоформы одного и того же транскрипта.\n",
    "features_sv_offtereget_joined_pr.groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])\n",
    "#features_sv_offtereget_joined_pr.drop_duplicates(subset='sv_id').groupby(['join_type', 'Feature_offtarget'])['Feature_offtarget'].agg(['count'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d8bf654",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">miRNA join</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1e4a88",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">features_sv_no_offtereget_joined_pr copy</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e88e7aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Offtarget stats\n",
    "features_sv_no_offtereget_joined_pr_cp = features_sv_no_offtereget_joined_pr.copy()\n",
    "features_sv_no_offtereget_joined_pr_cp[['Start_tmp', 'End_tmp']] = features_sv_no_offtereget_joined_pr_cp[['Start', 'End']] # Save old positions\n",
    "features_sv_no_offtereget_joined_pr_cp[['Start', 'End']] = features_sv_no_offtereget_joined_pr_cp[['join_start', 'join_end']] # SV positions to main positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99af8a69",
   "metadata": {},
   "source": [
    "#### <span style=\"color:green;\">Join</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c871563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join TarBase\n",
    "features_sv_no_offtereget_tarbase_joined_pr = features_sv_no_offtereget_joined_pr_cp.join_overlaps(tarbase_pr, suffix='_tarbase', strand_behavior='ignore', report_overlap_column='tarbase_join_len')\n",
    "\n",
    "features_sv_no_offtereget_tarbase_joined_pr[['Start', 'End']] = features_sv_no_offtereget_tarbase_joined_pr[['Start_tmp', 'End_tmp']]\n",
    "features_sv_no_offtereget_tarbase_joined_pr = features_sv_no_offtereget_tarbase_joined_pr.drop(['Start_tmp', 'End_tmp'], axis=1)\n",
    "\n",
    "features_sv_no_offtereget_tarbase_joined_pr['tarbase_id'] = 'tarbase_' + features_sv_no_offtereget_tarbase_joined_pr['Chromosome'].astype('str') + ':' + features_sv_no_offtereget_tarbase_joined_pr['Start_tarbase'].astype('str') + '-' + features_sv_no_offtereget_tarbase_joined_pr['End_tarbase'].astype('str') + ':' + features_sv_no_offtereget_tarbase_joined_pr['mirna_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6984057b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#features_sv_no_offtereget_tarbase_joined_pr['AF'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "653cb4b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_location\n",
       "3UTR    22692\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 367,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['gene_location'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae28975b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "interaction_group\n",
       "primary      16773\n",
       "secondary     5919\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 368,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['interaction_group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e230a118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "regulation\n",
       "Negative    22692\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 369,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['regulation'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "084ddca5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LNPEP-201</td>\n",
       "      <td>974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARPP19-208</td>\n",
       "      <td>795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SDE2-201</td>\n",
       "      <td>646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LDLR-208</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GNS-201</td>\n",
       "      <td>511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>COMMD9-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>NHLRC1-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>NATD1-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td>FLRT2-201</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>OXA1L-212</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>444 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    transcript_name  count\n",
       "0         LNPEP-201    974\n",
       "1        ARPP19-208    795\n",
       "2          SDE2-201    646\n",
       "3          LDLR-208    525\n",
       "4           GNS-201    511\n",
       "..              ...    ...\n",
       "439      COMMD9-201      1\n",
       "440      NHLRC1-201      1\n",
       "441       NATD1-201      1\n",
       "442       FLRT2-201      1\n",
       "443       OXA1L-212      1\n",
       "\n",
       "[444 rows x 2 columns]"
      ]
     },
     "execution_count": 370,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['transcript_name'].value_counts().to_frame().reset_index().to_csv(output_dir / 'mirna_genes.csv', index=False)\n",
    "features_sv_no_offtereget_tarbase_joined_pr['transcript_name'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eaea44e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "experimental_method\n",
       "HITS-CLIP                    13805\n",
       "PAR-CLIP                      8680\n",
       "Chimeric fragments             103\n",
       "qCLASH                          51\n",
       "Luciferase Reporter Assay       41\n",
       "CLASH                            9\n",
       "IMPACT-Seq                       2\n",
       "3LIFE                            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 371,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['experimental_method'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53825b57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "join_type\n",
       "sv_in_feature    10776\n",
       "sv_left_free      6169\n",
       "sv_right_free     5747\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 372,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['join_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e67fb804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "join_id\n",
       "join_5:97031222-97037155       974\n",
       "join_15:52547044-52550754      795\n",
       "join_19:11132469-11133820      720\n",
       "join_12:64713448-64715411      511\n",
       "join_1:179103702-179105699     445\n",
       "                              ... \n",
       "join_13:112885389-112885440      1\n",
       "join_17:81912306-81912359        1\n",
       "join_17:77215831-77216366        1\n",
       "join_17:77215830-77216398        1\n",
       "join_17:77215801-77216351        1\n",
       "Name: count, Length: 774, dtype: int64"
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['join_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f2cecd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome\n",
       "1     2807\n",
       "19    2102\n",
       "5     2007\n",
       "12    1933\n",
       "15    1434\n",
       "8     1392\n",
       "3     1352\n",
       "10    1194\n",
       "4     1079\n",
       "2     1053\n",
       "17     909\n",
       "6      901\n",
       "14     760\n",
       "22     664\n",
       "16     546\n",
       "11     469\n",
       "18     394\n",
       "7      393\n",
       "20     368\n",
       "13     311\n",
       "9      290\n",
       "21     177\n",
       "X      157\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 374,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['Chromosome'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e3fc4b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count   22692.0000000000\n",
       "mean     3803.5000881368\n",
       "std      4465.6255417306\n",
       "min        50.0000000000\n",
       "10%       226.0000000000\n",
       "25%       948.0000000000\n",
       "50%      2567.0000000000\n",
       "75%      5933.0000000000\n",
       "90%      8139.0000000000\n",
       "95%     10396.0000000000\n",
       "99%     15269.0000000000\n",
       "99.9%   78230.0000000000\n",
       "max     78230.0000000000\n",
       "Name: sv_len, dtype: float64"
      ]
     },
     "execution_count": 375,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_sv_no_offtereget_tarbase_joined_pr['sv_len'].describe(percentiles=[.1, .25, .5, .75, .9, .95, .99, .999]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9903c45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>Alias</th>\n",
       "      <th>external_name</th>\n",
       "      <th>logic_name</th>\n",
       "      <th>Name</th>\n",
       "      <th>biotype</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>version</th>\n",
       "      <th>Parent</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>constitutive</th>\n",
       "      <th>ensembl_end_phase</th>\n",
       "      <th>ensembl_phase</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>Start_sv</th>\n",
       "      <th>End_sv</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>caller</th>\n",
       "      <th>sv_len</th>\n",
       "      <th>sv_id</th>\n",
       "      <th>SID_list</th>\n",
       "      <th>SID_count</th>\n",
       "      <th>DP_list</th>\n",
       "      <th>DP_min</th>\n",
       "      <th>DP_median</th>\n",
       "      <th>DP_mean</th>\n",
       "      <th>DP_max</th>\n",
       "      <th>AD_list</th>\n",
       "      <th>AD_min</th>\n",
       "      <th>AD_median</th>\n",
       "      <th>AD_mean</th>\n",
       "      <th>AD_max</th>\n",
       "      <th>AF_list</th>\n",
       "      <th>AF_min</th>\n",
       "      <th>AF_median</th>\n",
       "      <th>AF_mean</th>\n",
       "      <th>AF_max</th>\n",
       "      <th>join_len</th>\n",
       "      <th>join_start</th>\n",
       "      <th>join_end</th>\n",
       "      <th>join_id</th>\n",
       "      <th>join_type</th>\n",
       "      <th>joins_count</th>\n",
       "      <th>species</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id_tarbase</th>\n",
       "      <th>gene_location</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_id_tarbase</th>\n",
       "      <th>Start_tarbase</th>\n",
       "      <th>End_tarbase</th>\n",
       "      <th>Strand_tarbase</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>regulation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>article_pubmed_id</th>\n",
       "      <th>confidence</th>\n",
       "      <th>interaction_group</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>microt_score</th>\n",
       "      <th>comment</th>\n",
       "      <th>tarbase_join_len</th>\n",
       "      <th>tarbase_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1327763</td>\n",
       "      <td>1328896</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 8)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1133</td>\n",
       "      <td>feature_1:1327763-1328896</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>361</td>\n",
       "      <td>sv_1:1328245-1328606</td>\n",
       "      <td>['000007143790', '000007094630']</td>\n",
       "      <td>2</td>\n",
       "      <td>[121, 88]</td>\n",
       "      <td>88</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>121</td>\n",
       "      <td>[34, 26]</td>\n",
       "      <td>26</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>34</td>\n",
       "      <td>[0.2809917355371901, 0.29545454545454547]</td>\n",
       "      <td>0.2809917355</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2954545455</td>\n",
       "      <td>361</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>join_1:1328245-1328606</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-1843</td>\n",
       "      <td>MIMAT0039764</td>\n",
       "      <td>CPTP</td>\n",
       "      <td>ENSG00000224051</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>ENST00000343938</td>\n",
       "      <td>1328265</td>\n",
       "      <td>1328277</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Primary Visual Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30455455.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4500000000</td>\n",
       "      <td>Gestation week 16.5</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1328265-1328277:hsa-miR-1843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1327763</td>\n",
       "      <td>1328896</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 8)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1133</td>\n",
       "      <td>feature_1:1327763-1328896</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>361</td>\n",
       "      <td>sv_1:1328245-1328606</td>\n",
       "      <td>['000007143790', '000007094630']</td>\n",
       "      <td>2</td>\n",
       "      <td>[121, 88]</td>\n",
       "      <td>88</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>121</td>\n",
       "      <td>[34, 26]</td>\n",
       "      <td>26</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>34</td>\n",
       "      <td>[0.2809917355371901, 0.29545454545454547]</td>\n",
       "      <td>0.2809917355</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2954545455</td>\n",
       "      <td>361</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>join_1:1328245-1328606</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-148a-3p</td>\n",
       "      <td>MIMAT0000243</td>\n",
       "      <td>CPTP</td>\n",
       "      <td>ENSG00000224051</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>ENST00000343938</td>\n",
       "      <td>1328306</td>\n",
       "      <td>1328318</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Intestine</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>26701625.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>0.3900000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1328306-1328318:hsa-miR-148a-3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1327763</td>\n",
       "      <td>1328896</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 8)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1133</td>\n",
       "      <td>feature_1:1327763-1328896</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>361</td>\n",
       "      <td>sv_1:1328245-1328606</td>\n",
       "      <td>['000007143790', '000007094630']</td>\n",
       "      <td>2</td>\n",
       "      <td>[121, 88]</td>\n",
       "      <td>88</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>121</td>\n",
       "      <td>[34, 26]</td>\n",
       "      <td>26</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>34</td>\n",
       "      <td>[0.2809917355371901, 0.29545454545454547]</td>\n",
       "      <td>0.2809917355</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2954545455</td>\n",
       "      <td>361</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>join_1:1328245-1328606</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-148b-3p</td>\n",
       "      <td>MIMAT0000759</td>\n",
       "      <td>CPTP</td>\n",
       "      <td>ENSG00000224051</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>ENST00000343938</td>\n",
       "      <td>1328306</td>\n",
       "      <td>1328318</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Intestine</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>26701625.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>0.3900000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1328306-1328318:hsa-miR-148b-3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1327763</td>\n",
       "      <td>1328896</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 8)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1133</td>\n",
       "      <td>feature_1:1327763-1328896</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>361</td>\n",
       "      <td>sv_1:1328245-1328606</td>\n",
       "      <td>['000007143790', '000007094630']</td>\n",
       "      <td>2</td>\n",
       "      <td>[121, 88]</td>\n",
       "      <td>88</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>121</td>\n",
       "      <td>[34, 26]</td>\n",
       "      <td>26</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>34</td>\n",
       "      <td>[0.2809917355371901, 0.29545454545454547]</td>\n",
       "      <td>0.2809917355</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2954545455</td>\n",
       "      <td>361</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>join_1:1328245-1328606</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-101-3p</td>\n",
       "      <td>MIMAT0000099</td>\n",
       "      <td>CPTP</td>\n",
       "      <td>ENSG00000224051</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>ENST00000343938</td>\n",
       "      <td>1328307</td>\n",
       "      <td>1328319</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Intestine</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>26701625.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1328307-1328319:hsa-miR-101-3p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1327763</td>\n",
       "      <td>1328896</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000343938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2 (assigned to previous version 8)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1133</td>\n",
       "      <td>feature_1:1327763-1328896</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>361</td>\n",
       "      <td>sv_1:1328245-1328606</td>\n",
       "      <td>['000007143790', '000007094630']</td>\n",
       "      <td>2</td>\n",
       "      <td>[121, 88]</td>\n",
       "      <td>88</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>104.5000000000</td>\n",
       "      <td>121</td>\n",
       "      <td>[34, 26]</td>\n",
       "      <td>26</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>30.0000000000</td>\n",
       "      <td>34</td>\n",
       "      <td>[0.2809917355371901, 0.29545454545454547]</td>\n",
       "      <td>0.2809917355</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2882231405</td>\n",
       "      <td>0.2954545455</td>\n",
       "      <td>361</td>\n",
       "      <td>1328245</td>\n",
       "      <td>1328606</td>\n",
       "      <td>join_1:1328245-1328606</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-423-5p</td>\n",
       "      <td>MIMAT0004748</td>\n",
       "      <td>CPTP</td>\n",
       "      <td>ENSG00000224051</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>CPTP-201</td>\n",
       "      <td>ENST00000343938</td>\n",
       "      <td>1328344</td>\n",
       "      <td>1328356</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>33406413.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>0.5900000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_1:1328344-1328356:hsa-miR-423-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2281</th>\n",
       "      <td>X</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>151924618</td>\n",
       "      <td>151925170</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000276344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGEA4-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000276344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>552</td>\n",
       "      <td>feature_X:151924618-151925170</td>\n",
       "      <td>151925098</td>\n",
       "      <td>151927335</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>2237</td>\n",
       "      <td>sv_X:151925098-151927335</td>\n",
       "      <td>['000007090080', '000007140250']</td>\n",
       "      <td>2</td>\n",
       "      <td>[29, 153]</td>\n",
       "      <td>29</td>\n",
       "      <td>91.0000000000</td>\n",
       "      <td>91.0000000000</td>\n",
       "      <td>153</td>\n",
       "      <td>[29, 53]</td>\n",
       "      <td>29</td>\n",
       "      <td>41.0000000000</td>\n",
       "      <td>41.0000000000</td>\n",
       "      <td>53</td>\n",
       "      <td>[1.0, 0.3464052287581699]</td>\n",
       "      <td>0.3464052288</td>\n",
       "      <td>0.6732026144</td>\n",
       "      <td>0.6732026144</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>72</td>\n",
       "      <td>151925098</td>\n",
       "      <td>151925170</td>\n",
       "      <td>join_X:151925098-151925170</td>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-191-5p</td>\n",
       "      <td>MIMAT0000440</td>\n",
       "      <td>MAGEA4</td>\n",
       "      <td>ENSG00000147381</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000360243</td>\n",
       "      <td>151925106</td>\n",
       "      <td>151925118</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>27292025.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:151925106-151925118:hsa-miR-191-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2281</th>\n",
       "      <td>X</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>151924618</td>\n",
       "      <td>151925170</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000276344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MAGEA4-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000276344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>552</td>\n",
       "      <td>feature_X:151924618-151925170</td>\n",
       "      <td>151925098</td>\n",
       "      <td>151927335</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>2237</td>\n",
       "      <td>sv_X:151925098-151927335</td>\n",
       "      <td>['000007090080', '000007140250']</td>\n",
       "      <td>2</td>\n",
       "      <td>[29, 153]</td>\n",
       "      <td>29</td>\n",
       "      <td>91.0000000000</td>\n",
       "      <td>91.0000000000</td>\n",
       "      <td>153</td>\n",
       "      <td>[29, 53]</td>\n",
       "      <td>29</td>\n",
       "      <td>41.0000000000</td>\n",
       "      <td>41.0000000000</td>\n",
       "      <td>53</td>\n",
       "      <td>[1.0, 0.3464052287581699]</td>\n",
       "      <td>0.3464052288</td>\n",
       "      <td>0.6732026144</td>\n",
       "      <td>0.6732026144</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>72</td>\n",
       "      <td>151925098</td>\n",
       "      <td>151925170</td>\n",
       "      <td>join_X:151925098-151925170</td>\n",
       "      <td>sv_right_free</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-181a-5p</td>\n",
       "      <td>MIMAT0000256</td>\n",
       "      <td>MAGEA4</td>\n",
       "      <td>ENSG00000147381</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENST00000360243</td>\n",
       "      <td>151925120</td>\n",
       "      <td>151925132</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>27292025.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>Epithelial cells</td>\n",
       "      <td>0.6000000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:151925120-151925132:hsa-miR-181a-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>153508445</td>\n",
       "      <td>153509546</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000331595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BGN-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000331595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 8)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1101</td>\n",
       "      <td>feature_X:153508445-153509546</td>\n",
       "      <td>153509176</td>\n",
       "      <td>153509287</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>111</td>\n",
       "      <td>sv_X:153509176-153509287</td>\n",
       "      <td>['000027009070', '000028000360', '000028000370', '000029004340', '000029003420', '000029003430', '000007202770']</td>\n",
       "      <td>7</td>\n",
       "      <td>[77, 63, 11, 40, 38, 16, 44]</td>\n",
       "      <td>11</td>\n",
       "      <td>40.0000000000</td>\n",
       "      <td>41.2857142857</td>\n",
       "      <td>77</td>\n",
       "      <td>[18, 18, 11, 13, 7, 16, 8]</td>\n",
       "      <td>7</td>\n",
       "      <td>13.0000000000</td>\n",
       "      <td>13.0000000000</td>\n",
       "      <td>18</td>\n",
       "      <td>[0.23376623376623376, 0.2857142857142857, 1.0, 0.325, 0.18421052631578946, 1.0, 0.18181818181818182]</td>\n",
       "      <td>0.1818181818</td>\n",
       "      <td>0.2857142857</td>\n",
       "      <td>0.4586441754</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>111</td>\n",
       "      <td>153509176</td>\n",
       "      <td>153509287</td>\n",
       "      <td>join_X:153509176-153509287</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-423-5p</td>\n",
       "      <td>MIMAT0004748</td>\n",
       "      <td>BGN</td>\n",
       "      <td>ENSG00000182492</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>BGN-201</td>\n",
       "      <td>ENST00000331595</td>\n",
       "      <td>153509179</td>\n",
       "      <td>153509191</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5700000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:153509179-153509191:hsa-miR-423-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2285</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055979</td>\n",
       "      <td>155056738</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>759</td>\n",
       "      <td>sv_X:155055979-155056738</td>\n",
       "      <td>['000007090320', '000027021000']</td>\n",
       "      <td>2</td>\n",
       "      <td>[108, 48]</td>\n",
       "      <td>48</td>\n",
       "      <td>78.0000000000</td>\n",
       "      <td>78.0000000000</td>\n",
       "      <td>108</td>\n",
       "      <td>[50, 48]</td>\n",
       "      <td>48</td>\n",
       "      <td>49.0000000000</td>\n",
       "      <td>49.0000000000</td>\n",
       "      <td>50</td>\n",
       "      <td>[0.46296296296296297, 1.0]</td>\n",
       "      <td>0.4629629630</td>\n",
       "      <td>0.7314814815</td>\n",
       "      <td>0.7314814815</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>759</td>\n",
       "      <td>155055979</td>\n",
       "      <td>155056738</td>\n",
       "      <td>join_X:155055979-155056738</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-32-5p</td>\n",
       "      <td>MIMAT0000090</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155056053</td>\n",
       "      <td>155056065</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4200000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155056053-155056065:hsa-miR-32-5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2285</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>155054672</td>\n",
       "      <td>155060304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcript:ENST00000369498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>1 (assigned to previous version 7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5632</td>\n",
       "      <td>feature_X:155054672-155060304</td>\n",
       "      <td>155055979</td>\n",
       "      <td>155056738</td>\n",
       "      <td>DEL</td>\n",
       "      <td>manta</td>\n",
       "      <td>759</td>\n",
       "      <td>sv_X:155055979-155056738</td>\n",
       "      <td>['000007090320', '000027021000']</td>\n",
       "      <td>2</td>\n",
       "      <td>[108, 48]</td>\n",
       "      <td>48</td>\n",
       "      <td>78.0000000000</td>\n",
       "      <td>78.0000000000</td>\n",
       "      <td>108</td>\n",
       "      <td>[50, 48]</td>\n",
       "      <td>48</td>\n",
       "      <td>49.0000000000</td>\n",
       "      <td>49.0000000000</td>\n",
       "      <td>50</td>\n",
       "      <td>[0.46296296296296297, 1.0]</td>\n",
       "      <td>0.4629629630</td>\n",
       "      <td>0.7314814815</td>\n",
       "      <td>0.7314814815</td>\n",
       "      <td>1.0000000000</td>\n",
       "      <td>759</td>\n",
       "      <td>155055979</td>\n",
       "      <td>155056738</td>\n",
       "      <td>join_X:155055979-155056738</td>\n",
       "      <td>sv_in_feature</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-miR-92b-3p</td>\n",
       "      <td>MIMAT0003218</td>\n",
       "      <td>FUNDC2</td>\n",
       "      <td>ENSG00000165775</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>FUNDC2-201</td>\n",
       "      <td>ENST00000369498</td>\n",
       "      <td>155056053</td>\n",
       "      <td>155056065</td>\n",
       "      <td>+</td>\n",
       "      <td>HITS-CLIP</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Brain - Motor cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24389009.0000000000</td>\n",
       "      <td>1</td>\n",
       "      <td>secondary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5000000000</td>\n",
       "      <td>Male - 44 years</td>\n",
       "      <td>12</td>\n",
       "      <td>tarbase_X:155056053-155056065:hsa-miR-92b-3p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22692 rows × 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome          Source          Feature      Start        End Score  \\\n",
       "0             1  ensembl_havana  three_prime_UTR    1327763    1328896     .   \n",
       "0             1  ensembl_havana  three_prime_UTR    1327763    1328896     .   \n",
       "0             1  ensembl_havana  three_prime_UTR    1327763    1328896     .   \n",
       "0             1  ensembl_havana  three_prime_UTR    1327763    1328896     .   \n",
       "0             1  ensembl_havana  three_prime_UTR    1327763    1328896     .   \n",
       "...         ...             ...              ...        ...        ...   ...   \n",
       "2281          X          havana  three_prime_UTR  151924618  151925170     .   \n",
       "2281          X          havana  three_prime_UTR  151924618  151925170     .   \n",
       "2284          X  ensembl_havana  three_prime_UTR  153508445  153509546     .   \n",
       "2285          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "2285          X  ensembl_havana  three_prime_UTR  155054672  155060304     .   \n",
       "\n",
       "     Strand Frame                          ID Alias external_name logic_name  \\\n",
       "0         +     .  transcript:ENST00000343938   NaN           NaN        NaN   \n",
       "0         +     .  transcript:ENST00000343938   NaN           NaN        NaN   \n",
       "0         +     .  transcript:ENST00000343938   NaN           NaN        NaN   \n",
       "0         +     .  transcript:ENST00000343938   NaN           NaN        NaN   \n",
       "0         +     .  transcript:ENST00000343938   NaN           NaN        NaN   \n",
       "...     ...   ...                         ...   ...           ...        ...   \n",
       "2281      +     .  transcript:ENST00000276344   NaN           NaN        NaN   \n",
       "2281      +     .  transcript:ENST00000276344   NaN           NaN        NaN   \n",
       "2284      +     .  transcript:ENST00000331595   NaN           NaN        NaN   \n",
       "2285      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "2285      +     .  transcript:ENST00000369498   NaN           NaN        NaN   \n",
       "\n",
       "            Name         biotype description gene_id version  \\\n",
       "0       CPTP-201  protein_coding         NaN     NaN     NaN   \n",
       "0       CPTP-201  protein_coding         NaN     NaN     NaN   \n",
       "0       CPTP-201  protein_coding         NaN     NaN     NaN   \n",
       "0       CPTP-201  protein_coding         NaN     NaN     NaN   \n",
       "0       CPTP-201  protein_coding         NaN     NaN     NaN   \n",
       "...          ...             ...         ...     ...     ...   \n",
       "2281  MAGEA4-201  protein_coding         NaN     NaN     NaN   \n",
       "2281  MAGEA4-201  protein_coding         NaN     NaN     NaN   \n",
       "2284     BGN-201  protein_coding         NaN     NaN     NaN   \n",
       "2285  FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "2285  FUNDC2-201  protein_coding         NaN     NaN     NaN   \n",
       "\n",
       "                          Parent transcript_id constitutive ensembl_end_phase  \\\n",
       "0     transcript:ENST00000343938           NaN          NaN               NaN   \n",
       "0     transcript:ENST00000343938           NaN          NaN               NaN   \n",
       "0     transcript:ENST00000343938           NaN          NaN               NaN   \n",
       "0     transcript:ENST00000343938           NaN          NaN               NaN   \n",
       "0     transcript:ENST00000343938           NaN          NaN               NaN   \n",
       "...                          ...           ...          ...               ...   \n",
       "2281  transcript:ENST00000276344           NaN          NaN               NaN   \n",
       "2281  transcript:ENST00000276344           NaN          NaN               NaN   \n",
       "2284  transcript:ENST00000331595           NaN          NaN               NaN   \n",
       "2285  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "2285  transcript:ENST00000369498           NaN          NaN               NaN   \n",
       "\n",
       "     ensembl_phase exon_id rank  \\\n",
       "0              NaN     NaN  NaN   \n",
       "0              NaN     NaN  NaN   \n",
       "0              NaN     NaN  NaN   \n",
       "0              NaN     NaN  NaN   \n",
       "0              NaN     NaN  NaN   \n",
       "...            ...     ...  ...   \n",
       "2281           NaN     NaN  NaN   \n",
       "2281           NaN     NaN  NaN   \n",
       "2284           NaN     NaN  NaN   \n",
       "2285           NaN     NaN  NaN   \n",
       "2285           NaN     NaN  NaN   \n",
       "\n",
       "                                                              tag  \\\n",
       "0     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "0     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "0     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "0     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "0     gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "...                                                           ...   \n",
       "2281  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "2281  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "2284  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "2285  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "2285  gencode_basic,gencode_primary,Ensembl_canonical,MANE_Select   \n",
       "\n",
       "                transcript_support_level ccdsid protein_id  feature_len  \\\n",
       "0     2 (assigned to previous version 8)    NaN        NaN         1133   \n",
       "0     2 (assigned to previous version 8)    NaN        NaN         1133   \n",
       "0     2 (assigned to previous version 8)    NaN        NaN         1133   \n",
       "0     2 (assigned to previous version 8)    NaN        NaN         1133   \n",
       "0     2 (assigned to previous version 8)    NaN        NaN         1133   \n",
       "...                                  ...    ...        ...          ...   \n",
       "2281                                   2    NaN        NaN          552   \n",
       "2281                                   2    NaN        NaN          552   \n",
       "2284  1 (assigned to previous version 8)    NaN        NaN         1101   \n",
       "2285  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "2285  1 (assigned to previous version 7)    NaN        NaN         5632   \n",
       "\n",
       "                         feature_id   Start_sv     End_sv SVTYPE caller  \\\n",
       "0         feature_1:1327763-1328896    1328245    1328606    DEL  manta   \n",
       "0         feature_1:1327763-1328896    1328245    1328606    DEL  manta   \n",
       "0         feature_1:1327763-1328896    1328245    1328606    DEL  manta   \n",
       "0         feature_1:1327763-1328896    1328245    1328606    DEL  manta   \n",
       "0         feature_1:1327763-1328896    1328245    1328606    DEL  manta   \n",
       "...                             ...        ...        ...    ...    ...   \n",
       "2281  feature_X:151924618-151925170  151925098  151927335    DEL  manta   \n",
       "2281  feature_X:151924618-151925170  151925098  151927335    DEL  manta   \n",
       "2284  feature_X:153508445-153509546  153509176  153509287    DEL  manta   \n",
       "2285  feature_X:155054672-155060304  155055979  155056738    DEL  manta   \n",
       "2285  feature_X:155054672-155060304  155055979  155056738    DEL  manta   \n",
       "\n",
       "      sv_len                     sv_id  \\\n",
       "0        361      sv_1:1328245-1328606   \n",
       "0        361      sv_1:1328245-1328606   \n",
       "0        361      sv_1:1328245-1328606   \n",
       "0        361      sv_1:1328245-1328606   \n",
       "0        361      sv_1:1328245-1328606   \n",
       "...      ...                       ...   \n",
       "2281    2237  sv_X:151925098-151927335   \n",
       "2281    2237  sv_X:151925098-151927335   \n",
       "2284     111  sv_X:153509176-153509287   \n",
       "2285     759  sv_X:155055979-155056738   \n",
       "2285     759  sv_X:155055979-155056738   \n",
       "\n",
       "                                                                                                              SID_list  \\\n",
       "0                                                                                     ['000007143790', '000007094630']   \n",
       "0                                                                                     ['000007143790', '000007094630']   \n",
       "0                                                                                     ['000007143790', '000007094630']   \n",
       "0                                                                                     ['000007143790', '000007094630']   \n",
       "0                                                                                     ['000007143790', '000007094630']   \n",
       "...                                                                                                                ...   \n",
       "2281                                                                                  ['000007090080', '000007140250']   \n",
       "2281                                                                                  ['000007090080', '000007140250']   \n",
       "2284  ['000027009070', '000028000360', '000028000370', '000029004340', '000029003420', '000029003430', '000007202770']   \n",
       "2285                                                                                  ['000007090320', '000027021000']   \n",
       "2285                                                                                  ['000007090320', '000027021000']   \n",
       "\n",
       "      SID_count                       DP_list  DP_min      DP_median  \\\n",
       "0             2                     [121, 88]      88 104.5000000000   \n",
       "0             2                     [121, 88]      88 104.5000000000   \n",
       "0             2                     [121, 88]      88 104.5000000000   \n",
       "0             2                     [121, 88]      88 104.5000000000   \n",
       "0             2                     [121, 88]      88 104.5000000000   \n",
       "...         ...                           ...     ...            ...   \n",
       "2281          2                     [29, 153]      29  91.0000000000   \n",
       "2281          2                     [29, 153]      29  91.0000000000   \n",
       "2284          7  [77, 63, 11, 40, 38, 16, 44]      11  40.0000000000   \n",
       "2285          2                     [108, 48]      48  78.0000000000   \n",
       "2285          2                     [108, 48]      48  78.0000000000   \n",
       "\n",
       "            DP_mean  DP_max                     AD_list  AD_min     AD_median  \\\n",
       "0    104.5000000000     121                    [34, 26]      26 30.0000000000   \n",
       "0    104.5000000000     121                    [34, 26]      26 30.0000000000   \n",
       "0    104.5000000000     121                    [34, 26]      26 30.0000000000   \n",
       "0    104.5000000000     121                    [34, 26]      26 30.0000000000   \n",
       "0    104.5000000000     121                    [34, 26]      26 30.0000000000   \n",
       "...             ...     ...                         ...     ...           ...   \n",
       "2281  91.0000000000     153                    [29, 53]      29 41.0000000000   \n",
       "2281  91.0000000000     153                    [29, 53]      29 41.0000000000   \n",
       "2284  41.2857142857      77  [18, 18, 11, 13, 7, 16, 8]       7 13.0000000000   \n",
       "2285  78.0000000000     108                    [50, 48]      48 49.0000000000   \n",
       "2285  78.0000000000     108                    [50, 48]      48 49.0000000000   \n",
       "\n",
       "           AD_mean  AD_max  \\\n",
       "0    30.0000000000      34   \n",
       "0    30.0000000000      34   \n",
       "0    30.0000000000      34   \n",
       "0    30.0000000000      34   \n",
       "0    30.0000000000      34   \n",
       "...            ...     ...   \n",
       "2281 41.0000000000      53   \n",
       "2281 41.0000000000      53   \n",
       "2284 13.0000000000      18   \n",
       "2285 49.0000000000      50   \n",
       "2285 49.0000000000      50   \n",
       "\n",
       "                                                                                                   AF_list  \\\n",
       "0                                                                [0.2809917355371901, 0.29545454545454547]   \n",
       "0                                                                [0.2809917355371901, 0.29545454545454547]   \n",
       "0                                                                [0.2809917355371901, 0.29545454545454547]   \n",
       "0                                                                [0.2809917355371901, 0.29545454545454547]   \n",
       "0                                                                [0.2809917355371901, 0.29545454545454547]   \n",
       "...                                                                                                    ...   \n",
       "2281                                                                             [1.0, 0.3464052287581699]   \n",
       "2281                                                                             [1.0, 0.3464052287581699]   \n",
       "2284  [0.23376623376623376, 0.2857142857142857, 1.0, 0.325, 0.18421052631578946, 1.0, 0.18181818181818182]   \n",
       "2285                                                                            [0.46296296296296297, 1.0]   \n",
       "2285                                                                            [0.46296296296296297, 1.0]   \n",
       "\n",
       "           AF_min    AF_median      AF_mean       AF_max  join_len  \\\n",
       "0    0.2809917355 0.2882231405 0.2882231405 0.2954545455       361   \n",
       "0    0.2809917355 0.2882231405 0.2882231405 0.2954545455       361   \n",
       "0    0.2809917355 0.2882231405 0.2882231405 0.2954545455       361   \n",
       "0    0.2809917355 0.2882231405 0.2882231405 0.2954545455       361   \n",
       "0    0.2809917355 0.2882231405 0.2882231405 0.2954545455       361   \n",
       "...           ...          ...          ...          ...       ...   \n",
       "2281 0.3464052288 0.6732026144 0.6732026144 1.0000000000        72   \n",
       "2281 0.3464052288 0.6732026144 0.6732026144 1.0000000000        72   \n",
       "2284 0.1818181818 0.2857142857 0.4586441754 1.0000000000       111   \n",
       "2285 0.4629629630 0.7314814815 0.7314814815 1.0000000000       759   \n",
       "2285 0.4629629630 0.7314814815 0.7314814815 1.0000000000       759   \n",
       "\n",
       "      join_start   join_end                     join_id      join_type  \\\n",
       "0        1328245    1328606      join_1:1328245-1328606  sv_in_feature   \n",
       "0        1328245    1328606      join_1:1328245-1328606  sv_in_feature   \n",
       "0        1328245    1328606      join_1:1328245-1328606  sv_in_feature   \n",
       "0        1328245    1328606      join_1:1328245-1328606  sv_in_feature   \n",
       "0        1328245    1328606      join_1:1328245-1328606  sv_in_feature   \n",
       "...          ...        ...                         ...            ...   \n",
       "2281   151925098  151925170  join_X:151925098-151925170  sv_right_free   \n",
       "2281   151925098  151925170  join_X:151925098-151925170  sv_right_free   \n",
       "2284   153509176  153509287  join_X:153509176-153509287  sv_in_feature   \n",
       "2285   155055979  155056738  join_X:155055979-155056738  sv_in_feature   \n",
       "2285   155055979  155056738  join_X:155055979-155056738  sv_in_feature   \n",
       "\n",
       "      joins_count       species       mirna_name      mirna_id gene_name  \\\n",
       "0               1  Homo sapiens     hsa-miR-1843  MIMAT0039764      CPTP   \n",
       "0               1  Homo sapiens  hsa-miR-148a-3p  MIMAT0000243      CPTP   \n",
       "0               1  Homo sapiens  hsa-miR-148b-3p  MIMAT0000759      CPTP   \n",
       "0               1  Homo sapiens   hsa-miR-101-3p  MIMAT0000099      CPTP   \n",
       "0               1  Homo sapiens   hsa-miR-423-5p  MIMAT0004748      CPTP   \n",
       "...           ...           ...              ...           ...       ...   \n",
       "2281            1  Homo sapiens   hsa-miR-191-5p  MIMAT0000440    MAGEA4   \n",
       "2281            1  Homo sapiens  hsa-miR-181a-5p  MIMAT0000256    MAGEA4   \n",
       "2284            1  Homo sapiens   hsa-miR-423-5p  MIMAT0004748       BGN   \n",
       "2285            1  Homo sapiens    hsa-miR-32-5p  MIMAT0000090    FUNDC2   \n",
       "2285            1  Homo sapiens   hsa-miR-92b-3p  MIMAT0003218    FUNDC2   \n",
       "\n",
       "      gene_id_tarbase gene_location transcript_name transcript_id_tarbase  \\\n",
       "0     ENSG00000224051          3UTR        CPTP-201       ENST00000343938   \n",
       "0     ENSG00000224051          3UTR        CPTP-201       ENST00000343938   \n",
       "0     ENSG00000224051          3UTR        CPTP-201       ENST00000343938   \n",
       "0     ENSG00000224051          3UTR        CPTP-201       ENST00000343938   \n",
       "0     ENSG00000224051          3UTR        CPTP-201       ENST00000343938   \n",
       "...               ...           ...             ...                   ...   \n",
       "2281  ENSG00000147381          3UTR             NaN       ENST00000360243   \n",
       "2281  ENSG00000147381          3UTR             NaN       ENST00000360243   \n",
       "2284  ENSG00000182492          3UTR         BGN-201       ENST00000331595   \n",
       "2285  ENSG00000165775          3UTR      FUNDC2-201       ENST00000369498   \n",
       "2285  ENSG00000165775          3UTR      FUNDC2-201       ENST00000369498   \n",
       "\n",
       "      Start_tarbase  End_tarbase Strand_tarbase experimental_method  \\\n",
       "0           1328265      1328277              +           HITS-CLIP   \n",
       "0           1328306      1328318              +            PAR-CLIP   \n",
       "0           1328306      1328318              +            PAR-CLIP   \n",
       "0           1328307      1328319              +            PAR-CLIP   \n",
       "0           1328344      1328356              +           HITS-CLIP   \n",
       "...             ...          ...            ...                 ...   \n",
       "2281      151925106    151925118              +            PAR-CLIP   \n",
       "2281      151925120    151925132              +            PAR-CLIP   \n",
       "2284      153509179    153509191              +           HITS-CLIP   \n",
       "2285      155056053    155056065              +           HITS-CLIP   \n",
       "2285      155056053    155056065              +           HITS-CLIP   \n",
       "\n",
       "     regulation                         tissue cell_line   article_pubmed_id  \\\n",
       "0      Negative  Brain - Primary Visual Cortex       NaN 30455455.0000000000   \n",
       "0      Negative                      Intestine    HCT116 26701625.0000000000   \n",
       "0      Negative                      Intestine    HCT116 26701625.0000000000   \n",
       "0      Negative                      Intestine    HCT116 26701625.0000000000   \n",
       "0      Negative                       Prostate     22RV1 33406413.0000000000   \n",
       "...         ...                            ...       ...                 ...   \n",
       "2281   Negative                       Prostate     22RV1 27292025.0000000000   \n",
       "2281   Negative                       Prostate     22RV1 27292025.0000000000   \n",
       "2284   Negative           Brain - Motor cortex       NaN 24389009.0000000000   \n",
       "2285   Negative           Brain - Motor cortex       NaN 24389009.0000000000   \n",
       "2285   Negative           Brain - Motor cortex       NaN 24389009.0000000000   \n",
       "\n",
       "      confidence interaction_group         cell_type  microt_score  \\\n",
       "0              1           primary               NaN  0.4500000000   \n",
       "0              1           primary  Epithelial cells  0.3900000000   \n",
       "0              1           primary  Epithelial cells  0.3900000000   \n",
       "0              1           primary  Epithelial cells           NaN   \n",
       "0              1           primary  Epithelial cells  0.5900000000   \n",
       "...          ...               ...               ...           ...   \n",
       "2281           1           primary  Epithelial cells           NaN   \n",
       "2281           1           primary  Epithelial cells  0.6000000000   \n",
       "2284           1           primary               NaN  0.5700000000   \n",
       "2285           1           primary               NaN  0.4200000000   \n",
       "2285           1         secondary               NaN  0.5000000000   \n",
       "\n",
       "                  comment  tarbase_join_len  \\\n",
       "0     Gestation week 16.5                12   \n",
       "0                     NaN                12   \n",
       "0                     NaN                12   \n",
       "0                     NaN                12   \n",
       "0                     NaN                12   \n",
       "...                   ...               ...   \n",
       "2281                  NaN                12   \n",
       "2281                  NaN                12   \n",
       "2284      Male - 44 years                12   \n",
       "2285      Male - 44 years                12   \n",
       "2285      Male - 44 years                12   \n",
       "\n",
       "                                         tarbase_id  \n",
       "0            tarbase_1:1328265-1328277:hsa-miR-1843  \n",
       "0         tarbase_1:1328306-1328318:hsa-miR-148a-3p  \n",
       "0         tarbase_1:1328306-1328318:hsa-miR-148b-3p  \n",
       "0          tarbase_1:1328307-1328319:hsa-miR-101-3p  \n",
       "0          tarbase_1:1328344-1328356:hsa-miR-423-5p  \n",
       "...                                             ...  \n",
       "2281   tarbase_X:151925106-151925118:hsa-miR-191-5p  \n",
       "2281  tarbase_X:151925120-151925132:hsa-miR-181a-5p  \n",
       "2284   tarbase_X:153509179-153509191:hsa-miR-423-5p  \n",
       "2285    tarbase_X:155056053-155056065:hsa-miR-32-5p  \n",
       "2285   tarbase_X:155056053-155056065:hsa-miR-92b-3p  \n",
       "\n",
       "[22692 rows x 82 columns]"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(features_sv_no_offtereget_tarbase_joined_pr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1dec50d",
   "metadata": {},
   "source": [
    "### <span style=\"color:#00ff00;\">GFF output for IGV</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1769e350",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/2115558342.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  output_features_pr['Source'] = 'GFF'\n",
      "/var/folders/rf/k3pdmpqx0bjfvwyn2q6sw5v00000gn/T/ipykernel_3125/2115558342.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  output_features_pr['color'] = 'blue'\n"
     ]
    }
   ],
   "source": [
    "pr_for_igv = features_sv_no_offtereget_tarbase_joined_pr\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Add GFF\n",
    "output_features_pr = pr_for_igv[['Chromosome', 'Feature', 'Start', 'End', 'Strand', 'Name', 'biotype', 'tag', 'ID']]\n",
    "output_features_pr = output_features_pr.drop_duplicates()\n",
    "output_features_pr['Source'] = 'GFF'\n",
    "output_features_pr['color'] = 'blue'\n",
    "\n",
    "# Add gnomAD SV\n",
    "'''output_sv_pr = pr_for_igv[['Chromosome', 'Start_sv', 'End_sv', 'svtype', 'name', 'sv_id', 'AF', 'N_HET', 'N_HOMALT']]\n",
    "output_sv_pr = output_sv_pr.drop_duplicates()\n",
    "output_sv_pr['Source'] = 'gnomAD_SV'\n",
    "output_sv_pr['color'] = 'red'\n",
    "output_sv_pr = pr.PyRanges(output_sv_pr.rename(columns={'Start_sv': 'Start', 'End_sv': 'End', 'svtype': 'Feature', 'name': 'Name', 'sv_id': 'ID'}))'''\n",
    "\n",
    "# Add RCMG BTK SV\n",
    "output_sv_pr = pr_for_igv[['Chromosome', 'Start_sv', 'End_sv', 'SVTYPE', 'SID_count', 'DP_median', 'AF_median', 'SID_list', 'sv_id']]\n",
    "output_sv_pr = output_sv_pr.drop_duplicates()\n",
    "output_sv_pr['Source'] = 'RCMG_BTK_SV'\n",
    "output_sv_pr['color'] = 'red'\n",
    "output_sv_pr = pr.PyRanges(output_sv_pr.rename(columns={'Start_sv': 'Start', 'End_sv': 'End', 'svtype': 'Feature', 'name': 'Name', 'sv_id': 'ID'}))\n",
    "\n",
    "# Add joins data\n",
    "output_joins_pr = pr_for_igv[['Chromosome', 'join_start', 'join_end', 'join_type', 'join_id']]\n",
    "output_joins_pr = output_joins_pr.drop_duplicates()\n",
    "output_joins_pr['Source'] = 'join'\n",
    "output_joins_pr['color'] = 'orange'\n",
    "output_joins_pr = pr.PyRanges(output_joins_pr.rename(columns={'join_start': 'Start', 'join_end': 'End', 'join_type': 'Feature', 'join_id': 'ID'}))\n",
    "\n",
    "# Add miRNA TarBase\n",
    "output_tarbase_pr = pr_for_igv[['Chromosome', 'Start_tarbase', 'End_tarbase', 'Strand_tarbase', 'mirna_name', 'interaction_group', 'mirna_id', 'tarbase_id']]\n",
    "output_tarbase_pr = output_tarbase_pr.drop_duplicates(subset=['mirna_id', 'tarbase_id'])\n",
    "output_tarbase_pr['Source'] = 'tarbase'\n",
    "output_tarbase_pr['color'] = 'green'\n",
    "output_tarbase_pr = pr.PyRanges(output_tarbase_pr.rename(columns={'Start_tarbase': 'Start', 'End_tarbase': 'End', 'Strand_tarbase': 'Strand', 'interaction_group': 'Feature', 'mirna_name': 'Name', 'tarbase_id': 'ID'}))\n",
    "\n",
    "# Add offtarget GFF (features) data\n",
    "output_offtargets_pr = features_sv_offtereget_joined_pr[['Chromosome', 'Start_offtarget', 'End_offtarget', 'Feature_offtarget', 'offtarget_id']]\n",
    "output_offtargets_pr = output_offtargets_pr.query('Start_offtarget == Start_offtarget')\n",
    "output_offtargets_pr = output_offtargets_pr.drop_duplicates()\n",
    "output_offtargets_pr['Source'] = 'GFF_offtarget'\n",
    "output_offtargets_pr['color'] = '#d9d9d9'\n",
    "output_offtargets_pr = pr.PyRanges(output_offtargets_pr.rename(columns={'Start_offtarget': 'Start', 'End_offtarget': 'End', 'Feature_offtarget': 'Feature', 'offtarget_id': 'ID'}))\n",
    "\n",
    "output_igv_pr = pd.concat([output_features_pr, output_sv_pr, output_joins_pr, output_tarbase_pr, output_offtargets_pr], ignore_index=True)\n",
    "\n",
    "# ==============================================================================================================================================\n",
    "\n",
    "# Fill NA and save\n",
    "output_igv_pr = output_igv_pr.reset_index(drop=True)\n",
    "#output_igv_pr = pr.PyRanges(output_igv_pr.fillna('.'))\n",
    "\n",
    "output_igv_pr.to_gff3(output_dir / 'regions_for_igv.gff3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b16cad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tarbase</td>\n",
       "      <td>10326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCMG_BTK_SV</td>\n",
       "      <td>802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GFF_offtarget</td>\n",
       "      <td>790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>join</td>\n",
       "      <td>774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFF</td>\n",
       "      <td>497</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Source  count\n",
       "0        tarbase  10326\n",
       "1    RCMG_BTK_SV    802\n",
       "2  GFF_offtarget    790\n",
       "3           join    774\n",
       "4            GFF    497"
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stats\n",
    "output_igv_pr['Source'].value_counts().to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541f18c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a29f19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "668dfa8e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "662de559",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e9d4e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1fd440d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "utr3.venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
